{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "from cobra.core import Gene, Metabolite, Reaction\n",
    "import cobra.test\n",
    "\n",
    "model_pf = cobra.io.read_sbml_model(\"/Users/maureencarey/local_documents/work/comparative_parasite_models/paradigm/data/2017_pf_model.xml\")\n",
    "#mapping = pd.read_table(\"/Users/maureencarey/local_documents/work/comparative_parasite_models/paradigm/data/orthology_mapping_jun22/pf3d7_plasmodb.txt\")\n",
    "mapping = pd.read_csv('/Users/maureencarey/local_documents/work/comparative_parasite_models/paradigm/data/plasmodium_gene_mapping.csv')\n",
    "annotations_dict = dict()\n",
    "columns = ['query_gene', 'BiGG_gene', 'pident', 'length', 'mismatch', 'gapopen','qstart', 'qend', 'sstart', 'send', 'evalue', 'score']\n",
    "path = \"/Users/maureencarey/local_documents/work/comparative_parasite_models/code/data/genomes_from_EuPathDB/version37/diamond_output/\"\n",
    "os.chdir(path)\n",
    "for filename in glob.glob(os.path.join(path, '*.tsv')):\n",
    "    annotations_dict[filename.split('/')[len(filename.split('/'))-1]] = pd.read_table(filename, sep = '\\t', names=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "### move to curation script\n",
    "# aliases = pd.read_csv('/Users/maureencarey/local_documents/work/comparative_parasite_models/paradigm/data/Pfalciparum3D7_GeneAliases.csv')\n",
    "# # update gene identifiers\n",
    "\n",
    "# for gene in model_pf.genes:\n",
    "#     gene_id = gene.id\n",
    "\n",
    "#     for column in aliases:\n",
    "#         if (gene_id == aliases[column]).any():\n",
    "#             indx = aliases[aliases[column] == gene_id].index.tolist()[0]\n",
    "#     gene.id = aliases['name1'][indx]\n",
    "#     if '.' in gene.id:\n",
    "#         gene.id = gene.id.split('.')[0]\n",
    "# model_pf.repair()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # modify spcies column in mapping dataframe to map to filenames NECESSARY IF READING TXT MAPPING FILE\n",
    "# file_names = [*annotations_dict]\n",
    "# file_names = [x.split('_')[0] for x in file_names]\n",
    "# mapping['[Organism]'] = [x.replace(\". \", \"\") for x in mapping['[Organism]']]\n",
    "# mapping['[Organism]']  = [x.replace(\"strain\", \"\") for x in mapping['[Organism]'] ]\n",
    "# mapping['[Organism]']  = [x.replace(\" \", \"\") for x in mapping['[Organism]'] ]\n",
    "# # mapping['[Organism]']  = [x+'_May20.tsv' for x in mapping['[Organism]'] ]\n",
    "# mapping['[Organism]']  = [x.replace(\"Pfragilenilgiri\", \"PfragileNilgiri\") for x in mapping['[Organism]'] ]\n",
    "# mapping['[Organism]']  = [x.replace(\"PknowlesiMalayanStrainPk1A\", \"PknowlesiMalayanPk1A\") for x in mapping['[Organism]'] ]\n",
    "# mapping['[Organism]']  = [x.replace(\"PvivaxSal-1\", \"PvivaxSal1\") for x in mapping['[Organism]'] ]\n",
    "# if [] != [x for x in mapping['[Organism]'] if x not in set(mapping['[Organism]'] ).intersection(set(file_names))]:\n",
    "#     print('filename issue')\n",
    "# mapping.to_csv('plasmodium_gene_mapping.csv', header=True, index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "add_these = ['pe_prod1', 'pe_prod10', 'pe_prod11', 'pe_prod12', 'pe_prod13', 'pe_prod14',\n",
    " 'pe_prod15', 'pe_prod16', 'pe_prod17', 'pe_prod18', 'pe_prod19', 'pe_prod20', 'pe_prod21', 'pe_prod22','pe_prod2',\n",
    "'pe_prod3','pe_prod4','pe_prod5','pe_prod6','pe_prod7',\n",
    " 'pe_prod8', 'pe_prod9', 'pg_prod1', 'pg_prod2', 'pg_prod3', 'pg_prod4', 'pg_prod5', 'pg_prod6', 'pg_prod7','pi_prod1',\n",
    " 'pi_prod2', 'pi_prod3','pi_prod4','progly_t', 'ps_prod1', 'ps_prod2', 'ps_prod3', 'ps_prod4', 'ps_prod5',\n",
    " 'ps_prod6', 'ps_prod7','lipid1', 'lipid10', 'lipid2', 'lipid3','lipid4', 'lipid5', 'lipid6', 'lipid7', 'lipid8',\n",
    " 'lipid9','PA120tap', 'PA140tap', 'PA141tap', 'PA160tap', 'PA161tap', 'PA180tap', 'PA181tap',\n",
    " '2agpe120_LSQBKT_e_RSQBKT__t', '2agpe140_LSQBKT_e_RSQBKT__t', '2agpe141_LSQBKT_e_RSQBKT__t',\n",
    " '2agpe160_LSQBKT_e_RSQBKT__t', '2agpe161_LSQBKT_e_RSQBKT__t', '2agpe180_LSQBKT_e_RSQBKT__t',\n",
    " '2agpe181_LSQBKT_e_RSQBKT__t', '2agpg120_LSQBKT_e_RSQBKT__t', '2agpg140_LSQBKT_e_RSQBKT__t',\n",
    " '2agpg141_LSQBKT_e_RSQBKT__t', '2agpg160_LSQBKT_e_RSQBKT__t', '2agpg161_LSQBKT_e_RSQBKT__t',\n",
    " '2agpg180_LSQBKT_e_RSQBKT__t', '2ddecg3p_LSQBKT_e_RSQBKT__t','acoa_prod1',\n",
    " 'acoa_prod10', 'acoa_prod11', 'acoa_prod12', 'acoa_prod13', 'acoa_prod14', 'acoa_prod15','acoa_prod16',\n",
    " 'acoa_prod2', 'acoa_prod3', 'acoa_prod4', 'acoa_prod5', 'acoa_prod6', 'acoa_prod7',\n",
    " 'acoa_prod8', 'acoa_prod9', 'acylpg_prod1', 'acylpg_prod2', 'acylpg_prod3', 'acylpg_prod4', 'acylpg_prod5',\n",
    " 'acylpg_prod6','acylpg_prod7', \n",
    " 'dolichol_t','TTDCAt','XOLEST2te','PCt','pail_uptake','pc_prod',\n",
    " 'Htap','FEROpp','Htmt','2agpg181_LSQBKT_e_RSQBKT__t','CHSTEROLt','HBtr','SM_host',\n",
    "             'HMBZex','Hfv','superox_ex']\n",
    "\n",
    "# bm models already have biomass_generic 'protein_t_e', 'dgl_prod1','dgl_prod2', 'Lipid_prod','dgl_prod3', 'dgl_prod4', 'dgl_prod5','dgl_prod6', 'dgl_prod7','Protein',"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "pf_model_dict = dict()\n",
    "path = \"/Users/maureencarey/local_documents/work/comparative_parasite_models/paradigm/data/\"\n",
    "os.chdir(path)\n",
    "for filename in glob.glob(os.path.join(path, '*.json')):\n",
    "    key = filename.split('/')[len(filename.split('/'))-1]\n",
    "    key = key[:-5]\n",
    "    if key.startswith('DIY7bm_P'):\n",
    "        key = key[7:]\n",
    "        key = key[:-10]\n",
    "        if key != 'PconfusumCUL13' and key != 'PneurophiliaMK1':\n",
    "            pf_model_dict[key] = cobra.io.load_json_model(filename) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PfalciparumIT\n",
      "1294 423 2717\n",
      "1406 423 2850\n",
      "1827 769 3121\n",
      "Pyoeliiyoelii17XNL\n",
      "1277 395 2727\n",
      "1389 395 2863\n",
      "1782 673 3128\n",
      "PinuiSanAntonio1\n",
      "1292 400 2723\n",
      "1404 400 2853\n",
      "1793 676 3118\n",
      "Pyoeliiyoelii17X\n",
      "1305 404 2736\n",
      "1417 404 2870\n",
      "1813 676 3139\n",
      "PknowlesiMalayanPk1A\n",
      "1317 403 2720\n",
      "1429 403 2854\n",
      "1822 671 3123\n",
      "PreichenowiCDC\n",
      "1294 418 2701\n",
      "1406 418 2834\n",
      "1803 720 3106\n",
      "PrelictumSGS1-like\n",
      "1324 410 2724\n",
      "1436 410 2857\n",
      "1835 685 3121\n",
      "PgaboniSY75\n",
      "1294 411 2700\n",
      "1406 411 2833\n",
      "1801 714 3102\n",
      "PbillcollinsiG01\n",
      "525 129 2204\n",
      "637 129 2356\n",
      "1144 490 2741\n",
      "PcoatneyiHackeri\n",
      "1293 402 2722\n",
      "1405 402 2857\n",
      "1795 672 3122\n",
      "PvivaxSal1\n",
      "1319 413 2712\n",
      "1431 413 2845\n",
      "1822 692 3113\n",
      "PadleriG01\n",
      "1294 415 2697\n",
      "1406 415 2830\n",
      "1801 731 3101\n",
      "PbergheiANKA\n",
      "1337 401 2759\n",
      "1449 401 2894\n",
      "1849 667 3158\n",
      "PvivaxP01\n",
      "1332 406 2726\n",
      "1444 406 2859\n",
      "1836 678 3127\n",
      "Pvinckeivinckeivinckei\n",
      "1329 405 2766\n",
      "1441 405 2902\n",
      "1832 677 3167\n",
      "PblacklockiG01\n",
      "1293 405 2704\n",
      "1405 405 2838\n",
      "1798 710 3108\n",
      "PpraefalciparumG01\n",
      "1291 414 2705\n",
      "1403 414 2838\n",
      "1802 712 3109\n",
      "PcynomolgiM\n",
      "1303 407 2712\n",
      "1415 407 2846\n",
      "1801 681 3113\n",
      "PgaboniG01\n",
      "1293 407 2692\n",
      "1405 407 2825\n",
      "1800 710 3092\n",
      "PcynomolgiB\n",
      "1252 382 2670\n",
      "1364 382 2804\n",
      "1748 652 3071\n",
      "Pchabaudichabaudi\n",
      "1339 415 2765\n",
      "1451 415 2900\n",
      "1845 686 3169\n",
      "PreichenowiG01\n",
      "1287 410 2698\n",
      "1399 410 2831\n",
      "1796 708 3103\n",
      "PovalecurtisiGH01\n",
      "1292 407 2713\n",
      "1404 407 2848\n",
      "1798 675 3116\n",
      "Pfalciparum3D7\n",
      "1296 422 2717\n",
      "1408 422 2850\n",
      "1492 497 2913\n",
      "PfragileNilgiri\n",
      "1320 405 2718\n",
      "1432 405 2851\n",
      "1813 678 3113\n",
      "PknowlesiH\n",
      "1324 408 2724\n",
      "1436 408 2858\n",
      "1830 679 3128\n",
      "PmalariaeUG01\n",
      "1291 402 2693\n",
      "1403 402 2826\n",
      "1804 682 3095\n",
      "PvinckeipetteriCR\n",
      "1355 413 2761\n",
      "1467 413 2891\n",
      "1849 685 3157\n",
      "PyoeliiyoeliiYM\n",
      "1303 403 2734\n",
      "1415 403 2868\n",
      "1811 675 3137\n",
      "Pgallinaceum8A\n",
      "1322 412 2721\n",
      "1434 412 2854\n",
      "1835 690 3120\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/maureencarey/.virtualenvs/cobrapy/lib/python3.6/site-packages/pandas/core/indexing.py:194 \u001b[1;31mSettingWithCopyWarning\u001b[0m: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "columns = ['species','starting_genes','reactions_added', 'mets_added','genes_added']\n",
    "modifications_ortho = pd.DataFrame(index = pf_model_dict.keys(), columns=columns)\n",
    "path = \"/Users/maureencarey/local_documents/work/comparative_parasite_models/paradigm/data/\"\n",
    "os.chdir(path)\n",
    "\n",
    "# keep bounds from universal model\n",
    "\n",
    "for species, model in pf_model_dict.items():\n",
    "\n",
    "    print(species)\n",
    "    x1 = len(model.reactions)\n",
    "    y1 = len(model.genes)\n",
    "    z1 = len(model.metabolites)\n",
    "    print(x1,y1,z1)\n",
    "\n",
    "    for x in add_these:\n",
    "        rxn = model_pf.reactions.get_by_id(x).copy()\n",
    "        if rxn.gene_reaction_rule:\n",
    "            print(rxn.gene_reaction_rule)\n",
    "        for met in rxn.metabolites:\n",
    "            if met.id not in [m.id for m in model.metabolites]:\n",
    "                model.add_metabolites([met.copy()])\n",
    "        model.add_reactions([rxn])\n",
    "\n",
    "    rxn = model_pf.reactions.get_by_id('biomass').copy()\n",
    "    for met in rxn.metabolites:\n",
    "        if met.id not in [m.id for m in model.metabolites]:\n",
    "            model.add_metabolites([met.copy()])\n",
    "    model.add_reactions([rxn])\n",
    "\n",
    "    x2 = len(model.reactions)\n",
    "    y2 = len(model.genes)\n",
    "    z2 = len(model.metabolites)\n",
    "    print(x2,y2,z2)\n",
    "\n",
    "    species_specific_mapping = mapping[mapping['[Organism]'] == species]\n",
    "    for index, row in species_specific_mapping.iterrows():\n",
    "        new_gene = row['[Gene ID]']\n",
    "        if ',' not in row['[Input Ortholog(s)]']:\n",
    "            gene = row['[Input Ortholog(s)]'].strip()\n",
    "            if gene in [x.id for x in model_pf.genes]:\n",
    "                gene = model_pf.genes.get_by_id(gene)\n",
    "                rxn_to_add = gene.reactions.copy()\n",
    "                for rxn in rxn_to_add:\n",
    "                    if rxn.id not in [x.id for x in model.reactions]:\n",
    "#                         print('no comma')\n",
    "#                         print(rxn.id)\n",
    "                        rxn2 = rxn.copy()\n",
    "                        test = len(model.reactions)\n",
    "                        model.add_reactions([rxn2])\n",
    "#                         print(len(model.reactions)-test)\n",
    "                        model.reactions.get_by_id(rxn2.id).gene_reaction_rule = new_gene\n",
    "                    else: # reaction is in model\n",
    "                        if new_gene not in model.reactions.get_by_id(rxn.id).gene_reaction_rule:\n",
    "                            if model.reactions.get_by_id(rxn.id).gene_reaction_rule == '':\n",
    "                                model.reactions.get_by_id(rxn.id).gene_reaction_rule = new_gene\n",
    "                            else: model.reactions.get_by_id(rxn.id).gene_reaction_rule = \\\n",
    "                                model.reactions.get_by_id(rxn.id).gene_reaction_rule + ' or ' +new_gene\n",
    "#                         else:\n",
    "#                             print('gene already there')\n",
    "        else:\n",
    "            for i in range(0,len(row['[Input Ortholog(s)]'].split(', ')),1):\n",
    "                gene = row['[Input Ortholog(s)]'].split(', ')[i].strip()\n",
    "                if gene in [x.id for x in model_pf.genes]:\n",
    "                    gene = model_pf.genes.get_by_id(gene)\n",
    "                    rxn_to_add = gene.reactions.copy()\n",
    "                    for rxn in rxn_to_add:\n",
    "                        if rxn.id not in [x.id for x in model.reactions]:\n",
    "#                             print('comma')\n",
    "#                             print(rxn.id)\n",
    "                            rxn2 = rxn.copy()\n",
    "                            test = len(model.reactions)\n",
    "                            model.add_reactions([rxn2])\n",
    "#                             print(len(model.reactions)-test)\n",
    "                            model.reactions.get_by_id(rxn2.id).gene_reaction_rule = new_gene\n",
    "                        else:\n",
    "                            if new_gene not in model.reactions.get_by_id(rxn.id).gene_reaction_rule:\n",
    "                                if model.reactions.get_by_id(rxn.id).gene_reaction_rule == '':\n",
    "                                    model.reactions.get_by_id(rxn.id).gene_reaction_rule = new_gene\n",
    "                                else: model.reactions.get_by_id(rxn.id).gene_reaction_rule = \\\n",
    "                                    model.reactions.get_by_id(rxn.id).gene_reaction_rule + ' or ' +new_gene\n",
    "#                             else:\n",
    "#                                 print('gene already there')\n",
    "    if len(model.reactions) != len(set(model.reactions)):\n",
    "        print('duplicate reactions')\n",
    "\n",
    "    if species.startswith('Pfalciparum3D7'):\n",
    "        s = 1\n",
    "    else:\n",
    "        gene_list = [x for x in model.genes if x.id.startswith('PF3D7_')]\n",
    "        cobra.manipulation.delete.remove_genes(model, gene_list, remove_reactions=True)\n",
    "        gene_list = [x for x in model.genes if x.id in ['mal_mito_1','mal_mito_2','mal_mito_3']]\n",
    "        cobra.manipulation.delete.remove_genes(model, gene_list, remove_reactions=True)\n",
    "        \n",
    "        \n",
    "    x3 = len(model.reactions)\n",
    "    y3 = len(model.genes)\n",
    "    z3 = len(model.metabolites)\n",
    "    print(x3,y3,z3)\n",
    "\n",
    "    modifications_ortho.species.loc[species] = species\n",
    "    modifications_ortho.starting_genes.loc[species] = y1\n",
    "    modifications_ortho.reactions_added.loc[species] = len(model.reactions) - x1\n",
    "    modifications_ortho.mets_added.loc[species] = len(model.metabolites) - z1\n",
    "    modifications_ortho.genes_added.loc[species] = len(model.genes) - y1\n",
    "\n",
    "    model.objective = 'biomass'\n",
    "    pf_model_dict[species] = model\n",
    "    cobra.io.save_json_model(model, \"DIY_ortho_\"+species+\".json\")\n",
    "modifications_ortho.to_csv('orthology_modifications_plasmodium_july23.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>starting_genes</th>\n",
       "      <th>reactions_added</th>\n",
       "      <th>mets_added</th>\n",
       "      <th>genes_added</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PfalciparumIT</th>\n",
       "      <td>PfalciparumIT</td>\n",
       "      <td>423</td>\n",
       "      <td>533</td>\n",
       "      <td>404</td>\n",
       "      <td>346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pyoeliiyoelii17XNL</th>\n",
       "      <td>Pyoeliiyoelii17XNL</td>\n",
       "      <td>395</td>\n",
       "      <td>505</td>\n",
       "      <td>401</td>\n",
       "      <td>278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PinuiSanAntonio1</th>\n",
       "      <td>PinuiSanAntonio1</td>\n",
       "      <td>400</td>\n",
       "      <td>501</td>\n",
       "      <td>395</td>\n",
       "      <td>276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pyoeliiyoelii17X</th>\n",
       "      <td>Pyoeliiyoelii17X</td>\n",
       "      <td>404</td>\n",
       "      <td>508</td>\n",
       "      <td>403</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PknowlesiMalayanPk1A</th>\n",
       "      <td>PknowlesiMalayanPk1A</td>\n",
       "      <td>403</td>\n",
       "      <td>505</td>\n",
       "      <td>403</td>\n",
       "      <td>268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PreichenowiCDC</th>\n",
       "      <td>PreichenowiCDC</td>\n",
       "      <td>418</td>\n",
       "      <td>509</td>\n",
       "      <td>405</td>\n",
       "      <td>302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PrelictumSGS1-like</th>\n",
       "      <td>PrelictumSGS1-like</td>\n",
       "      <td>410</td>\n",
       "      <td>511</td>\n",
       "      <td>397</td>\n",
       "      <td>275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PgaboniSY75</th>\n",
       "      <td>PgaboniSY75</td>\n",
       "      <td>411</td>\n",
       "      <td>507</td>\n",
       "      <td>402</td>\n",
       "      <td>303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PbillcollinsiG01</th>\n",
       "      <td>PbillcollinsiG01</td>\n",
       "      <td>129</td>\n",
       "      <td>619</td>\n",
       "      <td>537</td>\n",
       "      <td>361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PcoatneyiHackeri</th>\n",
       "      <td>PcoatneyiHackeri</td>\n",
       "      <td>402</td>\n",
       "      <td>502</td>\n",
       "      <td>400</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PvivaxSal1</th>\n",
       "      <td>PvivaxSal1</td>\n",
       "      <td>413</td>\n",
       "      <td>503</td>\n",
       "      <td>401</td>\n",
       "      <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PadleriG01</th>\n",
       "      <td>PadleriG01</td>\n",
       "      <td>415</td>\n",
       "      <td>507</td>\n",
       "      <td>404</td>\n",
       "      <td>316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PbergheiANKA</th>\n",
       "      <td>PbergheiANKA</td>\n",
       "      <td>401</td>\n",
       "      <td>512</td>\n",
       "      <td>399</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PvivaxP01</th>\n",
       "      <td>PvivaxP01</td>\n",
       "      <td>406</td>\n",
       "      <td>504</td>\n",
       "      <td>401</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pvinckeivinckeivinckei</th>\n",
       "      <td>Pvinckeivinckeivinckei</td>\n",
       "      <td>405</td>\n",
       "      <td>503</td>\n",
       "      <td>401</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PblacklockiG01</th>\n",
       "      <td>PblacklockiG01</td>\n",
       "      <td>405</td>\n",
       "      <td>505</td>\n",
       "      <td>404</td>\n",
       "      <td>305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PpraefalciparumG01</th>\n",
       "      <td>PpraefalciparumG01</td>\n",
       "      <td>414</td>\n",
       "      <td>511</td>\n",
       "      <td>404</td>\n",
       "      <td>298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PcynomolgiM</th>\n",
       "      <td>PcynomolgiM</td>\n",
       "      <td>407</td>\n",
       "      <td>498</td>\n",
       "      <td>401</td>\n",
       "      <td>274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PgaboniG01</th>\n",
       "      <td>PgaboniG01</td>\n",
       "      <td>407</td>\n",
       "      <td>507</td>\n",
       "      <td>400</td>\n",
       "      <td>303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PcynomolgiB</th>\n",
       "      <td>PcynomolgiB</td>\n",
       "      <td>382</td>\n",
       "      <td>496</td>\n",
       "      <td>401</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pchabaudichabaudi</th>\n",
       "      <td>Pchabaudichabaudi</td>\n",
       "      <td>415</td>\n",
       "      <td>506</td>\n",
       "      <td>404</td>\n",
       "      <td>271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PreichenowiG01</th>\n",
       "      <td>PreichenowiG01</td>\n",
       "      <td>410</td>\n",
       "      <td>509</td>\n",
       "      <td>405</td>\n",
       "      <td>298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PovalecurtisiGH01</th>\n",
       "      <td>PovalecurtisiGH01</td>\n",
       "      <td>407</td>\n",
       "      <td>506</td>\n",
       "      <td>403</td>\n",
       "      <td>268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pfalciparum3D7</th>\n",
       "      <td>Pfalciparum3D7</td>\n",
       "      <td>422</td>\n",
       "      <td>196</td>\n",
       "      <td>196</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PfragileNilgiri</th>\n",
       "      <td>PfragileNilgiri</td>\n",
       "      <td>405</td>\n",
       "      <td>493</td>\n",
       "      <td>395</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PknowlesiH</th>\n",
       "      <td>PknowlesiH</td>\n",
       "      <td>408</td>\n",
       "      <td>506</td>\n",
       "      <td>404</td>\n",
       "      <td>271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PmalariaeUG01</th>\n",
       "      <td>PmalariaeUG01</td>\n",
       "      <td>402</td>\n",
       "      <td>513</td>\n",
       "      <td>402</td>\n",
       "      <td>280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PvinckeipetteriCR</th>\n",
       "      <td>PvinckeipetteriCR</td>\n",
       "      <td>413</td>\n",
       "      <td>494</td>\n",
       "      <td>396</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PyoeliiyoeliiYM</th>\n",
       "      <td>PyoeliiyoeliiYM</td>\n",
       "      <td>403</td>\n",
       "      <td>508</td>\n",
       "      <td>403</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pgallinaceum8A</th>\n",
       "      <td>Pgallinaceum8A</td>\n",
       "      <td>412</td>\n",
       "      <td>513</td>\n",
       "      <td>399</td>\n",
       "      <td>278</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       species  starting_genes  \\\n",
       "PfalciparumIT                    PfalciparumIT             423   \n",
       "Pyoeliiyoelii17XNL          Pyoeliiyoelii17XNL             395   \n",
       "PinuiSanAntonio1              PinuiSanAntonio1             400   \n",
       "Pyoeliiyoelii17X              Pyoeliiyoelii17X             404   \n",
       "PknowlesiMalayanPk1A      PknowlesiMalayanPk1A             403   \n",
       "PreichenowiCDC                  PreichenowiCDC             418   \n",
       "PrelictumSGS1-like          PrelictumSGS1-like             410   \n",
       "PgaboniSY75                        PgaboniSY75             411   \n",
       "PbillcollinsiG01              PbillcollinsiG01             129   \n",
       "PcoatneyiHackeri              PcoatneyiHackeri             402   \n",
       "PvivaxSal1                          PvivaxSal1             413   \n",
       "PadleriG01                          PadleriG01             415   \n",
       "PbergheiANKA                      PbergheiANKA             401   \n",
       "PvivaxP01                            PvivaxP01             406   \n",
       "Pvinckeivinckeivinckei  Pvinckeivinckeivinckei             405   \n",
       "PblacklockiG01                  PblacklockiG01             405   \n",
       "PpraefalciparumG01          PpraefalciparumG01             414   \n",
       "PcynomolgiM                        PcynomolgiM             407   \n",
       "PgaboniG01                          PgaboniG01             407   \n",
       "PcynomolgiB                        PcynomolgiB             382   \n",
       "Pchabaudichabaudi            Pchabaudichabaudi             415   \n",
       "PreichenowiG01                  PreichenowiG01             410   \n",
       "PovalecurtisiGH01            PovalecurtisiGH01             407   \n",
       "Pfalciparum3D7                  Pfalciparum3D7             422   \n",
       "PfragileNilgiri                PfragileNilgiri             405   \n",
       "PknowlesiH                          PknowlesiH             408   \n",
       "PmalariaeUG01                    PmalariaeUG01             402   \n",
       "PvinckeipetteriCR            PvinckeipetteriCR             413   \n",
       "PyoeliiyoeliiYM                PyoeliiyoeliiYM             403   \n",
       "Pgallinaceum8A                  Pgallinaceum8A             412   \n",
       "\n",
       "                        reactions_added  mets_added  genes_added  \n",
       "PfalciparumIT                       533         404          346  \n",
       "Pyoeliiyoelii17XNL                  505         401          278  \n",
       "PinuiSanAntonio1                    501         395          276  \n",
       "Pyoeliiyoelii17X                    508         403          272  \n",
       "PknowlesiMalayanPk1A                505         403          268  \n",
       "PreichenowiCDC                      509         405          302  \n",
       "PrelictumSGS1-like                  511         397          275  \n",
       "PgaboniSY75                         507         402          303  \n",
       "PbillcollinsiG01                    619         537          361  \n",
       "PcoatneyiHackeri                    502         400          270  \n",
       "PvivaxSal1                          503         401          279  \n",
       "PadleriG01                          507         404          316  \n",
       "PbergheiANKA                        512         399          266  \n",
       "PvivaxP01                           504         401          272  \n",
       "Pvinckeivinckeivinckei              503         401          272  \n",
       "PblacklockiG01                      505         404          305  \n",
       "PpraefalciparumG01                  511         404          298  \n",
       "PcynomolgiM                         498         401          274  \n",
       "PgaboniG01                          507         400          303  \n",
       "PcynomolgiB                         496         401          270  \n",
       "Pchabaudichabaudi                   506         404          271  \n",
       "PreichenowiG01                      509         405          298  \n",
       "PovalecurtisiGH01                   506         403          268  \n",
       "Pfalciparum3D7                      196         196           75  \n",
       "PfragileNilgiri                     493         395          273  \n",
       "PknowlesiH                          506         404          271  \n",
       "PmalariaeUG01                       513         402          280  \n",
       "PvinckeipetteriCR                   494         396          272  \n",
       "PyoeliiyoeliiYM                     508         403          272  \n",
       "Pgallinaceum8A                      513         399          278  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "modifications_ortho\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FIND/ REMOVE DUPLICATE REACTIONS\n",
    "\n",
    "duplicates = dict()\n",
    "for species, model in pf_model_dict.items():\n",
    "\n",
    "    temp_dict = dict() # get products and reactants for every reaction\n",
    "    for rxn in model.reactions:\n",
    "        rxn_dict = dict()\n",
    "        check_rxn_products = rxn.products\n",
    "        check_rxn_reactants = rxn.reactants\n",
    "        rxn_dict['reactants'] = [x.id for x in check_rxn_reactants]\n",
    "        rxn_dict['products'] = [x.id for x in check_rxn_products]\n",
    "        temp_dict[rxn.id] = rxn_dict\n",
    "        \n",
    "        for key in temp_dict.keys():\n",
    "            if key != rxn.id:\n",
    "                if temp_dict[rxn.id]['reactants'] == temp_dict[key]['reactants'] and \\\n",
    "                temp_dict[rxn.id]['products'] == temp_dict[key]['products']:\n",
    "                    if rxn.id not in duplicates.keys():\n",
    "                        duplicates[rxn.id] = key\n",
    "                    elif duplicates[rxn.id] == key or key in duplicates[rxn.id]:\n",
    "                        continue\n",
    "                    else:\n",
    "                         duplicates[rxn.id] = duplicates[rxn.id]+', '+key\n",
    "                            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACCOAL2': 'ACCOAL',\n",
       " 'ACCOALm': 'ACS2',\n",
       " 'ACONTac': 'ACONTa',\n",
       " 'ACONTbc': 'ACONTb',\n",
       " 'AFAT': 'FMNAT',\n",
       " 'AKG_OAAtmt': 'OAAAKGtm',\n",
       " 'ALAS_mt': 'ALASm',\n",
       " 'ALCD19_D': 'ALCD19y',\n",
       " 'ATPH1e': 'ATPH1',\n",
       " 'ATPH2e': 'NDP1',\n",
       " 'ATPS4r': 'ATPS10_3',\n",
       " 'CS_mt': 'CSm',\n",
       " 'CYOOm': 'CYOO6m',\n",
       " 'CYOOm_mt': 'CYOO6m, CYOOm',\n",
       " 'CYTBO3': 'CYTBD',\n",
       " 'DUMPK': 'URIDK2r',\n",
       " 'FMNAT': 'AFAT',\n",
       " 'FPGS_tm': 'THFGLUS',\n",
       " 'GLCt2': 'GLCt2_2',\n",
       " 'GLYCL_mt': 'GLYCLm',\n",
       " 'GTHOr_c': 'GTHOr',\n",
       " 'GTPCI_2': 'GTPCI',\n",
       " 'HEX8': 'HEX4',\n",
       " 'ICDHyr_mt': 'ICDHym',\n",
       " 'LPLIPAL1E180pp': 'LPLPS1AGPE180',\n",
       " 'MI145P6K': 'AMITP',\n",
       " 'NADH5': 'NADHOR',\n",
       " 'NAt3_1': 'NAt3_15, NAt3_2',\n",
       " 'NAt3_2': 'NAt3_15',\n",
       " 'NO2t2rpp': 'NO2t2r',\n",
       " 'OIVD3_mt': 'OIVD3m',\n",
       " 'PGMT_2': 'PPM',\n",
       " 'PIOHex_mt': 'PIt5m',\n",
       " 'PPPGO_mt': 'PPPGOm',\n",
       " 'PPTT_mt': 'PPTTm',\n",
       " 'SPODMc': 'SPODM',\n",
       " 'SUCD2_u6m_mt': 'SUCDH_q8_m',\n",
       " 'SUCOAS1m_mt': 'SUCOAS1m',\n",
       " 'SUCOAS_mt': 'SUCOASm',\n",
       " 'XPRT': 'XPPT'}"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "key\n",
      "value\n",
      "key\n",
      "value\n",
      "key\n",
      "value\n",
      "key\n",
      "value\n",
      "key\n",
      "value\n",
      "value\n",
      "value\n",
      "key\n",
      "value\n",
      "value\n",
      "key\n",
      "value\n",
      "value\n",
      "value\n",
      "value\n",
      "value\n",
      "value\n",
      "key\n",
      "value\n",
      "value\n",
      "value\n",
      "value\n",
      "value\n",
      "value\n",
      "value\n",
      "value\n",
      "value\n",
      "value\n",
      "value\n",
      "value\n",
      "value\n",
      "key\n",
      "value\n",
      "key\n",
      "value\n",
      "value\n",
      "key\n",
      "value\n",
      "key\n",
      "value\n",
      "value\n",
      "key\n",
      "value\n",
      "key\n",
      "value\n",
      "key\n",
      "value\n",
      "key\n",
      "value\n"
     ]
    }
   ],
   "source": [
    "\n",
    "universal_model = cobra.io.load_json_model('universal_model_may2018.json')\n",
    "for key, value in duplicates.items():\n",
    "    if key in [x.id for x in universal_model.reactions]:\n",
    "        print('key')\n",
    "    if value in [x.id for x in universal_model.reactions]:\n",
    "        print('value')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# get met stoichiometry for each reaction\n",
    "# MANAGE DUPLICATE REACTIONS WITH DIFFERENT GENES ASSOCIATED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for species, model in pf_model_dict.items():\n",
    "    print(species)\n",
    "    if species ! = 'Pfalciparum3D7'\n",
    "    for x in model.genes:\n",
    "        if x.id.startswith('PF3D7'):\n",
    "            print(x.id) # SHOULD HAVE NONE"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
