{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "from cobra.core import Gene, Metabolite, Reaction\n",
    "import cobra.test\n",
    "\n",
    "universal_model = cobra.io.load_json_model('/Users/maureencarey/local_documents/work/comparative_parasite_models/data/models/universal_model_may2018.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_dict = dict()\n",
    "path = \"/Users/maureencarey/local_documents/work/comparative_parasite_models/paradigm/data/\"\n",
    "os.chdir(path)\n",
    "for filename in glob.glob(os.path.join(path, '*.json')):\n",
    "    key = filename.split('/')[len(filename.split('/'))-1]\n",
    "    key = key[:-5]\n",
    "    if key.startswith('DIY3_'):\n",
    "        key = key[5:]\n",
    "        model_dict[key] = cobra.io.load_json_model(filename)\n",
    "    elif key.startswith('DIY_ortho_'):\n",
    "        key = key[10:]\n",
    "        model_dict[key] = cobra.io.load_json_model(filename) # REPLACE MODELS in DICT IF ORTHOLOGY UPDATED\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24311\n",
      "13249\n",
      "24798\n",
      "13545\n"
     ]
    }
   ],
   "source": [
    "# universal = cobra.io.read_sbml_model(\"/Users/maureencarey/local_documents/work/comparative_parasite_models/data/curation/temp_use_CARVEME_universe_bacteria.xml\")\n",
    "universal = cobra.io.load_json_model('/Users/maureencarey/local_documents/work/comparative_parasite_models/data/models/universal_model_may2018.json')\n",
    "pf_curated = cobra.io.read_sbml_model(\"/Users/maureencarey/local_documents/work/comparative_parasite_models/data/models/2018_updated_pf_model.xml\")\n",
    "\n",
    "print(len(universal.reactions))\n",
    "print(len(universal.metabolites))\n",
    "for rxn in pf_curated.reactions:\n",
    "    if rxn.id not in [r.id for r in universal.reactions]:\n",
    "        if rxn.reaction not in [r.reaction for r in universal.reactions]:\n",
    "            rxn_to_add = rxn.copy()\n",
    "            if rxn_to_add.gene_reaction_rule is not '':\n",
    "                rxn_to_add.gene_reaction_rule = ''\n",
    "            universal.add_reactions([rxn_to_add])\n",
    "            for met in rxn_to_add.metabolites:\n",
    "                if met not in universal.metabolites:\n",
    "                    print(met.id)\n",
    "\n",
    "print(len(universal.reactions))\n",
    "print(len(universal.metabolites))\n",
    "cobra.io.save_json_model(universal, \"/Users/maureencarey/local_documents/work/comparative_parasite_models/paradigm/data/universal_with_pf_specific_rxns_july22.json\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from __future__ import absolute_import\n",
    "\n",
    "from optlang.symbolics import add, Zero\n",
    "\n",
    "from optlang.interface import OPTIMAL\n",
    "from cobra.core import Model\n",
    "from cobra.util import fix_objective_as_constraint\n",
    "\n",
    "from warnings import warn\n",
    "\n",
    "# UPDATED TO 0.13.0\n",
    "\n",
    "class GapFiller(object):\n",
    "    \"\"\"Class for performing gap filling.\n",
    "    This class implements gap filling based on a mixed-integer approach,\n",
    "    very similar to that described in [1]_ and the 'no-growth but growth'\n",
    "    part of [2]_ but with minor adjustments. In short, we add indicator\n",
    "    variables for using the reactions in the universal model, z_i and then\n",
    "    solve problem\n",
    "    minimize \\sum_i c_i * z_i\n",
    "    s.t. Sv = 0\n",
    "         v_o >= t\n",
    "         lb_i <= v_i <= ub_i\n",
    "         v_i = 0 if z_i = 0\n",
    "    where lb, ub are the upper, lower flux bounds for reaction i, c_i is a\n",
    "    cost parameter and the objective v_o is greater than the lower bound t.\n",
    "    The default costs are 1 for reactions from the universal model, 100 for\n",
    "    exchange (uptake) reactions added and 1 for added demand reactions.\n",
    "    Note that this is a mixed-integer linear program and as such will\n",
    "    expensive to solve for large models. Consider using alternatives [3]_\n",
    "    such as CORDA instead [4,5]_.\n",
    "    Parameters\n",
    "    ----------\n",
    "    model : cobra.Model\n",
    "        The model to perform gap filling on.\n",
    "    universal : cobra.Model\n",
    "        A universal model with reactions that can be used to complete the\n",
    "        model.\n",
    "    lower_bound : float\n",
    "        The minimally accepted flux for the objective in the filled model.\n",
    "    penalties : dict, None\n",
    "        A dictionary with keys being 'universal' (all reactions included in\n",
    "        the universal model), 'exchange' and 'demand' (all additionally\n",
    "        added exchange and demand reactions) for the three reaction types.\n",
    "        Can also have reaction identifiers for reaction specific costs.\n",
    "        Defaults are 1, 100 and 1 respectively.\n",
    "    integer_threshold : float\n",
    "        The threshold at which a value is considered non-zero (aka\n",
    "        integrality threshold). If gapfilled models fail to validate,\n",
    "        you may want to lower this value.\n",
    "    exchange_reactions : bool\n",
    "        Consider adding exchange (uptake) reactions for all metabolites\n",
    "        in the model.\n",
    "    demand_reactions : bool\n",
    "        Consider adding demand reactions for all metabolites.\n",
    "    References\n",
    "    ----------\n",
    "    .. [1] Reed, Jennifer L., Trina R. Patel, Keri H. Chen, Andrew R. Joyce,\n",
    "       Margaret K. Applebee, Christopher D. Herring, Olivia T. Bui, Eric M.\n",
    "       Knight, Stephen S. Fong, and Bernhard O. Palsson. “Systems Approach\n",
    "       to Refining Genome Annotation.” Proceedings of the National Academy\n",
    "       of Sciences 103, no. 46 (2006): 17480–17484.\n",
    "       [2] Kumar, Vinay Satish, and Costas D. Maranas. “GrowMatch: An\n",
    "       Automated Method for Reconciling In Silico/In Vivo Growth\n",
    "       Predictions.” Edited by Christos A. Ouzounis. PLoS Computational\n",
    "       Biology 5, no. 3 (March 13, 2009): e1000308.\n",
    "       doi:10.1371/journal.pcbi.1000308.\n",
    "       [3] http://opencobra.github.io/cobrapy/tags/gapfilling/\n",
    "       [4] Schultz, André, and Amina A. Qutub. “Reconstruction of\n",
    "       Tissue-Specific Metabolic Networks Using CORDA.” Edited by Costas D.\n",
    "       Maranas. PLOS Computational Biology 12, no. 3 (March 4, 2016):\n",
    "       e1004808. doi:10.1371/journal.pcbi.1004808.\n",
    "       [5] Diener, Christian https://github.com/cdiener/corda\n",
    "     \"\"\"\n",
    "\n",
    "    def __init__(self, model, universal=None, lower_bound=0.05,\n",
    "                 penalties=None, exchange_reactions=False,\n",
    "                 demand_reactions=True, integer_threshold=1e-6):\n",
    "        self.original_model = model\n",
    "        self.lower_bound = lower_bound\n",
    "        self.model = model.copy()\n",
    "        tolerances = self.model.solver.configuration.tolerances\n",
    "        tolerances.integrality = integer_threshold\n",
    "        self.universal = universal.copy() if universal else Model('universal')\n",
    "        self.penalties = dict(universal=1, exchange=100, demand=1)\n",
    "        if penalties is not None:\n",
    "            self.penalties.update(penalties)\n",
    "        self.integer_threshold = integer_threshold\n",
    "        self.indicators = list()\n",
    "        self.costs = dict()\n",
    "        self.extend_model(exchange_reactions, demand_reactions)\n",
    "        fix_objective_as_constraint(self.model, bound=lower_bound)\n",
    "        self.add_switches_and_objective()\n",
    "\n",
    "    def extend_model(self, exchange_reactions=False, demand_reactions=True):\n",
    "        \"\"\"Extend gapfilling model.\n",
    "        Add reactions from universal model and optionally exchange and\n",
    "        demand reactions for all metabolites in the model to perform\n",
    "        gapfilling on.\n",
    "        Parameters\n",
    "        ----------\n",
    "        exchange_reactions : bool\n",
    "            Consider adding exchange (uptake) reactions for all metabolites\n",
    "            in the model.\n",
    "        demand_reactions : bool\n",
    "            Consider adding demand reactions for all metabolites.\n",
    "        \"\"\"\n",
    "        for rxn in self.universal.reactions:\n",
    "            rxn.gapfilling_type = 'universal'\n",
    "        new_metabolites = self.universal.metabolites.query( #### NEW ####\n",
    "                lambda metabolite: metabolite not in self.model.metabolites\n",
    "        )\n",
    "        self.model.add_metabolites(new_metabolites) #### NEW ####\n",
    "        for met in self.model.metabolites:    \n",
    "            if exchange_reactions:\n",
    "                rxn = self.universal.add_boundary(\n",
    "                    met, type='exchange_smiley', lb=-1000, ub=0,\n",
    "                    reaction_id='EX_{}'.format(met.id))\n",
    "                rxn.gapfilling_type = 'exchange'\n",
    "            if demand_reactions:\n",
    "                rxn = self.universal.add_boundary(\n",
    "                    met, type='demand_smiley', lb=0, ub=1000,\n",
    "                    reaction_id='DM_{}'.format(met.id))\n",
    "                rxn.gapfilling_type = 'demand'\n",
    "\n",
    "        new_reactions = self.universal.reactions.query(\n",
    "            lambda reaction: reaction not in self.model.reactions\n",
    "        )\n",
    "        self.model.add_reactions(new_reactions)\n",
    "\n",
    "    def update_costs(self):\n",
    "        \"\"\"Update the coefficients for the indicator variables in the objective.\n",
    "        Done incrementally so that second time the function is called,\n",
    "        active indicators in the current solutions gets higher cost than the\n",
    "        unused indicators.\n",
    "        \"\"\"\n",
    "        for var in self.indicators:\n",
    "            if var not in self.costs:\n",
    "                self.costs[var] = var.cost\n",
    "            else:\n",
    "                if var._get_primal() > self.integer_threshold:\n",
    "                    self.costs[var] += var.cost\n",
    "        self.model.objective.set_linear_coefficients(self.costs)\n",
    "\n",
    "    def add_switches_and_objective(self):\n",
    "        \"\"\" Update gapfilling model with switches and the indicator objective.\n",
    "        \"\"\"\n",
    "        constraints = list()\n",
    "        big_m = max(max(abs(b) for b in r.bounds)\n",
    "                    for r in self.model.reactions)\n",
    "        prob = self.model.problem\n",
    "        for rxn in self.model.reactions:\n",
    "            if not hasattr(rxn, 'gapfilling_type'):# or rxn.id.startswith('DM_'):\n",
    "                continue\n",
    "            indicator = prob.Variable(\n",
    "                name='indicator_{}'.format(rxn.id), lb=0, ub=1, type='binary')\n",
    "            if rxn.id in self.penalties:\n",
    "                indicator.cost = self.penalties[rxn.id]\n",
    "            else:\n",
    "                indicator.cost = self.penalties[rxn.gapfilling_type]\n",
    "            indicator.rxn_id = rxn.id\n",
    "            self.indicators.append(indicator)\n",
    "\n",
    "            # if z = 1 v_i is allowed non-zero\n",
    "            # v_i - Mz <= 0   and   v_i + Mz >= 0\n",
    "            constraint_lb = prob.Constraint(\n",
    "                rxn.flux_expression - big_m * indicator, ub=0,\n",
    "                name='constraint_lb_{}'.format(rxn.id), sloppy=True)\n",
    "            constraint_ub = prob.Constraint(\n",
    "                rxn.flux_expression + big_m * indicator, lb=0,\n",
    "                name='constraint_ub_{}'.format(rxn.id), sloppy=True)\n",
    "\n",
    "            constraints.extend([constraint_lb, constraint_ub])\n",
    "\n",
    "        self.model.add_cons_vars(self.indicators)\n",
    "        self.model.add_cons_vars(constraints, sloppy=True)\n",
    "        self.model.objective = prob.Objective(\n",
    "            #Add(*self.indicators), direction='min') # UPDATED TO 0.13.0\n",
    "            Zero, direction='min', sloppy=True)\n",
    "        self.model.objective.set_linear_coefficients({ # UPDATED TO 0.13.0\n",
    "            i: 1 for i in self.indicators})\n",
    "        self.update_costs()\n",
    "\n",
    "    def fill(self, iterations=1):\n",
    "        \"\"\"Perform the gapfilling by iteratively solving the model, updating\n",
    "        the costs and recording the used reactions.\n",
    "        Parameters\n",
    "        ----------\n",
    "        iterations : int\n",
    "            The number of rounds of gapfilling to perform. For every\n",
    "            iteration, the penalty for every used reaction increases\n",
    "            linearly. This way, the algorithm is encouraged to search for\n",
    "            alternative solutions which may include previously used\n",
    "            reactions. I.e., with enough iterations pathways including 10\n",
    "            steps will eventually be reported even if the shortest pathway\n",
    "            is a single reaction.\n",
    "        Returns\n",
    "        -------\n",
    "        iterable\n",
    "            A list of lists where each element is a list reactions that were\n",
    "            used to gapfill the model.\n",
    "        Raises\n",
    "        ------\n",
    "        RuntimeError\n",
    "            If the model fails to be validated (i.e. the original model with\n",
    "            the proposed reactions added, still cannot get the required flux\n",
    "            through the objective).\n",
    "        \"\"\"\n",
    "        used_reactions = list()\n",
    "        for i in range(iterations):\n",
    "            self.model.slim_optimize(error_value=None,\n",
    "                                     message='gapfilling optimization failed')\n",
    "            solution = [self.model.reactions.get_by_id(ind.rxn_id)\n",
    "                        for ind in self.indicators if\n",
    "                        ind._get_primal() > self.integer_threshold]\n",
    "            if not self.validate(solution):\n",
    "                raise RuntimeError('failed to validate gapfilled model, '\n",
    "                                   'try lowering the integer_threshold')\n",
    "            used_reactions.append(solution)\n",
    "            self.update_costs()\n",
    "        return used_reactions\n",
    "\n",
    "    def validate(self, reactions):\n",
    "        with self.original_model as model:\n",
    "            model.add_reactions(reactions)\n",
    "            mets  = [x.metabolites for x in reactions] #### ADDED NEW ####\n",
    "            all_keys = set().union(*(d.keys() for d in mets)) #### ADDED NEW ####\n",
    "            model.add_metabolites(all_keys) #### ADDED NEW ####\n",
    "            model.slim_optimize()\n",
    "            return (model.solver.status == OPTIMAL and\n",
    "                    model.solver.objective.value >= self.lower_bound)\n",
    "\n",
    "\n",
    "def gapfill2(model, universal=None, lower_bound=0.05,\n",
    "            penalties=None, demand_reactions=True, exchange_reactions=False,\n",
    "            iterations=1):\n",
    "    \"\"\"Perform gapfilling on a model.\n",
    "    See documentation for the class GapFiller.\n",
    "    Parameters\n",
    "    ----------\n",
    "    model : cobra.Model\n",
    "        The model to perform gap filling on.\n",
    "    universal : cobra.Model, None\n",
    "        A universal model with reactions that can be used to complete the\n",
    "        model. Only gapfill considering demand and exchange reactions if\n",
    "        left missing.\n",
    "    lower_bound : float\n",
    "        The minimally accepted flux for the objective in the filled model.\n",
    "    penalties : dict, None\n",
    "        A dictionary with keys being 'universal' (all reactions included in\n",
    "        the universal model), 'exchange' and 'demand' (all additionally\n",
    "        added exchange and demand reactions) for the three reaction types.\n",
    "        Can also have reaction identifiers for reaction specific costs.\n",
    "        Defaults are 1, 100 and 1 respectively.\n",
    "    iterations : int\n",
    "        The number of rounds of gapfilling to perform. For every iteration,\n",
    "        the penalty for every used reaction increases linearly. This way,\n",
    "        the algorithm is encouraged to search for alternative solutions\n",
    "        which may include previously used reactions. I.e., with enough\n",
    "        iterations pathways including 10 steps will eventually be reported\n",
    "        even if the shortest pathway is a single reaction.\n",
    "    exchange_reactions : bool\n",
    "        Consider adding exchange (uptake) reactions for all metabolites\n",
    "        in the model.\n",
    "    demand_reactions : bool\n",
    "        Consider adding demand reactions for all metabolites.\n",
    "    Returns\n",
    "    -------\n",
    "    iterable\n",
    "        list of lists with on set of reactions that completes the model per\n",
    "        requested iteration.\n",
    "    Examples\n",
    "    --------\n",
    "    >>> import cobra.test as ct\n",
    "    >>> from cobra import Model\n",
    "    >>> from cobra.flux_analysis import gapfill\n",
    "    >>> model = ct.create_test_model(\"salmonella\")\n",
    "    >>> universal = Model('universal')\n",
    "    >>> universal.add_reactions(model.reactions.GF6PTA.copy())\n",
    "    >>> model.remove_reactions([model.reactions.GF6PTA])\n",
    "    >>> gapfill(model, universal)\n",
    "    \"\"\"\n",
    "    gapfiller = GapFiller(model, universal=universal,\n",
    "                          lower_bound=lower_bound, penalties=penalties,\n",
    "                          demand_reactions=demand_reactions,\n",
    "                          exchange_reactions=exchange_reactions)\n",
    "    return gapfiller.fill(iterations=iterations)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gapfill based on excretion of these metabolites\n",
    "columns = ['species','objective','reactions_added_it1', 'reactions_added_it2','reactions_added_it3','reactions_added_it4','reactions_added_it5']\n",
    "modifications_gapfilling = pd.DataFrame(index = pf_model_dict.keys(), columns=columns)\n",
    "\n",
    "for species, model in models_dict.items():\n",
    "    \n",
    "    if species == '':\n",
    "\n",
    "        atp = model.metabolites.get_by_id('atp_c')\n",
    "        atp_production = model.add_boundary(atp, type=\"exchange\", reaction_id='atp_production',\n",
    "                         lb=0, ub=1000.0)\n",
    "        model.objective = atp_production\n",
    "\n",
    "        print(species)\n",
    "        og_sol = model.slim_optimize()\n",
    "        if og_sol > 0.1:\n",
    "\n",
    "            #cobra.manipulation.modify.convert_to_irreversible(model)\n",
    "            model.solver = 'gurobi'\n",
    "            result = gapfill2(model,universal= universal_model, \n",
    "                              demand_reactions = False, exchange_reactions = False, iterations=5)\n",
    "        else:\n",
    "            result = list('none needed, ' + og_sol,'none needed, ' + og_sol,\n",
    "                          'none needed, ' + og_sol,'none needed, ' + og_sol,\n",
    "                          'none needed, ' + og_sol)\n",
    "\n",
    "        modifications_gapfilling.species.loc[species] = species\n",
    "        modifications_gapfilling.objective.loc[species] = 'atp_gf'\n",
    "        modifications_gapfilling.reactions_added_it1.loc[species] = result[0]\n",
    "        modifications_gapfilling.reactions_added_it2.loc[species] = result[1]\n",
    "        modifications_gapfilling.reactions_added_it3.loc[species] = result[2]\n",
    "        modifications_gapfilling.reactions_added_it4.loc[species] = result[3]\n",
    "        modifications_gapfilling.reactions_added_it5.loc[species] = result[4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gapfill based on biomass\n",
    "\n",
    "for species, model in models_dict.items():\n",
    "    \n",
    "    if species == '':\n",
    "        model.objective = 'biomass'\n",
    "\n",
    "        print(species)\n",
    "        og_sol = model.slim_optimize()\n",
    "        if og_sol > 0.1:\n",
    "\n",
    "            #cobra.manipulation.modify.convert_to_irreversible(model)\n",
    "            model.solver = 'gurobi'\n",
    "            result = gapfill2(model,universal= universal_model, \n",
    "                              demand_reactions = False, exchange_reactions = False, iterations=5)\n",
    "        else:\n",
    "            result = list('none needed, ' + og_sol,'none needed, ' + og_sol,\n",
    "                          'none needed, ' + og_sol,'none needed, ' + og_sol,\n",
    "                          'none needed, ' + og_sol)\n",
    "\n",
    "        modifications_gapfilling.species.loc[species] = species\n",
    "        modifications_gapfilling.objective.loc[species] = 'atp_gf'\n",
    "        modifications_gapfilling.reactions_added_it1.loc[species] = result[0]\n",
    "        modifications_gapfilling.reactions_added_it2.loc[species] = result[1]\n",
    "        modifications_gapfilling.reactions_added_it3.loc[species] = result[2]\n",
    "        modifications_gapfilling.reactions_added_it4.loc[species] = result[3]\n",
    "        modifications_gapfilling.reactions_added_it5.loc[species] = result[4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for species, model in models_dict.items():\n",
    "    \n",
    "        \n",
    "    if species == '':\n",
    "\n",
    "        gapfill_rxns = list()\n",
    "        if modifications_gapfilling.reactions_added_it1.loc[species].startswith('none'):\n",
    "            continue\n",
    "        else:\n",
    "            gapfill_rxns.append(modifications_gapfilling.reactions_added_it1.loc[species])\n",
    "            gapfill_rxns.append(modifications_gapfilling.reactions_added_it2.loc[species])\n",
    "            gapfill_rxns.append(modifications_gapfilling.reactions_added_it3.loc[species])\n",
    "            gapfill_rxns.append(modifications_gapfilling.reactions_added_it4.loc[species])\n",
    "            gapfill_rxns.append(modifications_gapfilling.reactions_added_it5.loc[species])\n",
    "            gapfill_rxns = [item for sublist in gapfill_rxns for item in sublist]\n",
    "\n",
    "            gapfill_dict = dict()\n",
    "            for x in list(set(gapfill_rxns)):\n",
    "                gapfill_dict[x] = gapfill_rxns.count(x)\n",
    "                model.add_reactions[universal_model.reactions.get_by_id(x)]\n",
    "                model.reactions.get_by_id(x).notes[modifications_gapfilling.objective.loc[species]] = gapfill_dict[x]+'/5'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
