{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import os\n",
    "import glob\n",
    "import pandas as pd\n",
    "\n",
    "# Matplotlib defaults\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import matplotlib.ticker as mticker\n",
    "matplotlib.rcParams[\"savefig.dpi\"] = 300\n",
    "matplotlib.rcParams['figure.dpi'] = 300\n",
    "#Rivanna fonts should include sans-serif now, but this could still raise an error\n",
    "matplotlib.rcParams['font.sans-serif'] = \"Arial\"\n",
    "matplotlib.rcParams['font.family'] = \"sans-serif\"\n",
    "\n",
    "SMALLER_SIZE = 8\n",
    "SMALL_SIZE = 14\n",
    "MEDIUM_SIZE = 16\n",
    "BIGGER_SIZE = 16\n",
    "\n",
    "matplotlib.rc('font', size=SMALL_SIZE)          # controls default text sizes\n",
    "matplotlib.rc('axes', titlesize=MEDIUM_SIZE)     # fontsize of the axes title\n",
    "matplotlib.rc('axes', labelsize=SMALL_SIZE)    # fontsize of the x and y labels\n",
    "matplotlib.rc('xtick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "matplotlib.rc('ytick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "matplotlib.rc('legend', fontsize=SMALLER_SIZE)    # legend fontsize\n",
    "matplotlib.rc('figure', titlesize=BIGGER_SIZE)  # fontsize of the figure title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_dict_json = dict()\n",
    "path = \"/home/mac9jc/paradigm/models/\"\n",
    "os.chdir(path)\n",
    "\n",
    "for filename in glob.glob(os.path.join(path, 'final_denovo_*.json')):\n",
    "    key = filename.split('/')[len(filename.split('/'))-1]\n",
    "    key = key[:-5]\n",
    "    key = key[13:]\n",
    "    model_dict_json[key] = cobra.io.load_json_model(filename)\n",
    "    \n",
    "model_dict_gf = dict()\n",
    "        \n",
    "for filename in glob.glob(os.path.join(path, 'gf_*.xml')):\n",
    "    key = filename.split('/')[len(filename.split('/'))-1]\n",
    "    key = key[:-4]\n",
    "    key = key[3:]\n",
    "    if key.startswith('P'):\n",
    "        key = key #GOOD\n",
    "    else: key = 'NA'\n",
    "    if key != 'NA':\n",
    "        model_dict_gf[key] = cobra.io.read_sbml_model(filename)\n",
    "        \n",
    "del model_dict_json['TgondiiRH'] #del model_dict_gf['TgondiiRH']\n",
    "del model_dict_json['GintestinalisAssemblageAWB'] #del model_dict_gf['GintestinalisAssemblageAWB']\n",
    "del model_dict_json['Pvivax-likePvl01']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rename_organisms(file_o_genes):\n",
    "    \n",
    "    #model_dict_gf['LmajorSD75.1'] = model_dict_gf.pop('LmajorSD75')\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('. ','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace(' strain ','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('subsp','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace(' ','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace(':','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('-','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('/','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('isolate','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('strain','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('LdonovaniCLSL','LdonovaniCL-SL', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('EhistolyticaHM1IMSSA','EhistolyticaHM1IMSS-A', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('AcastellaniistrNeff','AcastellaniiNeff', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TcruziCLBrenerNonEsmeraldolike','TcruziCLBrenerNon-Esmeraldo-like', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('NcaninumLiverpool','NcaninumLIV', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('PknowlesiMalayanStrainPk1A','PknowlesiMalayanPk1A', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TgondiiGAB22007GALDOM2','TgondiiGAB2-2007-GAL-DOM2', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('EcuniculiEcunIIIL','EcuniculiEcunIII-L', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('EhistolyticaHM1IMSSB','EhistolyticaHM1IMSS-B', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('HhammondiH.H.34','HhammondiHH34', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TcruziSylvioX1012012','TcruziSylvioX10-1-2012', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('BayalaiB08376','BayalaiB08-376', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TcruziSylvioX101','TcruziSylvioX10-1', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TcruziCLBrenerEsmeraldolike','TcruziCLBrenerEsmeraldo-like', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('GAssemblageBGS','GintestinalisAssemblageBGS', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('GAssemblageBGS_B','GintestinalisAssemblageBGS_B', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('GAssemblageEP15','GintestinalisAssemblageEP15', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TbruceibruceiTREU927','TbruceiTREU927', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('PrelictumSGS1like','PrelictumSGS1-like', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('GAssemblageA2DH','GintestinalisAssemblageADH', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TbruceiLister4272018','TbruceiLister427_2018', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('PvivaxlikePvl01','Pvivax-likePvl01', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('Pfragilenilgiri','PfragileNilgiri', regex=False)\n",
    "    #gene_file['Organism'] = gene_file['Organism'].str.replace('','TgondiiRH', regex=False)\n",
    "    #gene_file['Organism'] = gene_file['Organism'].str.replace('','GintestinalisAssemblageAWB', regex=False)\n",
    "    \n",
    "    return(file_o_genes)\n",
    "\n",
    "def id_genes_not_in_model(model_dict, file_o_genes, model_genes_denovo_input, model_genes_gf_input):\n",
    "    \n",
    "    # model_dict: dictionary of models with keys being the species ID and values are filled with the model\n",
    "    # the input for this script will typically be the de novo models\n",
    "\n",
    "    gene_in_model = dict()\n",
    "    gene_in_model_gf = dict()\n",
    "\n",
    "    for species, model in model_dict.items():\n",
    "\n",
    "        #subset gene_file for species\n",
    "        temp_gene_file = file_o_genes[file_o_genes['Organism'].str.lower() == species.lower()]\n",
    "\n",
    "        #preallocate - DE NOVO MODEL\n",
    "        species_gene_not_in_model = dict()\n",
    "        model_version = 'de novo'\n",
    "\n",
    "        #assess genes - DE NOVO MODEL\n",
    "        for row in temp_gene_file.iterrows():\n",
    "            row_use = row[1]\n",
    "            gene = row_use['Gene ID']\n",
    "            if gene not in model_genes_denovo_input[species]:\n",
    "                EC = row_use['EC numbers']\n",
    "                name = row_use['Product Description']\n",
    "                GO2 = row_use['Curated GO Functions']\n",
    "                GO1 = row_use['Computed GO Functions']\n",
    "                present = False\n",
    "            else: \n",
    "                present = True# gene in model\n",
    "            gene_in_model[gene] = {'species':species,'model_version':model_version,'EC':EC, 'name':name,'GO, curated':GO1,'GO, computed':GO2, 'present':present}\n",
    "\n",
    "        #preallocate - GAPFILLED MODEL\n",
    "        species_gene_not_in_model = dict()\n",
    "        species_count_with_EC = 0\n",
    "        model_version = 'gf'\n",
    "\n",
    "        #get other model\n",
    "        if species in model_dict_gf.keys():\n",
    "            #assess genes - GAPFILLED MODEL\n",
    "            for row in temp_gene_file.iterrows():\n",
    "                row_use = row[1]\n",
    "                gene = row_use['Gene ID']\n",
    "                if gene not in model_genes_gf_input[species]:\n",
    "                    EC = row_use['EC numbers']\n",
    "                    name = row_use['Product Description']\n",
    "                    GO2 = row_use['Curated GO Functions']\n",
    "                    GO1 = row_use['Computed GO Functions']\n",
    "                    present = False\n",
    "                else: \n",
    "                    present = True # gene in model\n",
    "                gene_in_model_gf[gene] = {'species':species,'model_version':model_version,'EC':EC, 'name':name,'GO, curated':GO1,'GO, computed':GO2, 'present':present}\n",
    "\n",
    "    output1 = pd.DataFrame.from_dict(gene_in_model, orient='index')\n",
    "    output1 = output1.rename(columns = {'index':'gene'})\n",
    "    output2 = pd.DataFrame.from_dict(gene_in_model_gf, orient='index')\n",
    "    output2 = output2.rename(columns = {'index':'gene'})\n",
    "    output = output1.append(output2, ignore_index=False)\n",
    "    \n",
    "    return(output)\n",
    "\n",
    "\n",
    "def get_gene_options(model_input):\n",
    "    \n",
    "    model_genes = list()\n",
    "    \n",
    "    for g in model_input.genes:\n",
    "        gene_id = g.id\n",
    "    \n",
    "        if gene_id[0:4] == 'rna_': gene_id = gene_id.replace('rna_', '')\n",
    "            \n",
    "        if gene_id[0:6] == 'mRNA1_': gene_id = gene_id.replace('mRNA1_', '')\n",
    "            \n",
    "        if '.?-p1' in gene_id: gene_id = gene_id.split('.?-p1')[0]\n",
    "        \n",
    "        if ':' in gene_id: gene_id = gene_id.split(':')[0]\n",
    "            \n",
    "        if \"-t\" in gene_id: gene_id = gene_id.split('-t')[0]\n",
    "            \n",
    "        if \".\" in gene_id: gene_id = gene_id.split('.')[0]\n",
    "            \n",
    "        if \"-RA-p1\" in gene_id: gene_id = gene_id.split('-RA-p1')[0]\n",
    "            \n",
    "        if \"-1-p1\" in gene_id: gene_id = gene_id.split('-1-p1')[0]\n",
    "            \n",
    "        if \"-T1-p1\" in gene_id: gene_id = gene_id.split('-T1-p1')[0]\n",
    "                 \n",
    "        if \"-p1\" in gene_id: gene_id = gene_id.split('-p1')[0]\n",
    "    \n",
    "        model_genes.append(gene_id)\n",
    "    #if gene_id == g.id: print(gene_id)\n",
    "                \n",
    "    return(model_genes)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarize_df(df_input, model_dict, gene_file_input, model_dict_gf_input, model_genes_denovo_dict_list, model_genes_gf_dict_list):\n",
    "    \n",
    "    df_denovo = df_input[df_input['model_version'] == 'de novo']\n",
    "    df_gf = df_input[df_input['model_version'] == 'gf']\n",
    "    \n",
    "    df_denovo_present = df_denovo[df_denovo['present'] == True]\n",
    "    df_denovo_absent = df_denovo[df_denovo['present'] == False]\n",
    "    \n",
    "    df_gf_present = df_gf[df_gf['present'] == True]\n",
    "    df_gf_absent = df_gf[df_gf['present'] == False]\n",
    "    \n",
    "    dict_genes = dict()\n",
    "    \n",
    "    for species in model_dict.keys():\n",
    "        \n",
    "        \n",
    "        # All EuPathDB genes\n",
    "        EuPathDB_genes = set(gene_file_input[gene_file_input['Organism'] == species]['Gene ID'])\n",
    "        # EuPathDB genes that WERE in de novo model\n",
    "        EuPathDB_genes_in_denovo = set(df_denovo_present[df_denovo_present['species'] == species].index)\n",
    "        # EuPathDB genes that WERE NOT in de novo model\n",
    "        EuPathDB_genes_NOT_in_denovo = set(df_denovo_absent[df_denovo_absent['species'] == species].index)\n",
    "\n",
    "        # All de novo genes\n",
    "        denovo_genes = set(model_genes_denovo_dict_list[species])\n",
    "        # De novo genes on EuPathDB (same list as EuPathDB_genes_in_denovo)\n",
    "        denovo_on_EuPathDB = denovo_genes.intersection(EuPathDB_genes)\n",
    "        # De novo genes NOT on EuPathDB\n",
    "        unique_to_denovo = denovo_genes.difference(EuPathDB_genes)\n",
    "\n",
    "        # check lists\n",
    "        if len(denovo_on_EuPathDB.difference(EuPathDB_genes_in_denovo)) > 0 or len(EuPathDB_genes_in_denovo.difference(denovo_on_EuPathDB)) >0:\n",
    "            print('error in de novo genes on EuPathDB: ',species)\n",
    "\n",
    "        if len(EuPathDB_genes_in_denovo) + len(EuPathDB_genes_NOT_in_denovo) == len(EuPathDB_genes): a = 0\n",
    "        else: print('error in EuPathDB split: ',species)\n",
    "\n",
    "        if len(unique_to_denovo) + len(denovo_on_EuPathDB) == len(denovo_genes): a = 0\n",
    "        else: print('error in de novo split: ',species)\n",
    "        \n",
    "        if species not in model_dict_gf_input.keys(): \n",
    "            N_EuPathDB_genes_in_gf = 'NaN'\n",
    "            N_EuPathDB_genes_NOT_in_gf = 'NaN'\n",
    "            N_gf_genes = 'NaN'\n",
    "            N_gf_on_EuPathDB = 'NaN' # same as EuPathDB_in_gf\n",
    "            N_unique_to_gf = 'NaN'\n",
    "        else:\n",
    "            # EuPathDB genes that WERE in de novo model\n",
    "            EuPathDB_genes_in_gf = set(df_gf_present[df_gf_present['species'] == species].index)\n",
    "            # EuPathDB genes that WERE NOT in de novo model\n",
    "            EuPathDB_genes_NOT_in_gf = set(df_gf_absent[df_gf_absent['species'] == species].index)\n",
    "            \n",
    "            # All gf genes\n",
    "            gf_genes = (set(model_genes_gf_dict_list[species]))\n",
    "            # gf genes on EuPathDB (same list as EuPathDB_genes_in_gf)\n",
    "            gf_on_EuPathDB = set(gf_genes.intersection(EuPathDB_genes)) #### HERE\n",
    "            # gf genes NOT on EuPathDB\n",
    "            unique_to_gf = set(gf_genes.difference(EuPathDB_genes))\n",
    "            \n",
    "            # check lists # ISSUE WITH FIRST TWO\n",
    "            if len(gf_on_EuPathDB.difference(EuPathDB_genes_in_gf)) > 0 or len(EuPathDB_genes_in_gf.difference(gf_on_EuPathDB)) >0:\n",
    "                print('gf - error in gf genes on EuPathDB: ',species) # PLASMODIUM ONLY\n",
    "\n",
    "            if len(EuPathDB_genes_in_gf) + len(EuPathDB_genes_NOT_in_gf) == len(EuPathDB_genes): a = 0\n",
    "            else: print('gf - error in EuPathDB split: ',species)\n",
    "\n",
    "            if len(unique_to_gf) + len(gf_on_EuPathDB) == len(gf_genes): a = 0\n",
    "            else: print('gf - error in de novo split: ',species)\n",
    "                \n",
    "            N_EuPathDB_genes_in_gf = len(EuPathDB_genes_in_gf)\n",
    "            N_EuPathDB_genes_NOT_in_gf = len(EuPathDB_genes_NOT_in_gf)\n",
    "            N_gf_genes = len(gf_genes)\n",
    "            N_gf_on_EuPathDB = len(gf_on_EuPathDB)\n",
    "            N_unique_to_gf = len(unique_to_gf)\n",
    "            \n",
    "\n",
    "        dict_genes[species] = {'total EuPathDB genes' : len(EuPathDB_genes),\n",
    "                               'total de novo genes' : len(denovo_genes),\n",
    "                               'total semi-curated genes': N_gf_genes,\n",
    "                               '# in de novo' : len(denovo_on_EuPathDB),\n",
    "                              '# in semi-curated' : N_gf_on_EuPathDB,\n",
    "                               '# not in de novo' : len(EuPathDB_genes_NOT_in_denovo),\n",
    "                              '# not in semi-curated' : N_EuPathDB_genes_NOT_in_gf,\n",
    "                              '# only in de novo' : len(unique_to_denovo), \n",
    "                              '# only in semi-curated' : N_unique_to_gf,\n",
    "                              \n",
    "                              'EuPathDB genes' : EuPathDB_genes,\n",
    "                               'de novo genes' : denovo_genes,\n",
    "                               'in de novo' : denovo_on_EuPathDB,\n",
    "                               'not in de novo' : EuPathDB_genes_NOT_in_denovo,\n",
    "                              'only in de novo' : unique_to_denovo} \n",
    "        \n",
    "    df_use = pd.DataFrame.from_dict(dict_genes, orient='index')\n",
    "\n",
    "    return(df_use)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_genes_denovo = dict()\n",
    "model_genes_gf = dict()\n",
    "for species, model in model_dict_json.items():\n",
    "    model_genes_denovo[species] = get_gene_options(model)\n",
    "for species, model in model_dict_gf.items():\n",
    "    model_genes_gf[species] = get_gene_options(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_file = pd.read_csv('/home/mac9jc/paradigm/data/EuPathDB_GOTerms_metabolic_process.txt', sep=\"\\t\")\n",
    "gene_file = rename_organisms(gene_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>model_version</th>\n",
       "      <th>EC</th>\n",
       "      <th>name</th>\n",
       "      <th>GO, curated</th>\n",
       "      <th>GO, computed</th>\n",
       "      <th>present</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13J3.01</th>\n",
       "      <td>TbruceiLister427</td>\n",
       "      <td>de novo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA polymerase III pseudogene, possible</td>\n",
       "      <td>DNA binding;DNA-directed 5'-3' RNA polymerase ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0H76_1008</th>\n",
       "      <td>Heriocheircanceri</td>\n",
       "      <td>de novo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>DNA binding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0H76_1010</th>\n",
       "      <td>Heriocheircanceri</td>\n",
       "      <td>de novo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CLPB</td>\n",
       "      <td>ATP binding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0H76_1016</th>\n",
       "      <td>Heriocheircanceri</td>\n",
       "      <td>de novo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRI2</td>\n",
       "      <td>DNA primase activity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0H76_102</th>\n",
       "      <td>Heriocheircanceri</td>\n",
       "      <td>de novo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RS28</td>\n",
       "      <td>structural constituent of ribosome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      species model_version   EC  \\\n",
       "13J3.01      TbruceiLister427       de novo  NaN   \n",
       "A0H76_1008  Heriocheircanceri       de novo  NaN   \n",
       "A0H76_1010  Heriocheircanceri       de novo  NaN   \n",
       "A0H76_1016  Heriocheircanceri       de novo  NaN   \n",
       "A0H76_102   Heriocheircanceri       de novo  NaN   \n",
       "\n",
       "                                               name  \\\n",
       "13J3.01     RNA polymerase III pseudogene, possible   \n",
       "A0H76_1008                     hypothetical protein   \n",
       "A0H76_1010                                     CLPB   \n",
       "A0H76_1016                                     PRI2   \n",
       "A0H76_102                                      RS28   \n",
       "\n",
       "                                                  GO, curated GO, computed  \\\n",
       "13J3.01     DNA binding;DNA-directed 5'-3' RNA polymerase ...          NaN   \n",
       "A0H76_1008                                        DNA binding          NaN   \n",
       "A0H76_1010                                        ATP binding          NaN   \n",
       "A0H76_1016                               DNA primase activity          NaN   \n",
       "A0H76_102                  structural constituent of ribosome          NaN   \n",
       "\n",
       "            present  \n",
       "13J3.01       False  \n",
       "A0H76_1008    False  \n",
       "A0H76_1010    False  \n",
       "A0H76_1016    False  \n",
       "A0H76_102     False  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = id_genes_not_in_model(model_dict_json, gene_file, model_genes_denovo, model_genes_gf)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_all = summarize_df(df, model_dict_json,  gene_file, model_dict_gf, model_genes_denovo, model_genes_gf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "0\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(sum(~(data_all['# not in de novo'] + data_all['# in de novo'] == data_all['total EuPathDB genes'])))\n",
    "print(sum(~(data_all['# only in de novo'] + data_all['# in de novo'] == data_all['total de novo genes'])))\n",
    "data_all_sub = data_all[~(data_all['# only in semi-curated'] == 'NaN')] \n",
    "print(sum(~(data_all_sub['# only in semi-curated'] + data_all_sub['# in semi-curated'] == data_all_sub['total semi-curated genes'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(189, 14)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_all_sub.shape\n",
    "data_all.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_file[gene_file['Organism'] == species].head() # source_id might not need converstion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for text in response to reviewers\n",
    "temp = df[(df['model_version'] =='de novo') & (df['species'] =='TgondiiME49')]\n",
    "print(temp.loc['TGME49_321660']['GO, curated'])\n",
    "print(temp.shape)\n",
    "#temp\n",
    "temp[pd.notna(temp['EC'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# percent of genes that have GO terms and EC and are not in model\n",
    "df_denovo = df[df['model_version'] =='de novo']\n",
    "df_gf = df[df['model_version'] =='gf']\n",
    "dict_genes_temp = dict()\n",
    "dict_genes = dict()\n",
    "for species in model_dict_json.keys():\n",
    "    species_df = gene_file[gene_file['Organism'] == species]\n",
    "    N_EuPathDB_genes_withoutEC = species_df[species_df['EC numbers'] != species_df['EC numbers']].shape[0]\n",
    "    percent_EuPathDB_genes_withoutEC = 100* N_EuPathDB_genes_withoutEC/species_df.shape[0]\n",
    "    species_denovo_df = df_denovo[df_denovo['species'] == species]\n",
    "    N_missing_from_denovo_withoutEC = species_denovo_df[species_denovo_df['EC'] != species_denovo_df['EC']].shape[0]\n",
    "    percent_missing_from_denovo_withoutEC = 100* N_missing_from_denovo_withoutEC/species_denovo_df.shape[0]\n",
    "    species_gf_df = df_gf[df_gf['species'] == species]\n",
    "    N_missing_from_gf_withoutEC = species_gf_df[species_gf_df['EC'] != species_gf_df['EC']].shape[0]\n",
    "    if species not in model_dict_gf.keys(): \n",
    "        N_missing_from_gf = 'NaN'\n",
    "        percent_missing_from_gf_withoutEC = 'NaN'\n",
    "    else:\n",
    "        percent_missing_from_gf_withoutEC = 100* N_missing_from_gf_withoutEC/species_gf_df.shape[0]\n",
    "    dict_genes[species] = {'all annotations' : percent_EuPathDB_genes_withoutEC,\n",
    "                           'omitted from\\nde novo' : percent_missing_from_denovo_withoutEC,\n",
    "                          'omitted from\\nsemi-curated' : percent_missing_from_gf_withoutEC}\n",
    "    dict_genes_temp[species] = {'all annotations' : N_EuPathDB_genes_withoutEC,\n",
    "                           'omitted from\\nde novo' : N_missing_from_denovo_withoutEC,\n",
    "                          'omitted from\\nsemi-curated' : N_missing_from_gf_withoutEC}\n",
    "    \n",
    "EC_data = pd.DataFrame.from_dict(dict_genes, orient='index')\n",
    "EC_data.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "GO_gene_IN_model = list()\n",
    "\n",
    "for species, model in model_dict_json.items():\n",
    "     \n",
    "    #subset gene_file for species\n",
    "    temp_gene_file = gene_file[gene_file['Organism'].str.lower() == species.lower()]\n",
    "    EuPathDB_list = temp_gene_file['Gene ID'].tolist()\n",
    "    \n",
    "    #assess genes - DE NOVO MODEL\n",
    "    model_version = 'de novo'\n",
    "    for g in model.genes:\n",
    "        gene = g.id\n",
    "        gene_options = [gene.split('-')[0], gene.split('.')[0], gene.split(':')[0], gene.split('A-')[0],\n",
    "                       gene.split('.1-p1')[0], gene.split('.mRNA')[0]]\n",
    "        if 'rna_' in gene[0:4]:\n",
    "            gene_options = gene_options + [gene.split('rna_')[1].split('-')[0]]\n",
    "        if any(item in EuPathDB_list for item in gene_options):\n",
    "            GO_gene_IN_model.append({'species':species,'gene':g.id,\n",
    "                                  'model_version':model_version,'present?':'present'})\n",
    "        else:\n",
    "            GO_gene_IN_model.append({'species':species,'gene':g.id,\n",
    "                                  'model_version':model_version,'present?':'absent'})\n",
    "    \n",
    "    #assess genes - GAPFILLED (SEMI-CURATED) MODEL\n",
    "    model_version = 'gf'\n",
    "    if species in model_dict_gf.keys():\n",
    "        model_gf = model_dict_gf[species]\n",
    "        \n",
    "        #assess genes - GAPFILLED MODEL\n",
    "        for g in model.genes:\n",
    "            gene = g.id\n",
    "            gene_options = [gene.split('-')[0], gene.split('.')[0], gene.split(':')[0], gene.split('A-')[0],\n",
    "                           gene.split('.1-p1')[0], gene.split('.mRNA')[0]]\n",
    "            if 'rna_' in gene[0:4]:\n",
    "                gene_options = gene_options + [gene.split('rna_')[1].split('-')[0]]\n",
    "            if any(item in EuPathDB_list for item in gene_options):\n",
    "                GO_gene_IN_model.append({'species':species,'gene':g.id,\n",
    "                                      'model_version':model_version,\n",
    "                                      'present?':'present'})\n",
    "            else:\n",
    "                GO_gene_IN_model.append({'species':species,'gene':g.id,\n",
    "                                      'model_version':model_version,\n",
    "                                      'present?':'absent'})\n",
    "            \n",
    "    \n",
    "df = pd.DataFrame(GO_gene_IN_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len([g.id for g in model_dict_json['AalgeraePRA109'].genes]))\n",
    "print(len(list(set([g.id for g in model_dict_json['AalgeraePRA109'].genes]))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[(df['model_version'] =='de novo') & (df['species'] =='AalgeraePRA109') & (df['present?'] == 'present')].shape[0] + df[(df['model_version'] =='de novo') & (df['species'] =='AalgeraePRA109') & (df['present?'] == 'absent')].shape[0] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " \n",
    "df_denovo = df[df['model_version'] == 'de novo']\n",
    "df_gf = df[df['model_version'] == 'gf']\n",
    "\n",
    "df_ab_denovo = df_denovo[df_denovo['present?'] =='absent']\n",
    "df_pres_denovo = df_denovo[df_denovo['present?'] =='present']\n",
    "\n",
    "df_ab_gf = df_gf[df_gf['present?'] =='absent']\n",
    "df_pres_gf = df_gf[df_gf['present?'] =='present']\n",
    "\n",
    "dict_genes = dict()\n",
    "for species in model_dict_json.keys():\n",
    "    \n",
    "    N_EuPathDB_genes_in_model_denovo = df_pres_denovo[df_pres_denovo['species'] == species].shape[0]\n",
    "    N_EuPathDB_genes_in_model_gf = df_pres_gf[df_pres_gf['species'] == species].shape[0]\n",
    "    N_genes_in_model_not_on_EuPathDB_denovo = df_ab_denovo[df_ab_denovo['species'] == species].shape[0]\n",
    "    N_genes_in_model_not_on_EuPathDB_gf = df_ab_gf[df_ab_gf['species'] == species].shape[0]\n",
    "    \n",
    "    if species not in model_dict_gf.keys(): \n",
    "        N_EuPathDB_genes_in_model_gf = 'NaN' \n",
    "        N_genes_in_model_not_on_EuPathDB_gf = 'NaN'\n",
    "        \n",
    "    dict_genes[species] = {'absent from\\nEuPathDB\\n(de novo)':N_genes_in_model_not_on_EuPathDB_denovo,\n",
    "                          'absent from\\nEuPathDB\\n(semi-curated)':N_genes_in_model_not_on_EuPathDB_gf,\n",
    "                          'on EuPathDB\\n(de novo)':N_EuPathDB_genes_in_model_denovo,\n",
    "                          'on EuPathDB\\n(semi-curated)':N_EuPathDB_genes_in_model_gf}\n",
    "    \n",
    "data_opp = pd.DataFrame.from_dict(dict_genes, orient='index')\n",
    "data_opp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_file = pd.read_csv('/home/mac9jc/paradigm/data/EuPathDB_GOTerms_amino_acid.txt', sep=\"\\t\")\n",
    "gene_file = rename_organisms(gene_file)\n",
    "[df, genes_in_model_AA] = id_genes_not_in_model(model_dict_json, gene_file, model_genes_denovo, model_genes_gf)\n",
    "data_amino = summarize_df(df, model_dict_json)\n",
    "data_amino.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_file = pd.read_csv('/home/mac9jc/paradigm/data/EuPathDB_GOTerm_nucleotide.txt', sep=\"\\t\")\n",
    "gene_file = rename_organisms(gene_file)\n",
    "[df, genes_in_model_nts] = id_genes_not_in_model(model_dict_json, gene_file, model_genes_denovo, model_genes_gf)\n",
    "data_nucleotides = summarize_df(df, model_dict_json)\n",
    "data_nucleotides.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_file = pd.read_csv('/home/mac9jc/paradigm/data/EuPathDB_GOTerm_transport.txt', sep=\"\\t\")\n",
    "gene_file = rename_organisms(gene_file)\n",
    "[df, genes_in_model_trans] = id_genes_not_in_model(model_dict_json, gene_file, model_genes_denovo, model_genes_gf)\n",
    "data_transporters = summarize_df(df, model_dict_json)\n",
    "data_transporters.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = ['white','grey','grey']\n",
    "\n",
    "# all annotations\n",
    "fig,ax = plt.subplots()\n",
    "x1 = data_all['EuPathDB genes'].tolist()\n",
    "x2 = data_all['# missing\\nfrom de novo'].tolist()\n",
    "x3 = [x for x in data_all['# missing\\nfrom semi-curated'].tolist() if x != 'NaN']\n",
    "bplot = ax.boxplot([x1,x2,x3],\n",
    "                   showfliers=False, patch_artist=True,\n",
    "                   labels = data_all.columns.tolist(),\n",
    "                   medianprops = dict(color=\"black\",linewidth=1.5)) \n",
    "ax.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "    patch.set_edgecolor(\"black\")\n",
    "ax.set_title('All metabolic annotations')\n",
    "ax.axes.get_xaxis().set_visible(False)\n",
    "plt.subplots_adjust(bottom = 0.35, right = 0.5)\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_all_annotations.svg\")\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_all_annotations.png\")\n",
    "\n",
    "# amino acid annotations \n",
    "fig,ax = plt.subplots()\n",
    "x1 = data_amino['EuPathDB genes'].tolist()\n",
    "x2 = data_amino['# missing\\nfrom de novo'].tolist()\n",
    "x3 = [x for x in data_amino['# missing\\nfrom semi-curated'].tolist() if x != 'NaN']\n",
    "bplot = ax.boxplot([x1,x2,x3],\n",
    "                   showfliers=False, patch_artist=True,\n",
    "                   labels = data_all.columns.tolist(),\n",
    "                   medianprops = dict(color=\"black\",linewidth=1.5)) \n",
    "ax.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "    patch.set_edgecolor(\"black\")\n",
    "ax.set_title('Amino acid annotations')\n",
    "ax.set_xticklabels(rotation=-25, ha='left', labels = ['Total EuPathDB genes','missing from de novo','missing from semi-curated'])\n",
    "plt.subplots_adjust(bottom = 0.35, right = 0.5)\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_AA_annotations.svg\")\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_AA_annotations.png\")\n",
    "\n",
    "# nucleotide annotations \n",
    "fig,ax = plt.subplots()\n",
    "x1 = data_nucleotides['EuPathDB genes'].tolist()\n",
    "x2 = data_nucleotides['# missing\\nfrom de novo'].tolist()\n",
    "x3 = [x for x in data_nucleotides['# missing\\nfrom semi-curated'].tolist() if x != 'NaN']\n",
    "bplot = ax.boxplot([x1,x2,x3],\n",
    "                   showfliers=False, patch_artist=True,\n",
    "                   labels = data_all.columns.tolist(),\n",
    "                   medianprops = dict(color=\"black\",linewidth=1.5)) \n",
    "ax.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "    patch.set_edgecolor(\"black\")\n",
    "ax.set_title('Nucleotide annotations')\n",
    "ax.axes.get_xaxis().set_visible(False)\n",
    "plt.subplots_adjust(bottom = 0.35, right = 0.5)\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_NTs_annotations.svg\")\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_NTs_annotations.png\")\n",
    "\n",
    "# transporter annotations \n",
    "fig,ax = plt.subplots()\n",
    "x1 = data_transporters['EuPathDB genes'].tolist()\n",
    "x2 = data_transporters['# missing\\nfrom de novo'].tolist()\n",
    "x3 = [x for x in data_transporters['# missing\\nfrom semi-curated'].tolist() if x != 'NaN']\n",
    "bplot = ax.boxplot([x1,x2,x3],\n",
    "                   showfliers=False, patch_artist=True,\n",
    "                   labels = data_all.columns.tolist(),\n",
    "                   medianprops = dict(color=\"black\",linewidth=1.5)) \n",
    "ax.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "    patch.set_edgecolor(\"black\")\n",
    "ax.set_title('Transporter annotations')\n",
    "ax.set_xticklabels(rotation=-25, ha='left', labels = ['Total EuPathDB genes','missing from de novo','missing from semi-curated'])\n",
    "plt.subplots_adjust(bottom = 0.35, right = 0.5)\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_transporter_annotations.svg\")\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_transporter_annotations.png\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots()\n",
    "x1 = EC_data['all annotations'].tolist()\n",
    "x2 = EC_data['omitted from\\nde novo'].tolist()\n",
    "x3 = [x for x in EC_data['omitted from\\nsemi-curated'].tolist() if x != 'NaN']\n",
    "bplot = ax.boxplot([x1,x2,x3],\n",
    "                   showfliers=False, patch_artist=True,\n",
    "                   labels = data_all.columns.tolist(),\n",
    "                   medianprops = dict(color=\"black\",linewidth=1.5)) \n",
    "ax.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "colors = ['white','grey','grey']\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "    patch.set_edgecolor(\"black\")\n",
    "ax.set_title('% genes without EC number')\n",
    "ax.set_xticklabels(rotation=-25, ha='left', labels = ['Total EuPathDB genes','missing from de novo','missing from semi-curated'])\n",
    "plt.subplots_adjust(bottom = 0.35, right = 0.5)\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_EC_eval.svg\")\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_EC_eval.png\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots()\n",
    "x1 = data_opp['absent from\\nEuPathDB\\n(de novo)'].tolist()\n",
    "x2 = [x for x in data_opp['on EuPathDB\\n(de novo)'].tolist() if x == x]\n",
    "x3 = [x for x in data_opp['absent from\\nEuPathDB\\n(semi-curated)'].tolist() if x != 'NaN']\n",
    "x4 = [x for x in data_opp['on EuPathDB\\n(semi-curated)'].tolist() if x != 'NaN']\n",
    "bplot = ax.boxplot([x1,x2,x3,x4],\n",
    "                   showfliers=False, patch_artist=True,\n",
    "                   labels = data_opp.columns.tolist(),\n",
    "                   medianprops = dict(color=\"black\",linewidth=1.5))  \n",
    "ax.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "colors = ['white','grey','white','grey']\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "    patch.set_edgecolor(\"black\")\n",
    "ax.set_title('Genes per model:')\n",
    "ax.set_xticklabels(rotation=-25, ha='left', labels = ['not on EuPathDB (de novo)','on EuPathDB (de novo)','not on EuPathDB (semi-curated)','on EuPathDB (semi-curated)'])\n",
    "plt.subplots_adjust(bottom = 0.35, right = 0.5)\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/Diamond_not_in_EuPathDB.svg\")\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/Diamond_not_in_EuPathDB.png\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-paradigm_env]",
   "language": "python",
   "name": "conda-env-.conda-paradigm_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
