{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import os\n",
    "import glob\n",
    "import pandas as pd\n",
    "\n",
    "# Matplotlib defaults\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import matplotlib.ticker as mticker\n",
    "matplotlib.rcParams[\"savefig.dpi\"] = 300\n",
    "matplotlib.rcParams['figure.dpi'] = 300\n",
    "#Rivanna fonts should include sans-serif now, but this could still raise an error\n",
    "matplotlib.rcParams['font.sans-serif'] = \"Arial\"\n",
    "matplotlib.rcParams['font.family'] = \"sans-serif\"\n",
    "\n",
    "SMALLER_SIZE = 8\n",
    "SMALL_SIZE = 14\n",
    "MEDIUM_SIZE = 16\n",
    "BIGGER_SIZE = 16\n",
    "\n",
    "matplotlib.rc('font', size=SMALL_SIZE)          # controls default text sizes\n",
    "matplotlib.rc('axes', titlesize=MEDIUM_SIZE)     # fontsize of the axes title\n",
    "matplotlib.rc('axes', labelsize=SMALL_SIZE)    # fontsize of the x and y labels\n",
    "matplotlib.rc('xtick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "matplotlib.rc('ytick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "matplotlib.rc('legend', fontsize=SMALLER_SIZE)    # legend fontsize\n",
    "matplotlib.rc('figure', titlesize=BIGGER_SIZE)  # fontsize of the figure title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_dict_json = dict()\n",
    "path = \"/home/mac9jc/paradigm/models/\"\n",
    "os.chdir(path)\n",
    "\n",
    "for filename in glob.glob(os.path.join(path, 'final_denovo_*.json')):\n",
    "    key = filename.split('/')[len(filename.split('/'))-1]\n",
    "    key = key[:-5]\n",
    "    key = key[13:]\n",
    "    model_dict_json[key] = cobra.io.load_json_model(filename)\n",
    "    \n",
    "model_dict_gf = dict()\n",
    "        \n",
    "for filename in glob.glob(os.path.join(path, 'gf_*.xml')):\n",
    "    key = filename.split('/')[len(filename.split('/'))-1]\n",
    "    key = key[:-4]\n",
    "    key = key[3:]\n",
    "    if key.startswith('P'):\n",
    "        key = key #GOOD\n",
    "    else: key = 'NA'\n",
    "    if key != 'NA':\n",
    "        model_dict_gf[key] = cobra.io.read_sbml_model(filename)\n",
    "        \n",
    "del model_dict_json['TgondiiRH'] #del model_dict_gf['TgondiiRH']\n",
    "del model_dict_json['GintestinalisAssemblageAWB'] #del model_dict_gf['GintestinalisAssemblageAWB']\n",
    "del model_dict_json['Pvivax-likePvl01']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rename_organisms(file_o_genes):\n",
    "    \n",
    "    #model_dict_gf['LmajorSD75.1'] = model_dict_gf.pop('LmajorSD75')\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('. ','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace(' strain ','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('subsp','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace(' ','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace(':','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('-','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('/','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('isolate','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('strain','', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('LdonovaniCLSL','LdonovaniCL-SL', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('EhistolyticaHM1IMSSA','EhistolyticaHM1IMSS-A', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('AcastellaniistrNeff','AcastellaniiNeff', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TcruziCLBrenerNonEsmeraldolike','TcruziCLBrenerNon-Esmeraldo-like', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('NcaninumLiverpool','NcaninumLIV', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('PknowlesiMalayanStrainPk1A','PknowlesiMalayanPk1A', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TgondiiGAB22007GALDOM2','TgondiiGAB2-2007-GAL-DOM2', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('EcuniculiEcunIIIL','EcuniculiEcunIII-L', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('EhistolyticaHM1IMSSB','EhistolyticaHM1IMSS-B', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('HhammondiH.H.34','HhammondiHH34', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TcruziSylvioX1012012','TcruziSylvioX10-1-2012', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('BayalaiB08376','BayalaiB08-376', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TcruziSylvioX101','TcruziSylvioX10-1', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TcruziCLBrenerEsmeraldolike','TcruziCLBrenerEsmeraldo-like', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('GAssemblageBGS','GintestinalisAssemblageBGS', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('GAssemblageBGS_B','GintestinalisAssemblageBGS_B', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('GAssemblageEP15','GintestinalisAssemblageEP15', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TbruceibruceiTREU927','TbruceiTREU927', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('PrelictumSGS1like','PrelictumSGS1-like', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('GAssemblageA2DH','GintestinalisAssemblageADH', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('TbruceiLister4272018','TbruceiLister427_2018', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('PvivaxlikePvl01','Pvivax-likePvl01', regex=False)\n",
    "    file_o_genes['Organism'] = file_o_genes['Organism'].str.replace('Pfragilenilgiri','PfragileNilgiri', regex=False)\n",
    "    #gene_file['Organism'] = gene_file['Organism'].str.replace('','TgondiiRH', regex=False)\n",
    "    #gene_file['Organism'] = gene_file['Organism'].str.replace('','GintestinalisAssemblageAWB', regex=False)\n",
    "    \n",
    "    return(file_o_genes)\n",
    "\n",
    "#def id_genes_not_in_model(model_dict, file_o_genes):\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_file = pd.read_csv('/home/mac9jc/paradigm/data/EuPathDB_GOTerms_metabolic_process.txt', sep=\"\\t\")\n",
    "gene_file = rename_organisms(gene_file)\n",
    "gene_file.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "gene_not_in_model = dict()\n",
    "\n",
    "for species, model in model_dict_json.items():\n",
    "     \n",
    "    #subset gene_file for species\n",
    "    temp_gene_file = gene_file[gene_file['Organism'].str.lower() == species.lower()]\n",
    "    \n",
    "    #preallocate - DE NOVO MODEL\n",
    "    species_gene_not_in_model = dict()\n",
    "    model_version = 'de novo'\n",
    "    \n",
    "    #assess genes - DE NOVO MODEL\n",
    "    for row in temp_gene_file.iterrows():\n",
    "        row_use = row[1]\n",
    "        gene = row_use['Gene ID']\n",
    "        model_genes0 = [g.id.split('-')[0] for g in model.genes]\n",
    "        model_genes1 = [g.id.split('.')[0] for g in model.genes]\n",
    "        model_genes2 = [g.id.split(':')[0] for g in model.genes]\n",
    "        if 'rna_' in [g.id[0:4] for g in model.genes]:\n",
    "            model_genes3 = [g.id.split('rna_')[1].split('-')[0] for g in model.genes if g.id.startswith('rna_')]\n",
    "        else: model_genes3 = []\n",
    "        model_genes4 = [g.id.split('A-')[0] for g in model.genes]\n",
    "        model_genes5 = [g.id.split('.1-p1')[0] for g in model.genes]\n",
    "        model_genes6 = [g.id.split('.mRNA')[0] for g in model.genes]\n",
    "        model_genes = model_genes0 + model_genes1 + model_genes2 + model_genes3 + model_genes4 + model_genes5 + model_genes6\n",
    "        if gene not in model_genes:\n",
    "            EC = row_use['EC numbers']\n",
    "            name = row_use['Product Description']\n",
    "            GO2 = row_use['Curated GO Functions']\n",
    "            GO1 = row_use['Computed GO Functions']\n",
    "            gene_not_in_model[gene] = {'species':species,'model_version':model_version,'EC':EC, 'name':name,'GO, curated':GO1,'GO, computed':GO2}\n",
    "    \n",
    "    #preallocate - GAPFILLED MODEL\n",
    "    species_gene_not_in_model = dict()\n",
    "    species_count_with_EC = 0\n",
    "    model_version = 'gf'\n",
    "    \n",
    "    #get other model\n",
    "    if species in model_dict_gf.keys():\n",
    "        model_gf = model_dict_gf[species]\n",
    "        \n",
    "        #assess genes - GAPFILLED MODEL\n",
    "        for row in temp_gene_file.iterrows():\n",
    "            row_use = row[1]\n",
    "            gene = row_use['Gene ID']\n",
    "            model_genes0 = [g.id.split('-')[0] for g in model_gf.genes]\n",
    "            model_genes1 = [g.id.split('.')[0] for g in model_gf.genes]\n",
    "            model_genes2 = [g.id.split(':')[0] for g in model_gf.genes]\n",
    "            if 'rna_' in [g.id[0:4] for g in model.genes]:\n",
    "                model_genes3 = [g.id.split('rna_')[1].split('-')[0] for g in model.genes if g.id.startswith('rna_')]\n",
    "            else: model_genes3 = []\n",
    "            model_genes4 = [g.id.split('A-')[0] for g in model_gf.genes]\n",
    "            model_genes5 = [g.id.split('.1-p1')[0] for g in model.genes]\n",
    "            model_genes6 = [g.id.split('.mRNA')[0] for g in model.genes]\n",
    "            model_genes = model_genes0 + model_genes1 + model_genes2 + model_genes3 + model_genes4 + model_genes5 + model_genes6\n",
    "            if gene not in model_genes:\n",
    "                EC = row_use['EC numbers']\n",
    "                name = row_use['Product Description']\n",
    "                GO2 = row_use['Curated GO Functions']\n",
    "                GO1 = row_use['Computed GO Functions']\n",
    "                gene_not_in_model[gene] = {'species':species,'model_version':model_version,'EC':EC, 'name':name,'GO, curated':GO1,'GO, computed':GO2}\n",
    "    \n",
    "df = pd.DataFrame.from_dict(gene_not_in_model, orient='index')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = df[(df['model_version'] =='de novo') & (df['species'] =='TgondiiME49')]\n",
    "print(temp.loc['TGME49_321660']['GO, curated'])\n",
    "print(temp.shape)\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp[pd.notna(temp['EC'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df['model_version'] == 'de novo']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#species_denovo_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_denovo = df[df['model_version'] == 'de novo']\n",
    "df_gf = df[df['model_version'] == 'gf']\n",
    "dict_genes = dict()\n",
    "for species in model_dict_json.keys():\n",
    "    if species not in model_dict_gf.keys(): N_gf = 'NaN'\n",
    "    N_denovo = len(model_dict_json[species].reactions)\n",
    "    species_df = gene_file[gene_file['Organism'] == species]\n",
    "    N_EuPathDB_genes = species_df.shape[0]\n",
    "    species_denovo_df = df_denovo[df_denovo['species'] == species]\n",
    "    N_missing_from_denovo = species_denovo_df.shape[0]\n",
    "    species_gf_df = df_gf[df_gf['species'] == species]\n",
    "    N_missing_from_gf = species_gf_df.shape[0]\n",
    "    if species not in model_dict_gf.keys(): N_missing_from_gf = 'NaN'\n",
    "    dict_genes[species] = {'EuPathDB genes' : N_EuPathDB_genes,\n",
    "                           '# missing\\nfrom de novo' : N_missing_from_denovo,\n",
    "                          '# missing\\nfrom semi-curated' : N_missing_from_gf,\n",
    "                          'de novo genes':N_denovo}\n",
    "    \n",
    "data_all = pd.DataFrame.from_dict(dict_genes, orient='index')\n",
    "data_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# percent of genes that have GO terms and EC and are not in model\n",
    "\n",
    "dict_genes_temp = dict()\n",
    "dict_genes = dict()\n",
    "for species in model_dict_json.keys():\n",
    "    species_df = gene_file[gene_file['Organism'] == species]\n",
    "    N_EuPathDB_genes_withoutEC = species_df[species_df['EC numbers'] != species_df['EC numbers']].shape[0]\n",
    "    percent_EuPathDB_genes_withoutEC = 100* N_EuPathDB_genes_withoutEC/species_df.shape[0]\n",
    "    species_denovo_df = df_denovo[df_denovo['species'] == species]\n",
    "    N_missing_from_denovo_withoutEC = species_denovo_df[species_denovo_df['EC'] != species_denovo_df['EC']].shape[0]\n",
    "    percent_missing_from_denovo_withoutEC = 100* N_missing_from_denovo_withoutEC/species_denovo_df.shape[0]\n",
    "    species_gf_df = df_gf[df_gf['species'] == species]\n",
    "    N_missing_from_gf_withoutEC = species_gf_df[species_gf_df['EC'] != species_gf_df['EC']].shape[0]\n",
    "    if species not in model_dict_gf.keys(): \n",
    "        N_missing_from_gf = 'NaN'\n",
    "        percent_missing_from_gf_withoutEC = 'NaN'\n",
    "    else:\n",
    "        percent_missing_from_gf_withoutEC = 100* N_missing_from_gf_withoutEC/species_gf_df.shape[0]\n",
    "    dict_genes[species] = {'all annotations' : percent_EuPathDB_genes_withoutEC,\n",
    "                           'omitted from\\nde novo' : percent_missing_from_denovo_withoutEC,\n",
    "                          'omitted from\\nsemi-curated' : percent_missing_from_gf_withoutEC}\n",
    "    dict_genes_temp[species] = {'all annotations' : N_EuPathDB_genes_withoutEC,\n",
    "                           'omitted from\\nde novo' : N_missing_from_denovo_withoutEC,\n",
    "                          'omitted from\\nsemi-curated' : N_missing_from_gf_withoutEC}\n",
    "    \n",
    "EC_data = pd.DataFrame.from_dict(dict_genes, orient='index')\n",
    "EC_data.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "gene_IN_model = list()\n",
    "\n",
    "for species, model in model_dict_json.items():\n",
    "     \n",
    "    #subset gene_file for species\n",
    "    temp_gene_file = gene_file[gene_file['Organism'].str.lower() == species.lower()]\n",
    "    EuPathDB_list = temp_gene_file['Gene ID'].tolist()\n",
    "    \n",
    "    #assess genes - DE NOVO MODEL\n",
    "    model_version = 'de novo'\n",
    "    for g in model.genes:\n",
    "        gene = g.id\n",
    "        gene_options = [gene.split('-')[0], gene.split('.')[0], gene.split(':')[0], gene.split('A-')[0],\n",
    "                       gene.split('.1-p1')[0], gene.split('.mRNA')[0]]\n",
    "        if 'rna_' in gene[0:4]:\n",
    "            gene_options = gene_options + [gene.split('rna_')[1].split('-')[0]]\n",
    "        if any(item in EuPathDB_list for item in gene_options):\n",
    "            gene_IN_model.append({'species':species,'gene':g.id,\n",
    "                                  'model_version':model_version,'present?':'present'})\n",
    "        else:\n",
    "            gene_IN_model.append({'species':species,'gene':g.id,\n",
    "                                  'model_version':model_version,'present?':'absent'})\n",
    "    \n",
    "    #assess genes - GAPFILLED (SEMI-CURATED) MODEL\n",
    "    model_version = 'gf'\n",
    "    if species in model_dict_gf.keys():\n",
    "        model_gf = model_dict_gf[species]\n",
    "        \n",
    "        #assess genes - GAPFILLED MODEL\n",
    "        for g in model.genes:\n",
    "            gene = g.id\n",
    "            gene_options = [gene.split('-')[0], gene.split('.')[0], gene.split(':')[0], gene.split('A-')[0],\n",
    "                           gene.split('.1-p1')[0], gene.split('.mRNA')[0]]\n",
    "            if 'rna_' in gene[0:4]:\n",
    "                gene_options = gene_options + [gene.split('rna_')[1].split('-')[0]]\n",
    "            if any(item in EuPathDB_list for item in gene_options):\n",
    "                gene_IN_model.append({'species':species,'gene':g.id,\n",
    "                                      'model_version':model_version,\n",
    "                                      'present?':'present'})\n",
    "            else:\n",
    "                gene_IN_model.append({'species':species,'gene':g.id,\n",
    "                                      'model_version':model_version,\n",
    "                                      'present?':'absent'})\n",
    "            \n",
    "    \n",
    "df = pd.DataFrame(gene_IN_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df['model_version'] == 'de novo'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " \n",
    "df_denovo = df[df['model_version'] == 'de novo']\n",
    "df_gf = df[df['model_version'] == 'gf']\n",
    "\n",
    "df_ab_denovo = df_denovo[df_denovo['present?'] =='absent']\n",
    "df_pres_denovo = df_denovo[df_denovo['present?'] =='present']\n",
    "\n",
    "df_ab_gf = df_gf[df_gf['present?'] =='absent']\n",
    "df_pres_gf = df_gf[df_gf['present?'] =='present']\n",
    "\n",
    "dict_genes = dict()\n",
    "for species in model_dict_json.keys():\n",
    "    \n",
    "    N_EuPathDB_genes_in_model_denovo = df_pres_denovo[df_pres_denovo['species'] == species].shape[0]\n",
    "    N_EuPathDB_genes_in_model_gf = df_pres_gf[df_pres_gf['species'] == species].shape[0]\n",
    "    N_genes_in_model_not_on_EuPathDB_denovo = df_ab_denovo[df_ab_denovo['species'] == species].shape[0]\n",
    "    N_genes_in_model_not_on_EuPathDB_gf = df_ab_gf[df_ab_gf['species'] == species].shape[0]\n",
    "    \n",
    "    if species not in model_dict_gf.keys(): \n",
    "        N_EuPathDB_genes_in_model_gf = 'NaN' \n",
    "        N_genes_in_model_not_on_EuPathDB_gf = 'NaN'\n",
    "        \n",
    "    dict_genes[species] = {'absent from\\nEuPathDB\\n(de novo)':N_genes_in_model_not_on_EuPathDB_denovo,\n",
    "                          'absent from\\nEuPathDB\\n(semi-curated)':N_genes_in_model_not_on_EuPathDB_gf,\n",
    "                          'on EuPathDB\\n(de novo)':N_EuPathDB_genes_in_model_denovo,\n",
    "                          'on EuPathDB\\n(semi-curated)':N_EuPathDB_genes_in_model_gf}\n",
    "    \n",
    "data_opp = pd.DataFrame.from_dict(dict_genes, orient='index')\n",
    "data_opp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_file = pd.read_csv('/home/mac9jc/paradigm/data/EuPathDB_GOTerms_amino_acid.txt', sep=\"\\t\")\n",
    "gene_file = rename_organisms(gene_file)\n",
    "\n",
    "gene_not_in_model = dict()\n",
    "\n",
    "for species, model in model_dict_json.items():\n",
    "     \n",
    "    #subset gene_file for species\n",
    "    temp_gene_file = gene_file[gene_file['Organism'].str.lower() == species.lower()]\n",
    "    \n",
    "    #preallocate - DE NOVO MODEL\n",
    "    species_gene_not_in_model = dict()\n",
    "    model_version = 'de novo'\n",
    "    \n",
    "    #assess genes - DE NOVO MODEL\n",
    "    for row in temp_gene_file.iterrows():\n",
    "        row_use = row[1]\n",
    "        gene = row_use['Gene ID']\n",
    "        model_genes0 = [g.id.split('-')[0] for g in model.genes]\n",
    "        model_genes1 = [g.id.split('.')[0] for g in model.genes]\n",
    "        model_genes2 = [g.id.split(':')[0] for g in model.genes]\n",
    "        if 'rna_' in [g.id[0:4] for g in model.genes]:\n",
    "            model_genes3 = [g.id.split('rna_')[1].split('-')[0] for g in model.genes if g.id.startswith('rna_')]\n",
    "        else: model_genes3 = []\n",
    "        model_genes4 = [g.id.split('A-')[0] for g in model.genes]\n",
    "        model_genes5 = [g.id.split('.1-p1')[0] for g in model.genes]\n",
    "        model_genes6 = [g.id.split('.mRNA')[0] for g in model.genes]\n",
    "        model_genes = model_genes0 + model_genes1 + model_genes2 + model_genes3 + model_genes4 + model_genes5 + model_genes6\n",
    "        if gene not in model_genes:\n",
    "            EC = row_use['EC numbers']\n",
    "            name = row_use['Product Description']\n",
    "            GO2 = row_use['Curated GO Functions']\n",
    "            GO1 = row_use['Computed GO Functions']\n",
    "            gene_not_in_model[gene] = {'species':species,'model_version':model_version,'EC':EC, 'name':name,'GO, curated':GO1,'GO, computed':GO2}\n",
    "    \n",
    "    #preallocate - GAPFILLED MODEL\n",
    "    species_gene_not_in_model = dict()\n",
    "    species_count_with_EC = 0\n",
    "    model_version = 'gf'\n",
    "    \n",
    "    #get other model\n",
    "    if species in model_dict_gf.keys():\n",
    "        model_gf = model_dict_gf[species]\n",
    "        \n",
    "        #assess genes - GAPFILLED MODEL\n",
    "        for row in temp_gene_file.iterrows():\n",
    "            row_use = row[1]\n",
    "            gene = row_use['Gene ID']\n",
    "            model_genes0 = [g.id.split('-')[0] for g in model_gf.genes]\n",
    "            model_genes1 = [g.id.split('.')[0] for g in model_gf.genes]\n",
    "            model_genes2 = [g.id.split(':')[0] for g in model_gf.genes]\n",
    "            if 'rna_' in [g.id[0:4] for g in model.genes]:\n",
    "                model_genes3 = [g.id.split('rna_')[1].split('-')[0] for g in model.genes if g.id.startswith('rna_')]\n",
    "            else: model_genes3 = []\n",
    "            model_genes4 = [g.id.split('A-')[0] for g in model_gf.genes]\n",
    "            model_genes5 = [g.id.split('.1-p1')[0] for g in model.genes]\n",
    "            model_genes6 = [g.id.split('.mRNA')[0] for g in model.genes]\n",
    "            model_genes = model_genes0 + model_genes1 + model_genes2 + model_genes3 + model_genes4 + model_genes5 + model_genes6\n",
    "            if gene not in model_genes:\n",
    "                EC = row_use['EC numbers']\n",
    "                name = row_use['Product Description']\n",
    "                GO2 = row_use['Curated GO Functions']\n",
    "                GO1 = row_use['Computed GO Functions']\n",
    "                gene_not_in_model[gene] = {'species':species,'model_version':model_version,'EC':EC, 'name':name,'GO, curated':GO1,'GO, computed':GO2}\n",
    "    \n",
    "df = pd.DataFrame.from_dict(gene_not_in_model, orient='index')\n",
    "\n",
    "df_denovo = df[df['model_version'] == 'de novo']\n",
    "df_gf = df[df['model_version'] == 'gf']\n",
    "dict_genes = dict()\n",
    "for species in model_dict_json.keys():\n",
    "    species_df = gene_file[gene_file['Organism'] == species]\n",
    "    N_EuPathDB_genes = species_df.shape[0]\n",
    "    species_denovo_df = df_denovo[df_denovo['species'] == species]\n",
    "    N_missing_from_denovo = species_denovo_df.shape[0]\n",
    "    species_gf_df = df_gf[df_gf['species'] == species]\n",
    "    N_missing_from_gf = species_gf_df.shape[0]\n",
    "    if species not in model_dict_gf.keys(): N_missing_from_gf = 'NaN'\n",
    "    dict_genes[species] = {'EuPathDB genes' : N_EuPathDB_genes,\n",
    "                           '# missing\\nfrom de novo' : N_missing_from_denovo,\n",
    "                          '# missing\\nfrom semi-curated' : N_missing_from_gf}\n",
    "    \n",
    "data_amino = pd.DataFrame.from_dict(dict_genes, orient='index')\n",
    "data_amino.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_file = pd.read_csv('/home/mac9jc/paradigm/data/EuPathDB_GOTerm_nucleotide.txt', sep=\"\\t\")\n",
    "gene_file = rename_organisms(gene_file)\n",
    "\n",
    "gene_not_in_model = dict()\n",
    "\n",
    "for species, model in model_dict_json.items():\n",
    "     \n",
    "    #subset gene_file for species\n",
    "    temp_gene_file = gene_file[gene_file['Organism'].str.lower() == species.lower()]\n",
    "    \n",
    "    #preallocate - DE NOVO MODEL\n",
    "    species_gene_not_in_model = dict()\n",
    "    model_version = 'de novo'\n",
    "    \n",
    "    #assess genes - DE NOVO MODEL\n",
    "    for row in temp_gene_file.iterrows():\n",
    "        row_use = row[1]\n",
    "        gene = row_use['Gene ID']\n",
    "        model_genes0 = [g.id.split('-')[0] for g in model.genes]\n",
    "        model_genes1 = [g.id.split('.')[0] for g in model.genes]\n",
    "        model_genes2 = [g.id.split(':')[0] for g in model.genes]\n",
    "        if 'rna_' in [g.id[0:4] for g in model.genes]:\n",
    "            model_genes3 = [g.id.split('rna_')[1].split('-')[0] for g in model.genes if g.id.startswith('rna_')]\n",
    "        else: model_genes3 = []\n",
    "        model_genes4 = [g.id.split('A-')[0] for g in model.genes]\n",
    "        model_genes5 = [g.id.split('.1-p1')[0] for g in model.genes]\n",
    "        model_genes6 = [g.id.split('.mRNA')[0] for g in model.genes]\n",
    "        model_genes = model_genes0 + model_genes1 + model_genes2 + model_genes3 + model_genes4 + model_genes5 + model_genes6\n",
    "        if gene not in model_genes:\n",
    "            EC = row_use['EC numbers']\n",
    "            name = row_use['Product Description']\n",
    "            GO2 = row_use['Curated GO Functions']\n",
    "            GO1 = row_use['Computed GO Functions']\n",
    "            gene_not_in_model[gene] = {'species':species,'model_version':model_version,'EC':EC, 'name':name,'GO, curated':GO1,'GO, computed':GO2}\n",
    "    \n",
    "    #preallocate - GAPFILLED MODEL\n",
    "    species_gene_not_in_model = dict()\n",
    "    species_count_with_EC = 0\n",
    "    model_version = 'gf'\n",
    "    \n",
    "    #get other model\n",
    "    if species in model_dict_gf.keys():\n",
    "        model_gf = model_dict_gf[species]\n",
    "        \n",
    "        #assess genes - GAPFILLED MODEL\n",
    "        for row in temp_gene_file.iterrows():\n",
    "            row_use = row[1]\n",
    "            gene = row_use['Gene ID']\n",
    "            model_genes0 = [g.id.split('-')[0] for g in model_gf.genes]\n",
    "            model_genes1 = [g.id.split('.')[0] for g in model_gf.genes]\n",
    "            model_genes2 = [g.id.split(':')[0] for g in model_gf.genes]\n",
    "            if 'rna_' in [g.id[0:4] for g in model.genes]:\n",
    "                model_genes3 = [g.id.split('rna_')[1].split('-')[0] for g in model.genes if g.id.startswith('rna_')]\n",
    "            else: model_genes3 = []\n",
    "            model_genes4 = [g.id.split('A-')[0] for g in model_gf.genes]\n",
    "            model_genes5 = [g.id.split('.1-p1')[0] for g in model.genes]\n",
    "            model_genes6 = [g.id.split('.mRNA')[0] for g in model.genes]\n",
    "            model_genes = model_genes0 + model_genes1 + model_genes2 + model_genes3 + model_genes4 + model_genes5 + model_genes6\n",
    "            if gene not in model_genes:\n",
    "                EC = row_use['EC numbers']\n",
    "                name = row_use['Product Description']\n",
    "                GO2 = row_use['Curated GO Functions']\n",
    "                GO1 = row_use['Computed GO Functions']\n",
    "                gene_not_in_model[gene] = {'species':species,'model_version':model_version,'EC':EC, 'name':name,'GO, curated':GO1,'GO, computed':GO2}\n",
    "    \n",
    "df = pd.DataFrame.from_dict(gene_not_in_model, orient='index')\n",
    "\n",
    "df_denovo = df[df['model_version'] == 'de novo']\n",
    "df_gf = df[df['model_version'] == 'gf']\n",
    "dict_genes = dict()\n",
    "for species in model_dict_json.keys():\n",
    "    species_df = gene_file[gene_file['Organism'] == species]\n",
    "    N_EuPathDB_genes = species_df.shape[0]\n",
    "    species_denovo_df = df_denovo[df_denovo['species'] == species]\n",
    "    N_missing_from_denovo = species_denovo_df.shape[0]\n",
    "    species_gf_df = df_gf[df_gf['species'] == species]\n",
    "    N_missing_from_gf = species_gf_df.shape[0]\n",
    "    if species not in model_dict_gf.keys(): N_missing_from_gf = 'NaN'\n",
    "    dict_genes[species] = {'EuPathDB genes' : N_EuPathDB_genes,\n",
    "                           '# missing\\nfrom de novo' : N_missing_from_denovo,\n",
    "                          '# missing\\nfrom semi-curated' : N_missing_from_gf}\n",
    "    \n",
    "data_nucleotides = pd.DataFrame.from_dict(dict_genes, orient='index')\n",
    "data_nucleotides.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_file = pd.read_csv('/home/mac9jc/paradigm/data/EuPathDB_GOTerm_transport.txt', sep=\"\\t\")\n",
    "gene_file = rename_organisms(gene_file)\n",
    "\n",
    "gene_not_in_model = dict()\n",
    "\n",
    "for species, model in model_dict_json.items():\n",
    "     \n",
    "    #subset gene_file for species\n",
    "    temp_gene_file = gene_file[gene_file['Organism'].str.lower() == species.lower()]\n",
    "    \n",
    "    #preallocate - DE NOVO MODEL\n",
    "    species_gene_not_in_model = dict()\n",
    "    model_version = 'de novo'\n",
    "    \n",
    "    #assess genes - DE NOVO MODEL\n",
    "    for row in temp_gene_file.iterrows():\n",
    "        row_use = row[1]\n",
    "        gene = row_use['Gene ID']\n",
    "        model_genes0 = [g.id.split('-')[0] for g in model.genes]\n",
    "        model_genes1 = [g.id.split('.')[0] for g in model.genes]\n",
    "        model_genes2 = [g.id.split(':')[0] for g in model.genes]\n",
    "        if 'rna_' in [g.id[0:4] for g in model.genes]:\n",
    "            model_genes3 = [g.id.split('rna_')[1].split('-')[0] for g in model.genes if g.id.startswith('rna_')]\n",
    "        else: model_genes3 = []\n",
    "        model_genes4 = [g.id.split('A-')[0] for g in model.genes]\n",
    "        model_genes5 = [g.id.split('.1-p1')[0] for g in model.genes]\n",
    "        model_genes6 = [g.id.split('.mRNA')[0] for g in model.genes]\n",
    "        model_genes = model_genes0 + model_genes1 + model_genes2 + model_genes3 + model_genes4 + model_genes5 + model_genes6\n",
    "        if gene not in model_genes:\n",
    "            EC = row_use['EC numbers']\n",
    "            name = row_use['Product Description']\n",
    "            GO2 = row_use['Curated GO Functions']\n",
    "            GO1 = row_use['Computed GO Functions']\n",
    "            gene_not_in_model[gene] = {'species':species,'model_version':model_version,'EC':EC, 'name':name,'GO, curated':GO1,'GO, computed':GO2}\n",
    "    \n",
    "    #preallocate - GAPFILLED MODEL\n",
    "    species_gene_not_in_model = dict()\n",
    "    species_count_with_EC = 0\n",
    "    model_version = 'gf'\n",
    "    \n",
    "    #get other model\n",
    "    if species in model_dict_gf.keys():\n",
    "        model_gf = model_dict_gf[species]\n",
    "        \n",
    "        #assess genes - GAPFILLED MODEL\n",
    "        for row in temp_gene_file.iterrows():\n",
    "            row_use = row[1]\n",
    "            gene = row_use['Gene ID']\n",
    "            model_genes0 = [g.id.split('-')[0] for g in model_gf.genes]\n",
    "            model_genes1 = [g.id.split('.')[0] for g in model_gf.genes]\n",
    "            model_genes2 = [g.id.split(':')[0] for g in model_gf.genes]\n",
    "            if 'rna_' in [g.id[0:4] for g in model.genes]:\n",
    "                model_genes3 = [g.id.split('rna_')[1].split('-')[0] for g in model.genes if g.id.startswith('rna_')]\n",
    "            else: model_genes3 = []\n",
    "            model_genes4 = [g.id.split('A-')[0] for g in model_gf.genes]\n",
    "            model_genes5 = [g.id.split('.1-p1')[0] for g in model.genes]\n",
    "            model_genes6 = [g.id.split('.mRNA')[0] for g in model.genes]\n",
    "            model_genes = model_genes0 + model_genes1 + model_genes2 + model_genes3 + model_genes4 + model_genes5 + model_genes6\n",
    "            if gene not in model_genes:\n",
    "                EC = row_use['EC numbers']\n",
    "                name = row_use['Product Description']\n",
    "                GO2 = row_use['Curated GO Functions']\n",
    "                GO1 = row_use['Computed GO Functions']\n",
    "                gene_not_in_model[gene] = {'species':species,'model_version':model_version,'EC':EC, 'name':name,'GO, curated':GO1,'GO, computed':GO2}\n",
    "    \n",
    "df = pd.DataFrame.from_dict(gene_not_in_model, orient='index')\n",
    "\n",
    "df_denovo = df[df['model_version'] == 'de novo']\n",
    "df_gf = df[df['model_version'] == 'gf']\n",
    "dict_genes = dict()\n",
    "for species in model_dict_json.keys():\n",
    "    species_df = gene_file[gene_file['Organism'] == species]\n",
    "    N_EuPathDB_genes = species_df.shape[0]\n",
    "    species_denovo_df = df_denovo[df_denovo['species'] == species]\n",
    "    N_missing_from_denovo = species_denovo_df.shape[0]\n",
    "    species_gf_df = df_gf[df_gf['species'] == species]\n",
    "    N_missing_from_gf = species_gf_df.shape[0]\n",
    "    if species not in model_dict_gf.keys(): N_missing_from_gf = 'NaN'\n",
    "    dict_genes[species] = {'EuPathDB genes' : N_EuPathDB_genes,\n",
    "                           '# missing\\nfrom de novo' : N_missing_from_denovo,\n",
    "                          '# missing\\nfrom semi-curated' : N_missing_from_gf}\n",
    "    \n",
    "data_transporters = pd.DataFrame.from_dict(dict_genes, orient='index')\n",
    "data_transporters.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = ['white','silver','dimgrey']\n",
    "\n",
    "# all annotations\n",
    "fig,ax = plt.subplots()\n",
    "x1 = data_all['EuPathDB genes'].tolist()\n",
    "x2 = data_all['# missing\\nfrom de novo'].tolist()\n",
    "x3 = [x for x in data_all['# missing\\nfrom semi-curated'].tolist() if x != 'NaN']\n",
    "bplot = ax.boxplot([x1,x2,x3],\n",
    "                   showfliers=False, patch_artist=True,\n",
    "                   labels = data_all.columns.tolist(),\n",
    "                   medianprops = dict(color=\"black\",linewidth=1.5)) \n",
    "ax.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "    patch.set_edgecolor(\"black\")\n",
    "ax.set_title('All metabolic annotations')\n",
    "ax.axes.get_xaxis().set_visible(False)\n",
    "plt.subplots_adjust(bottom = 0.35, right = 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = ['white','grey','grey']\n",
    "\n",
    "# all annotations\n",
    "fig,ax = plt.subplots()\n",
    "x1 = data_all['EuPathDB genes'].tolist()\n",
    "x2 = data_all['# missing\\nfrom de novo'].tolist()\n",
    "x3 = [x for x in data_all['# missing\\nfrom semi-curated'].tolist() if x != 'NaN']\n",
    "bplot = ax.boxplot([x1,x2,x3],\n",
    "                   showfliers=False, patch_artist=True,\n",
    "                   labels = data_all.columns.tolist(),\n",
    "                   medianprops = dict(color=\"black\",linewidth=1.5)) \n",
    "ax.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "    patch.set_edgecolor(\"black\")\n",
    "ax.set_title('All metabolic annotations')\n",
    "ax.axes.get_xaxis().set_visible(False)\n",
    "plt.subplots_adjust(bottom = 0.35, right = 0.5)\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_all_annotations.svg\")\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_all_annotations.png\")\n",
    "\n",
    "# amino acid annotations \n",
    "fig,ax = plt.subplots()\n",
    "x1 = data_amino['EuPathDB genes'].tolist()\n",
    "x2 = data_amino['# missing\\nfrom de novo'].tolist()\n",
    "x3 = [x for x in data_amino['# missing\\nfrom semi-curated'].tolist() if x != 'NaN']\n",
    "bplot = ax.boxplot([x1,x2,x3],\n",
    "                   showfliers=False, patch_artist=True,\n",
    "                   labels = data_all.columns.tolist(),\n",
    "                   medianprops = dict(color=\"black\",linewidth=1.5)) \n",
    "ax.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "    patch.set_edgecolor(\"black\")\n",
    "ax.set_title('Amino acid annotations')\n",
    "ax.set_xticklabels(rotation=-25, ha='left', labels = ['Total EuPathDB genes','missing from de novo','missing from semi-curated'])\n",
    "plt.subplots_adjust(bottom = 0.35, right = 0.5)\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_AA_annotations.svg\")\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_AA_annotations.png\")\n",
    "\n",
    "# nucleotide annotations \n",
    "fig,ax = plt.subplots()\n",
    "x1 = data_nucleotides['EuPathDB genes'].tolist()\n",
    "x2 = data_nucleotides['# missing\\nfrom de novo'].tolist()\n",
    "x3 = [x for x in data_nucleotides['# missing\\nfrom semi-curated'].tolist() if x != 'NaN']\n",
    "bplot = ax.boxplot([x1,x2,x3],\n",
    "                   showfliers=False, patch_artist=True,\n",
    "                   labels = data_all.columns.tolist(),\n",
    "                   medianprops = dict(color=\"black\",linewidth=1.5)) \n",
    "ax.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "    patch.set_edgecolor(\"black\")\n",
    "ax.set_title('Nucleotide annotations')\n",
    "ax.axes.get_xaxis().set_visible(False)\n",
    "plt.subplots_adjust(bottom = 0.35, right = 0.5)\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_NTs_annotations.svg\")\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_NTs_annotations.png\")\n",
    "\n",
    "# transporter annotations \n",
    "fig,ax = plt.subplots()\n",
    "x1 = data_transporters['EuPathDB genes'].tolist()\n",
    "x2 = data_transporters['# missing\\nfrom de novo'].tolist()\n",
    "x3 = [x for x in data_transporters['# missing\\nfrom semi-curated'].tolist() if x != 'NaN']\n",
    "bplot = ax.boxplot([x1,x2,x3],\n",
    "                   showfliers=False, patch_artist=True,\n",
    "                   labels = data_all.columns.tolist(),\n",
    "                   medianprops = dict(color=\"black\",linewidth=1.5)) \n",
    "ax.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "    patch.set_edgecolor(\"black\")\n",
    "ax.set_title('Transporter annotations')\n",
    "ax.set_xticklabels(rotation=-25, ha='left', labels = ['Total EuPathDB genes','missing from de novo','missing from semi-curated'])\n",
    "plt.subplots_adjust(bottom = 0.35, right = 0.5)\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_transporter_annotations.svg\")\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_comparison_transporter_annotations.png\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots()\n",
    "x1 = EC_data['all annotations'].tolist()\n",
    "x2 = EC_data['omitted from\\nde novo'].tolist()\n",
    "x3 = [x for x in EC_data['omitted from\\nsemi-curated'].tolist() if x != 'NaN']\n",
    "bplot = ax.boxplot([x1,x2,x3],\n",
    "                   showfliers=False, patch_artist=True,\n",
    "                   labels = data_all.columns.tolist(),\n",
    "                   medianprops = dict(color=\"black\",linewidth=1.5)) \n",
    "ax.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "colors = ['white','grey','grey']\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "    patch.set_edgecolor(\"black\")\n",
    "ax.set_title('% genes without EC number')\n",
    "ax.set_xticklabels(rotation=-25, ha='left', labels = ['Total EuPathDB genes','missing from de novo','missing from semi-curated'])\n",
    "plt.subplots_adjust(bottom = 0.35, right = 0.5)\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_EC_eval.svg\")\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/EuPathDB_EC_eval.png\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots()\n",
    "x1 = data_opp['absent from\\nEuPathDB\\n(de novo)'].tolist()\n",
    "x2 = [x for x in data_opp['on EuPathDB\\n(de novo)'].tolist() if x == x]\n",
    "x3 = [x for x in data_opp['absent from\\nEuPathDB\\n(semi-curated)'].tolist() if x != 'NaN']\n",
    "x4 = [x for x in data_opp['on EuPathDB\\n(semi-curated)'].tolist() if x != 'NaN']\n",
    "bplot = ax.boxplot([x1,x2,x3,x4],\n",
    "                   showfliers=False, patch_artist=True,\n",
    "                   labels = data_opp.columns.tolist(),\n",
    "                   medianprops = dict(color=\"black\",linewidth=1.5))  \n",
    "ax.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "colors = ['white','grey','white','grey']\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "    patch.set_edgecolor(\"black\")\n",
    "ax.set_title('Genes per model:')\n",
    "ax.set_xticklabels(rotation=-25, ha='left', labels = ['not on EuPathDB (de novo)','on EuPathDB (de novo)','not on EuPathDB (semi-curated)','on EuPathDB (semi-curated)'])\n",
    "plt.subplots_adjust(bottom = 0.35, right = 0.5)\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/Diamond_not_in_EuPathDB.svg\")\n",
    "fig.savefig(\"/home/mac9jc/paradigm/data/results/figures_for_reviewers/Diamond_not_in_EuPathDB.png\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-paradigm_env]",
   "language": "python",
   "name": "conda-env-.conda-paradigm_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
