{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.13.4\n"
     ]
    }
   ],
   "source": [
    "import cobra\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import glob\n",
    "from cobra import Model, Reaction, Metabolite\n",
    "import subprocess\n",
    "print (cobra.__version__) # Use Devel branch\n",
    "\n",
    "#metabolic_tasks_from_spreadsheet\n",
    "#gene_lists_to_sparse_matrix\n",
    "\n",
    "path = \"/Users/maureencarey/local_documents/work/comparative_parasite_models/paradigm\"\n",
    "data_path = \"/Users/maureencarey/local_documents/work/comparative_parasite_models/paradigm_preOct2018/data\"\n",
    "\n",
    "os.chdir(data_path)\n",
    "universal_model = cobra.io.load_json_model('universal_model_may2018.json')\n",
    "    \n",
    "for rxn in universal_model.reactions:\n",
    "    if rxn.id.startswith('EX_'):\n",
    "        rxn.lower_bound = -1000.\n",
    "        rxn.upper_bound = 1000.\n",
    "        \n",
    "for met in universal_model.metabolites:\n",
    "    if met.id.endswith('_e'):\n",
    "        if 'EX_'+met.id not in universal_model.reactions:\n",
    "            universal_model.add_boundary(met, type = 'exchange')\n",
    "\n",
    "for rxn in [r for r in universal_model.reactions if r.id.lower().startswith('biomass')]:\n",
    "    rxn.remove_from_model()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Helper functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# preliminary curation (iPfal17)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# THIS CELL ASSUMES preliminary_iPfal17_curation.ipynb WAS ALREADY RUN\n",
    "\n",
    "os.chdir(data_path+'/other_models')\n",
    "iPfal17 = cobra.io.read_sbml_model(\"iPfal17.xml\") # RUN iPfal17_curation_part1.ipyb\n",
    "iPfal18 = cobra.io.read_sbml_model(\"iPfal18_pregapfilling.xml\")\n",
    "iPfal17.solver = 'gurobi'\n",
    "iPfal18.solver = 'gurobi'\n",
    "\n",
    "pf_biomass = iPfal18.reactions.get_by_id('biomass').copy()\n",
    "pf_biomass_mets = [x.copy() for x in iPfal18.reactions.get_by_id('biomass').metabolites]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN biomass_comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# add curated reactions to reaction bag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN build_universal_reaction_bag"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# download genomes\n",
    "# in R\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map BiGG IDs a la Carve Me but de novo\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# build models on rivanna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SLURM SCRIPT NAME\n",
    "# build_de_novo_models.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# os.chdir(data_path+'/diy6_output')\n",
    "# # add _Jul2018_BiGG to DIY6_LmajorSD75\n",
    "# model_dict = dict()\n",
    "# skip = list()\n",
    "# for filename in glob.glob(os.path.join(data_path+'/diy6_output', '*.json')):\n",
    "#     if 'Jul2018_BiGG' in filename:\n",
    "#         key = filename.split('/')[len(filename.split('/'))-1]\n",
    "#         key = key[:-18] # get rid of _Jul2018_BiGG.json\n",
    "#         if key.startswith('DIY6_'):\n",
    "#             key = key[5:]\n",
    "#         model_dict[key] = cobra.io.load_json_model(filename)\n",
    "#         model_dict[key].solver = 'gurobi'\n",
    "# #     elif 'LmajorSD75' in filename:\n",
    "# #         key = filename.split('/')[len(filename.split('/'))-1]\n",
    "# #         key = key[:-5] # get rid of .json\n",
    "# #         if key.startswith('DIY6_'):\n",
    "# #             key = key[5:]\n",
    "# #         model_dict[key] = cobra.io.load_json_model(filename)\n",
    "# #         model_dict[key].solver = 'gurobi'\n",
    "#     else:\n",
    "#         skip.append(filename)\n",
    "# len(model_dict.keys())\n",
    "# model_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# matrix of reaction presence\n",
    "dict_o_lists = dict()\n",
    "for model_name, model in model_dict.items():\n",
    "    dict_o_lists[model_name] = [x.id for x in model.reactions] # get all reaction IDs\n",
    "unique_elements = set().union(*dict_o_lists.values()) # find all unique reeaction IDs\n",
    "df = pd.DataFrame(index = unique_elements,columns = dict_o_lists.keys()) # save as matrix/df\n",
    "for species,list_genes in dict_o_lists.items():\n",
    "    for x in unique_elements:\n",
    "        if x in list_genes:\n",
    "            y = 1\n",
    "        else:\n",
    "            y = 0\n",
    "        df[species][x] = y\n",
    "os.chdir(data_path)\n",
    "df.to_csv('reaction_matrix_july25.csv', header=True, index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # plot\n",
    "# proc=subprocess.Popen([\"Rscript\", \"/Users/maureencarey/local_documents/work/comparative_parasite_models/code/upsetR_models_uncurated.R\"])\n",
    "# # generating upsetR_all_models_OG_carvemeMAY25.png and others\n",
    "# SCRIPT NAME"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# add generic biomass function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# curated models\n",
    "os.chdir(data_path+'/other_models')\n",
    "chominis = cobra.io.read_sbml_model('chominis_vanee2010.xml')\n",
    "tgondii = cobra.io.read_sbml_model('tg_tymoshenko2015.xml')\n",
    "leish = cobra.io.read_sbml_model('iAC560_leishmania_Chavali2008.xml')\n",
    "pf_curated = cobra.io.read_sbml_model(\"iPfal18_pregapfilling.xml\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract biomass reactions\n",
    "pf_biomass_mets = [x.copy() for x in pf_curated.reactions.get_by_id('biomass').metabolites]\n",
    "chominis_biomass_mets = [x.copy() for x in chominis.reactions.get_by_id('R_biomass_target').metabolites]\n",
    "leish_biomass_mets = [x.copy() for x in leish.reactions.get_by_id('BIOMASS_LM3').metabolites]\n",
    "tgondii_biomass_mets = [x.copy() for x in tgondii.reactions.get_by_id('Biomass').metabolites]\n",
    "tgondii_also_add = [tgondii.reactions.get_by_id('DNA').copy(),tgondii.reactions.get_by_id('RNA').copy(),\n",
    "                    tgondii.reactions.get_by_id('GAM').copy(),tgondii.reactions.get_by_id('Acyl').copy(),\n",
    "                    tgondii.reactions.get_by_id('Proteins').copy(),tgondii.reactions.get_by_id('Lipids').copy(),\n",
    "                    tgondii.reactions.get_by_id('EMs').copy()]\n",
    "tgondii_also_add_mets = [x.copy() for x in [y.reactants for y in tgondii_also_add]]\n",
    "tgondii_also_add_mets = [item for sublist in tgondii_also_add_mets for item in sublist]\n",
    "\n",
    "# get biomass precursor names\n",
    "pf_biomass_mets_names = [x.name for x in pf_biomass_mets]\n",
    "chominis_biomass_mets_names = [x.name for x in chominis_biomass_mets]\n",
    "leish_biomass_mets_names = [x.name for x in leish_biomass_mets]\n",
    "tgondii_also_add_mets_names = [x.name for x in tgondii_also_add_mets]\n",
    "\n",
    "# make table of biomass precursor names\n",
    "max_length = max(len(pf_biomass_mets_names),len(chominis_biomass_mets_names),len(leish_biomass_mets_names),len(tgondii_also_add_mets_names))\n",
    "pf_biomass_mets_ids = sorted(pf_biomass_mets_names, key=str.lower) + ['' for _ in range(max_length-len(pf_biomass_mets_names))]\n",
    "for index, x in enumerate(pf_biomass_mets_ids):\n",
    "    if '_' in x: pf_biomass_mets_ids[index] = x.replace('_','-')\n",
    "chominis_biomass_mets_ids = sorted(chominis_biomass_mets_names, key=str.lower) + ['' for _ in range(max_length-len(chominis_biomass_mets_names))]\n",
    "leish_biomass_mets_ids = sorted(leish_biomass_mets_names, key=str.lower) + ['' for _ in range(max_length-len(leish_biomass_mets_names))]\n",
    "for index, x in enumerate(leish_biomass_mets_ids):\n",
    "    if x.startswith('M_'): \n",
    "        leish_biomass_mets_ids[index] = x[2:]\n",
    "        x = leish_biomass_mets_ids[index]\n",
    "    if x.endswith('_'): \n",
    "        leish_biomass_mets_ids[index] = x[:-1]\n",
    "        x = leish_biomass_mets_ids[index]\n",
    "    if '__' in x: \n",
    "        leish_biomass_mets_ids[index] = x.split('__')[0]\n",
    "        x = leish_biomass_mets_ids[index]\n",
    "    if '_' in x and x.split('_')[0] == x.split('_')[1]: \n",
    "        leish_biomass_mets_ids[index] = x.split('_')[0]\n",
    "        x = leish_biomass_mets_ids[index]\n",
    "    if 'Cysteine' not in x and '_C' in x: \n",
    "        leish_biomass_mets_ids[index] = x.split('_C')[0]\n",
    "        x = leish_biomass_mets_ids[index]\n",
    "    if '_' in x: leish_biomass_mets_ids[index] = x.replace('_','-')\n",
    "tgondii_biomass_mets_ids = sorted(tgondii_also_add_mets_names, key=str.lower) + ['' for _ in range(max_length-len(tgondii_also_add_mets_names))]\n",
    "df = pd.DataFrame( {'iPfal17': pf_biomass_mets_ids,'C. hominis 2010': chominis_biomass_mets_ids,\n",
    "     'T. gondii 2015': tgondii_biomass_mets_ids, 'Leishmania 2008': leish_biomass_mets_ids })\n",
    "df.to_csv('biomass_components_july23.csv', header=True, index=True)\n",
    "\n",
    "# get biomass precursor ids\n",
    "pf_biomass_mets_ids = [x.id for x in pf_biomass_mets]\n",
    "chominis_biomass_mets_ids = [x.id for x in chominis_biomass_mets]\n",
    "leish_biomass_mets_ids = [x.id for x in leish_biomass_mets]\n",
    "tgondii_also_add_mets_ids = [x.id for x in tgondii_also_add_mets]\n",
    "lower_pf = set([met.lower().split('_')[0] for met in pf_biomass_mets_ids])\n",
    "lower_lm = set([met.lower().split('_')[0] for met in leish_biomass_mets_ids])\n",
    "lower_ch = set([met.lower().split('_')[0] for met in chominis_biomass_mets_ids])\n",
    "biomass_mets = list(lower_ch.intersection(lower_lm).intersection(lower_pf))\n",
    "biomass_mets.append('biomass')\n",
    "biomass_mets.append('protein') \n",
    "biomass_mets.append('lipid') # plus DAG begause in all reactions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for species, model in model_dict.items():\n",
    "    #biomass\n",
    "    new_rxn = Reaction()\n",
    "    met_dict = dict()\n",
    "    for met in pf_curated.reactions.biomass.metabolites:\n",
    "        if met.id.lower().split('_')[0] in biomass_mets:\n",
    "            met_dict[met] = pf_curated.reactions.biomass.metabolites[met]\n",
    "    new_rxn.add_metabolites(met_dict)\n",
    "    new_rxn.name = 'generic_biomass'\n",
    "    new_rxn.id = 'generic_biomass'\n",
    "    new_rxn.lower_bound = 0.\n",
    "    new_rxn.upper_bound = 1000.\n",
    "    model.add_reactions([new_rxn])\n",
    "\n",
    "    # protein production, transport, exchange\n",
    "    # biomass 'sink', lipid exchange\n",
    "    for x in ['Protein', 'Protein_t','Protein_ex','DM_biomass_c','Lipid_ex','EX_protein_t_e']:\n",
    "        new_rxn = pf_curated.reactions.get_by_id(x).copy()\n",
    "        for met in new_rxn.metabolites:\n",
    "            if met.id not in [x.id for x in model.metabolites]:\n",
    "                model.add_metabolites([met])\n",
    "        model.add_reactions([new_rxn])\n",
    "\n",
    "    # lipid production\n",
    "    for r in pf_curated.metabolites.get_by_id('all_dgl_c').reactions:\n",
    "        if r.id != 'Lipid_prod':\n",
    "            new_rxn = pf_curated.reactions.get_by_id(r.id).copy()\n",
    "            for met in new_rxn.metabolites:\n",
    "                if met.id not in [x.id for x in model.metabolites]:\n",
    "                    model.add_metabolites([met])\n",
    "            model.add_reactions([new_rxn])\n",
    "    new_rxn = Reaction()\n",
    "    met_dict = {pf_curated.metabolites.get_by_id('all_dgl_c'): -1,\n",
    "               pf_curated.metabolites.get_by_id('lipid_c'): +1}\n",
    "    new_rxn.add_metabolites(met_dict)\n",
    "    new_rxn.name = 'Lipid_prod'\n",
    "    new_rxn.id = 'Lipid_prod'\n",
    "    new_rxn.lower_bound = 0.\n",
    "    new_rxn.upper_bound = 1000.\n",
    "    for met in new_rxn.metabolites:\n",
    "        if met.id not in [x.id for x in model.metabolites]:\n",
    "            model.add_metabolites([met])\n",
    "    model.add_reactions([new_rxn])\n",
    "\n",
    "    model.objective = model.reactions.get_by_id('generic_biomass')\n",
    "    \n",
    "    model_dict[species] = model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove unused metabolites\n",
    "\n",
    "# add exchange reactions for all extracellular metabolites to accelerate gapfilling\n",
    "for species, model in model_dict.items():\n",
    "    for met in model.metabolites:\n",
    "        if met.id.endswith('_e'):\n",
    "            if 'EX_'+met.id not in model.reactions:\n",
    "                model.add_boundary(met, type = 'exchange')\n",
    "    model_dict[species] = model\n",
    "    \n",
    "for rxn in universal_model.reactions:\n",
    "    if rxn.id.startswith('EX_'):\n",
    "        rxn.lower_bound = -1000.\n",
    "        rxn.upper_bound = 1000.\n",
    "        \n",
    "for met in universal_model.metabolites:\n",
    "    if met.id.endswith('_e'):\n",
    "        if 'EX_'+met.id not in universal_model.reactions:\n",
    "            universal_model.add_boundary(met, type = 'exchange')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(data_path+'/diy7bm_output')\n",
    "for species,model in model_dict.items():\n",
    "    if 'Lipid_prod' not in [x.id for x in model.reactions]:\n",
    "        print('missing lipid')\n",
    "        print(species)\n",
    "    if 'Growth' in [x.id for x in model.reactions]:\n",
    "        print('has growth rxn')\n",
    "        print(species)\n",
    "    # cobra.io.save_json_model(model, \"DIY7bm_\"+species+\".json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# orthology transformation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check next cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(data_path)\n",
    "\n",
    "iPfal18 = cobra.io.read_sbml_model(\"./other_models/iPfal18_pregapfilling.xml\")\n",
    "# mapping = pd.read_table(\"./orthology_mapping_jun22/pf3d7_plasmodb.txt\")\n",
    "# # modify spcies column in mapping dataframe to map to filenames NECESSARY IF READING TXT MAPPING FILE\n",
    "# file_names = [*model_dict] ## CHECK\n",
    "# file_names = [x.split('_')[0] for x in file_names] ## CHECK\n",
    "# mapping['[Organism]'] = [x.replace(\". \", \"\") for x in mapping['[Organism]']]\n",
    "# mapping['[Organism]']  = [x.replace(\"strain\", \"\") for x in mapping['[Organism]'] ]\n",
    "# mapping['[Organism]']  = [x.replace(\" \", \"\") for x in mapping['[Organism]'] ]\n",
    "# # mapping['[Organism]']  = [x+'_May20.tsv' for x in mapping['[Organism]'] ]\n",
    "# mapping['[Organism]']  = [x.replace(\"Pfragilenilgiri\", \"PfragileNilgiri\") for x in mapping['[Organism]'] ]\n",
    "# mapping['[Organism]']  = [x.replace(\"PknowlesiMalayanStrainPk1A\", \"PknowlesiMalayanPk1A\") for x in mapping['[Organism]'] ]\n",
    "# mapping['[Organism]']  = [x.replace(\"PvivaxSal-1\", \"PvivaxSal1\") for x in mapping['[Organism]'] ]\n",
    "# if [] != [x for x in mapping['[Organism]'] if x not in set(mapping['[Organism]'] ).intersection(set(file_names))]:\n",
    "#     print('filename issue')\n",
    "# mapping.to_csv('plasmodium_gene_mapping.csv', header=True, index=True)\n",
    "mapping = pd.read_csv('plasmodium_gene_mapping.csv')\n",
    "\n",
    "add_these = ['pe_prod1', 'pe_prod10', 'pe_prod11', 'pe_prod12', 'pe_prod13', 'pe_prod14',\n",
    "             'pe_prod15', 'pe_prod16', 'pe_prod17', 'pe_prod18', 'pe_prod19', 'pe_prod20', \n",
    "             'pe_prod21', 'pe_prod22','pe_prod2','pe_prod3','pe_prod4','pe_prod5',\n",
    "             'pe_prod6','pe_prod7','pe_prod8', 'pe_prod9', 'pg_prod1', 'pg_prod2', 'pg_prod3',\n",
    "             'pg_prod4', 'pg_prod5', 'pg_prod6', 'pg_prod7','pi_prod1', 'pi_prod2', \n",
    "             'pi_prod3','pi_prod4',#'progly_t', \n",
    "             'ps_prod1', 'ps_prod2', 'ps_prod3', \n",
    "             'ps_prod4', 'ps_prod5','ps_prod6', 'ps_prod7','lipid1', 'lipid10', 'lipid2', \n",
    "             'lipid3','lipid4', 'lipid5', 'lipid6', 'lipid7', 'lipid8','lipid9','PA120tap', \n",
    "             'PA140tap', 'PA141tap', 'PA160tap', 'PA161tap', 'PA180tap', 'PA181tap',\n",
    "             '2agpe120_LSQBKT_e_RSQBKT__t', '2agpe140_LSQBKT_e_RSQBKT__t', \n",
    "             '2agpe141_LSQBKT_e_RSQBKT__t','2agpe160_LSQBKT_e_RSQBKT__t', \n",
    "             '2agpe161_LSQBKT_e_RSQBKT__t', '2agpe180_LSQBKT_e_RSQBKT__t',\n",
    "             '2agpe181_LSQBKT_e_RSQBKT__t', '2agpg120_LSQBKT_e_RSQBKT__t', \n",
    "             '2agpg140_LSQBKT_e_RSQBKT__t','2agpg141_LSQBKT_e_RSQBKT__t', \n",
    "             '2agpg160_LSQBKT_e_RSQBKT__t', '2agpg161_LSQBKT_e_RSQBKT__t',\n",
    "             '2agpg180_LSQBKT_e_RSQBKT__t', '2ddecg3p_LSQBKT_e_RSQBKT__t','acoa_prod1',\n",
    "             'acoa_prod10', 'acoa_prod11', 'acoa_prod12', 'acoa_prod13', 'acoa_prod14', \n",
    "             'acoa_prod15','acoa_prod16','acoa_prod2', 'acoa_prod3', 'acoa_prod4', \n",
    "             'acoa_prod5', 'acoa_prod6', 'acoa_prod7', 'acoa_prod8', 'acoa_prod9', \n",
    "             'acylpg_prod1', 'acylpg_prod2', 'acylpg_prod3', 'acylpg_prod4', 'acylpg_prod5',\n",
    "             'acylpg_prod6','acylpg_prod7',  'dolichol_t',#'TTDCAt',\n",
    "             'XOLEST2te','PCt',\n",
    "             'pail_uptake','pc_prod', 'Htap',#'FEROpp',\n",
    "             'Htmt','2agpg181_LSQBKT_e_RSQBKT__t',\n",
    "             'CHSTEROLt','HBtr','SM_host','HMBZex','Hfv','superox_ex']\n",
    "# bm models already have biomass_generic 'protein_t_e', 'dgl_prod1','dgl_prod2', \n",
    "# 'Lipid_prod','dgl_prod3', 'dgl_prod4', 'dgl_prod5','dgl_prod6', 'dgl_prod7','Protein',\n",
    "\n",
    "pf_model_dict = dict()\n",
    "os.chdir(data_path+\"/diy7bm_output\")\n",
    "for filename in glob.glob(os.path.join(data_path+\"/diy7bm_output\", '*.json')):\n",
    "    key = filename.split('/')[len(filename.split('/'))-1]\n",
    "    key = key[:-5]\n",
    "    if key.startswith('DIY7bm_P'):\n",
    "        key = key[7:]\n",
    "        if key != 'PconfusumCUL13' and key != 'PneurophiliaMK1':\n",
    "            pf_model_dict[key] = cobra.io.load_json_model(filename) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = ['species','starting_genes','reactions_added', 'mets_added','genes_added']\n",
    "modifications_ortho = pd.DataFrame(index = pf_model_dict.keys(), columns=columns)\n",
    "os.chdir(data_path+'/diy_ortho_output')\n",
    "model_pf = iPfal17\n",
    "# keep bounds from universal model\n",
    "for species, model in pf_model_dict.items():\n",
    "    print(species)\n",
    "    x1 = len(model.reactions)\n",
    "    y1 = len(model.genes)\n",
    "    z1 = len(model.metabolites)\n",
    "\n",
    "    for x in add_these:\n",
    "        rxn = iPfal18.reactions.get_by_id(x).copy()\n",
    "        if rxn.gene_reaction_rule:\n",
    "            print(rxn.gene_reaction_rule)\n",
    "        for met in rxn.metabolites:\n",
    "            if met.id not in [m.id for m in model.metabolites]:\n",
    "                model.add_metabolites([met.copy()])\n",
    "        model.add_reactions([rxn])\n",
    "\n",
    "    rxn = iPfal18.reactions.get_by_id('biomass').copy()\n",
    "    for met in rxn.metabolites:\n",
    "        if met.id not in [m.id for m in model.metabolites]:\n",
    "            model.add_metabolites([met.copy()])\n",
    "    model.add_reactions([rxn])\n",
    "\n",
    "    x2 = len(model.reactions)\n",
    "    y2 = len(model.genes)\n",
    "    z2 = len(model.metabolites)\n",
    "\n",
    "    species_specific_mapping = mapping[mapping['[Organism]'] == species]\n",
    "    for index, row in species_specific_mapping.iterrows():\n",
    "        new_gene = row['[Gene ID]']\n",
    "        if ',' not in row['[Input Ortholog(s)]']:\n",
    "            gene = row['[Input Ortholog(s)]'].strip()\n",
    "            if gene in [x.id for x in model_pf.genes]:\n",
    "                gene = model_pf.genes.get_by_id(gene)\n",
    "                rxn_to_add = gene.reactions.copy()\n",
    "                for rxn in rxn_to_add:\n",
    "                    if rxn.id not in [x.id for x in model.reactions]:\n",
    "                        rxn2 = rxn.copy()\n",
    "                        test = len(model.reactions)\n",
    "                        model.add_reactions([rxn2])\n",
    "                        model.reactions.get_by_id(rxn2.id).gene_reaction_rule = new_gene\n",
    "                    else: # reaction is in model\n",
    "                        if new_gene not in model.reactions.get_by_id(rxn.id).gene_reaction_rule:\n",
    "                            if model.reactions.get_by_id(rxn.id).gene_reaction_rule == '':\n",
    "                                model.reactions.get_by_id(rxn.id).gene_reaction_rule = new_gene\n",
    "                            else: model.reactions.get_by_id(rxn.id).gene_reaction_rule = \\\n",
    "                                model.reactions.get_by_id(rxn.id).gene_reaction_rule + ' or ' +new_gene\n",
    "#                         else:print('gene already there')\n",
    "        else:\n",
    "            for i in range(0,len(row['[Input Ortholog(s)]'].split(', ')),1):\n",
    "                gene = row['[Input Ortholog(s)]'].split(', ')[i].strip()\n",
    "                if gene in [x.id for x in model_pf.genes]:\n",
    "                    gene = model_pf.genes.get_by_id(gene)\n",
    "                    rxn_to_add = gene.reactions.copy()\n",
    "                    for rxn in rxn_to_add:\n",
    "                        if rxn.id not in [x.id for x in model.reactions]:\n",
    "                            rxn2 = rxn.copy()\n",
    "                            test = len(model.reactions)\n",
    "                            model.add_reactions([rxn2])\n",
    "                            model.reactions.get_by_id(rxn2.id).gene_reaction_rule = new_gene\n",
    "                        else:\n",
    "                            if new_gene not in model.reactions.get_by_id(rxn.id).gene_reaction_rule:\n",
    "                                if model.reactions.get_by_id(rxn.id).gene_reaction_rule == '':\n",
    "                                    model.reactions.get_by_id(rxn.id).gene_reaction_rule = new_gene\n",
    "                                else: model.reactions.get_by_id(rxn.id).gene_reaction_rule = \\\n",
    "                                    model.reactions.get_by_id(rxn.id).gene_reaction_rule + ' or ' +new_gene\n",
    "#                             else: #gene already there\n",
    "    if len(model.reactions) != len(set(model.reactions)):\n",
    "        print('duplicate reactions')\n",
    "\n",
    "    if species.startswith('Pfalciparum3D7'):\n",
    "        s = 1\n",
    "    else:\n",
    "        gene_list = [x for x in model.genes if x.id.startswith('PF3D7_')]\n",
    "        cobra.manipulation.delete.remove_genes(model, gene_list, remove_reactions=True)\n",
    "        gene_list = [x for x in model.genes if x.id in ['mal_mito_1','mal_mito_2','mal_mito_3']]\n",
    "        cobra.manipulation.delete.remove_genes(model, gene_list, remove_reactions=True)\n",
    "        \n",
    "    x3 = len(model.reactions)\n",
    "    y3 = len(model.genes)\n",
    "    z3 = len(model.metabolites)\n",
    "\n",
    "    modifications_ortho.species.loc[species] = species\n",
    "    modifications_ortho.starting_genes.loc[species] = y1\n",
    "    modifications_ortho.reactions_added.loc[species] = len(model.reactions) - x1\n",
    "    modifications_ortho.mets_added.loc[species] = len(model.metabolites) - z1\n",
    "    modifications_ortho.genes_added.loc[species] = len(model.genes) - y1\n",
    "    \n",
    "    # add exchanges in preparation for gapfilling\n",
    "    l = list()\n",
    "    for rxn in model.reactions:\n",
    "        for met in rxn.metabolites:\n",
    "            if met.id.endswith('_e'):\n",
    "                l.append(met)\n",
    "    for met in l:\n",
    "        if 'EX_'+met.id not in [r.id for r in model.reactions]:\n",
    "            model.add_boundary(met, type = \"exchange\")\n",
    "            print(model.reactions.get_by_id('EX_'+met.id).bounds)\n",
    "    \n",
    "    model.objective = 'biomass'\n",
    "    pf_model_dict[species] = model\n",
    "    cobra.io.save_json_model(model, \"DIY_ortho_\"+species+\".json\")\n",
    "\n",
    "os.chdir(data_path)\n",
    "modifications_ortho.to_csv('orthology_modifications_plasmodium_july25.csv')\n",
    "         \n",
    "for species, model in pf_model_dict.items():\n",
    "    print(species)\n",
    "    if species != 'Pfalciparum3D7':\n",
    "        for x in model.genes:\n",
    "            if x.id.startswith('PF3D7'):\n",
    "                print(x.id) # SHOULD HAVE NONE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "START HERE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['PblacklockiG01', 'PfragileNilgiri', 'Pfalciparum3D7', 'PinuiSanAntonio1', 'PbillcollinsiG01', 'PvivaxP01', 'PcynomolgiB', 'Pyoeliiyoelii17X', 'PbergheiANKA', 'PgaboniG01', 'Pyoeliiyoelii17XNL', 'PadleriG01', 'PpraefalciparumG01', 'PreichenowiCDC', 'PyoeliiyoeliiYM', 'PrelictumSGS1-like', 'Pgallinaceum8A', 'PmalariaeUG01', 'PcynomolgiM', 'PfalciparumIT', 'Pvinckeivinckeivinckei', 'PcoatneyiHackeri', 'PknowlesiH', 'PvinckeipetteriCR', 'PreichenowiG01', 'PovalecurtisiGH01', 'PvivaxSal1', 'PgaboniSY75', 'PknowlesiMalayanPk1A', 'Pchabaudichabaudi'])"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pf_model_dict = dict()\n",
    "os.chdir(data_path+\"/diy_ortho_output\")\n",
    "for filename in glob.glob(os.path.join(data_path+\"/diy_ortho_output\", '*.json')):\n",
    "    key = filename.split('/')[len(filename.split('/'))-1]\n",
    "    key = key[:-5]\n",
    "    key = key[10:]\n",
    "    pf_model_dict[key] = cobra.io.load_json_model(filename) \n",
    "pf_model_dict.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# test_model = pf_model_dict['PvivaxSal1']\n",
    "\n",
    "# atp = test_model.metabolites.get_by_id('atp_c')\n",
    "# if 'EX_atp_c' not in [r.id for r in test_model.reactions]:\n",
    "#     atp_production = test_model.add_boundary(atp, type=\"exchange\", reaction_id='atp_production')\n",
    "#     test_model.objective = atp_production\n",
    "# else:\n",
    "#     print('already there')\n",
    "#     test_model.objective = test_model.reactions.get_by_id('EX_atp_c')\n",
    "\n",
    "# test_model.solver = 'gurobi'\n",
    "# test_model.solver.configuration.timeout = 3600 #360 worked for one iteration, not for 10\n",
    "# if test_model.slim_optimize() <0.1:\n",
    "#     result2 = gapfill2(test_model,universal= universal_model, \n",
    "#                       demand_reactions = False, exchange_reactions = True, iterations=10)\n",
    "# else:\n",
    "#     result2 = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SKIPT TODAY\n",
    "# FIND/ REMOVE DUPLICATE REACTIONS\n",
    "duplicates = dict()\n",
    "for species, model in pf_model_dict.items():\n",
    "\n",
    "    temp_dict = dict() # get products and reactants for every reaction\n",
    "    for rxn in model.reactions:\n",
    "        rxn_dict = dict()\n",
    "        check_rxn_products = rxn.products\n",
    "        check_rxn_reactants = rxn.reactants\n",
    "        rxn_dict['reactants'] = [x.id for x in check_rxn_reactants]\n",
    "        rxn_dict['products'] = [x.id for x in check_rxn_products]\n",
    "        temp_dict[rxn.id] = rxn_dict\n",
    "        \n",
    "        for key in temp_dict.keys():\n",
    "            if key != rxn.id:\n",
    "                if temp_dict[rxn.id]['reactants'] == temp_dict[key]['reactants'] and \\\n",
    "                temp_dict[rxn.id]['products'] == temp_dict[key]['products']:\n",
    "                    if rxn.id not in duplicates.keys():\n",
    "                        duplicates[rxn.id] = key\n",
    "                    elif duplicates[rxn.id] == key or key in duplicates[rxn.id]:\n",
    "                        continue\n",
    "                    else:\n",
    "                         duplicates[rxn.id] = duplicates[rxn.id]+', '+key\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SKIPT TODAY\n",
    "os.chdir(data_path)\n",
    "universal_model = cobra.io.load_json_model('universal_model_may2018.json')\n",
    "for rxn in pf_curated.reactions:\n",
    "    if rxn.id not in [r.id for r in universal_model.reactions]:\n",
    "        for met in rxn.metabolites:\n",
    "            if met.id not in [m.id for m in universal_model.metabolites]:\n",
    "                universal_model.add_metabolites(met.copy())\n",
    "        universal_model.add_reactions([rxn.copy()])\n",
    "universal_model.remove_reactions(['biomass','biomass_s])\n",
    "\n",
    "cobra.io.save_json_model(universal_model, \"universal_model_extended_by_pf_july23.json\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "universal_model.remove_reactions(['biomass'])\n",
    "os.chdir(data_path)\n",
    "cobra.io.save_json_model(universal_model, \"universal_model_extended_by_pf_july23.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# curate to metabolic tasks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# curation_USE.xlsx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# [x.reaction for x in pf_curated.metabolites.get_by_id('dxyl5p_ap').reactions]\n",
    "# # pf_curated.metabolites.thmpp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = ['model','metabolite','metabolite group','can scavange?','synthesize?','essential?']\n",
    "auxotrophy_table = pd.DataFrame(columns=columns)\n",
    "\n",
    "met_group_dict = {\n",
    "    'glc__D':'carbon source',\n",
    "    'fru':'carbon source',\n",
    "    'cit':'carbon source',\n",
    "    'glyc':'carbon source',\n",
    "    'orn':'carbon source',\n",
    "    'arg__L':'carbon source, amino acid',\n",
    "    'ala__L':'carbon source, amino acid',\n",
    "    'his__L':'carbon source, amino acid',\n",
    "    'lys__L':'carbon source, amino acid',\n",
    "    'asp__L':'carbon source, amino acid',\n",
    "    'gln__L':'carbon source, amino acid',\n",
    "    'glu__L':'carbon source, amino acid',\n",
    "    'ser__L':'carbon source, amino acid',\n",
    "    'thr__L':'carbon source, amino acid',\n",
    "    'asn__L':'carbon source, amino acid',\n",
    "    'cys__L':'carbon source, amino acid',\n",
    "    'gly__L':'carbon source, amino acid',\n",
    "    'pro__L':'carbon source, amino acid',\n",
    "    'val__L':'carbon source, amino acid',\n",
    "    'ile__L':'carbon source, amino acid',\n",
    "    'leu__L':'carbon source, amino acid',\n",
    "    'met__L':'carbon source, amino acid',\n",
    "    'phe__L':'carbon source, amino acid',\n",
    "    'tyr__L':'carbon source, amino acid',\n",
    "    'trp__L':'carbon source, amino acid',\n",
    "    'thym':'nucleotide, pyrimidine',\n",
    "    'csn':'nucleotide, pyrimidine',\n",
    "    'ura':'nucleotide, pyrimidine',\n",
    "    'ade':'nucleotide, purines',\n",
    "    'gua':'nucleotide, purines',\n",
    "    'hxan':'nucleotide, purines',\n",
    "    'xan':'nucleotide, purines',\n",
    "    \n",
    "} \n",
    "\n",
    "met_long_name_dict = {\n",
    "    'glc__D':'glucose',\n",
    "    'fru': 'fructose',\n",
    "    'cit':'citric acid'\n",
    "    'glyc':'glycerol',\n",
    "    'orn':'ornithine'\n",
    "    'ala__L':'alanine',\n",
    "    'arg__L':'arginine',\n",
    "    'his__L':'histidine',\n",
    "    'lys__L':'lysine',\n",
    "    'asp__L':'aspartate',\n",
    "    'gln__L':'glutamine',\n",
    "    'glu__L':'glutamate',\n",
    "    'ser__L':'serine',\n",
    "    'thr__L':'threonine',\n",
    "    'asn__L':'asparagine',\n",
    "    'cys__L':'cysteine',\n",
    "    'gly__L':'glycine',\n",
    "    'pro__L':'proline',\n",
    "    'val__L':'valine',\n",
    "    'ile__L':'isoleucine',\n",
    "    'leu__L':'leucine',\n",
    "    'met__L':'methionine',\n",
    "    'phe__L':'phenylalanine',\n",
    "    'tyr__L':'tyrosine',\n",
    "    'trp__L':'tryptophane',\n",
    "    'thym':'thymine',\n",
    "    'csn':'cytosine',\n",
    "    'ura':'uracil',\n",
    "    'ade':'adenine',\n",
    "    'gua':'guanine',\n",
    "    'hxan':'hypoxanthine',\n",
    "    'xan':'xanthine'\n",
    "}\n",
    "\n",
    "for key, model in models_dict.items():\n",
    "    for met in met_group_dict.keys():\n",
    "        \n",
    "        met_group = met_group_dict[met]\n",
    "        met_long_name = met_long_name_dict[met]\n",
    "        \n",
    "        result0 = \n",
    "        result1 = \n",
    "        result2 = \n",
    "        \n",
    "        auxotrophy_table = auxotrophy_table.append({'model':key, \n",
    "                                                    'metabolite':met_long_name,\n",
    "                                                    'metabolite group': met_group,\n",
    "                                                    'can scavange?':result0,\n",
    "                                                    'synthesize?':result1,\n",
    "                                                    'essential?':result2}, \n",
    "                                                   ignore_index = True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PLASMODIUM MODELS DONT HAVE APPROPRIATE BIOMASS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/maureencarey/.virtualenvs/cobrapy/lib/python3.6/site-packages/pandas/core/indexing.py:194 \u001b[1;31mSettingWithCopyWarning\u001b[0m: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "genome_ids = pd.read_csv(\"~/local_documents/work/comparative_parasite_models/paradigm/model_generation/data/auxotrophies_mapping_to_genomeID.csv\",\n",
    "                        header = None).T\n",
    "new_header = genome_ids.iloc[0] \n",
    "genome_ids = genome_ids[1:] \n",
    "genome_ids.columns = new_header\n",
    "\n",
    "met_ids = pd.read_csv(\"~/local_documents/work/comparative_parasite_models/paradigm/model_generation/data/auxotrophies_mapping_to_metID.csv\")\n",
    "\n",
    "gapfilling_tasks = pd.read_excel(\"/Users/maureencarey/local_documents/work/metabolism_review/data/auxotrophies_references.xlsx\",\n",
    "                                 skiprows = 1)\n",
    "idx2 = met_ids['BiGG'].notnull()\n",
    "temp_met_ids = met_ids.loc[idx2]\n",
    "met_dict = pd.Series(temp_met_ids.BiGG.values, index = temp_met_ids.Metabolite).to_dict()\n",
    "\n",
    "# change met ids to model met ids if possible\n",
    "for x in gapfilling_tasks.index:\n",
    "    if gapfilling_tasks['Metabolite'].iloc[x] in met_dict.keys():\n",
    "        gapfilling_tasks['Metabolite'].iloc[x] = met_dict[gapfilling_tasks['Metabolite'].iloc[x]]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "tasks_dict = dict()\n",
    "for species, model in pf_model_dict.items():\n",
    "    if sum(genome_ids['strain_ID'].isin([species])): # does the model have any tasks?\n",
    "        idx = genome_ids.index[genome_ids['strain_ID'] == species].tolist()\n",
    "        sp = genome_ids.loc[idx]['species'] # get species name\n",
    "        species_string = sp[sp.index[0]]\n",
    "        \n",
    "        # tasks for that species\n",
    "        gf_species = gapfilling_tasks[[\"Metabolite\",species_string]]\n",
    "        indx = gf_species[species_string].notnull()\n",
    "        if sum(indx) >0:\n",
    "            gf_species.loc[indx] \n",
    "\n",
    "\n",
    "            # types of tasks\n",
    "            consumption_species = gf_species[gf_species[species_string].str.contains(\"uptake|essential|both|rescue\")==True]\n",
    "            production_species =gf_species[gf_species[species_string].str.contains(\"prod|both\")==True]\n",
    "\n",
    "            tasks_dict[species] = {\"consumption\":consumption_species['Metabolite'],\n",
    "                                  \"production\":production_species['Metabolite']}\n",
    "\n",
    "# tasks_dict # key = species, value = dictionary\n",
    "# tasks_dict[species][consumption] = dataframe of metabolite Ids with column name 'Metabolite'\n",
    "# tasks_dict[species][production] = dataframe of metabolite Ids with column name 'Metabolite'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pfalciparum3D7\n",
      "\n",
      "ser__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['PGCD', 'UREA', 'PSP_L', 'RBPC', 'PSERT', 'PRUK']\n",
      "\n",
      "his__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['THRHISHISt', 'PGCD', 'EX_glucan1500_e', 'EX_thrhishis_e', 'GLS2', 'GLS2e', 'GLNHISHISr', 'SPODMm', 'GLUCAN2abc', 'THRHISHISr', 'PSP_L', 'ACALD', 'THRA', 'EX_glnhishis_e', 'G3PT', 'CYS', 'ACChex', 'PSERT', 'O2t', 'PPDK', 'GLNHISHISt', 'HCO3E', 'LDH_D', 'THFGLUH']\n",
      "\n",
      "ile__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['FDH7', 'PGCD', 'PUTA3', 'r2110', 'UREA', 'EX_ileargile_e', 'ILEARGILEr', 'GNMT', 'SPODMm', 'CYSTGL', 'METMETILEt', 'aratyr2', 'ACOAD1z', 'PSP_L', 'RBPC', 'TYRTAi', 'CYSTS', 'SARCOX', 'GLYCK2', 'CYS', 'ACChex', 'PSERT', 'EX_metmetile_e', 'DHAD2', 'FldAct', 'ILEARGILEt', 'KARA2', 'PRUK', 'LDH_D', 'PYDXO', 'ACHBS', 'GLYALDDy', 'METMETILEr']\n",
      "consume\n",
      "possible_rxns\n",
      "ILEtec: ile__L_e --> ile__L_c\n",
      "\n",
      "atp\n",
      "add exchange\n",
      "consume\n",
      "possible_rxns\n",
      "r1116: atp_c --> atp_e\n",
      "\n",
      "citr__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['CBMK_m', 'DHDDS', 'OCTm', 'GLUDym', 'OCBT', 'ORNt3m', 'EX_argarg_e', 'SPODMm', 'ARGARGt', 'MALS', 'DEDOLR', 'DEDOLP2', 'ACGKm', 'HMR_4957', 'ORNTACim', 'ARGARGr', 'ACChex', 'ARGDr', 'PPDK', 'DEDOLP1', 'ACONTm', 'ICL', 'CITRtm', 'PYDXO', 'ACOTAim', 'SUCOAACTr', 'CBMKr', 'AGPRim']\n",
      "\n",
      "glc__D\n",
      "consume\n",
      "no gapfilling needed\n",
      "\n",
      "coa\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['FACOAE60', 'FACOAE80', 'FACOAE120', 'FACOAE100', 'FACOAE141', 'ARTCOAL1', 'ACChex']\n",
      "\n",
      "gly\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['PGCD', 'UREA', 'PSP_L', 'PSERT']\n",
      "consume\n",
      "possible_rxns\n",
      "\n",
      "\n",
      "fum\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['ALCD19y', 'MALS', 'ICL', 'PYDXO']\n",
      "\n",
      "glu__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['FDH7', 'ABUTD', 'FAO3', 'FAO10', 'PTRCTA', 'SPODMm', 'SO4CLtex', 'CLFORtex', 'ACOAD1z', 'SO4t_e', 'ACChex', 'PPDK', 'O2tm', 'PYDXO', 'SSALx']\n",
      "consume\n",
      "possible_rxns\n",
      "\n",
      "\n",
      "cit\n",
      "produce\n",
      "no gapfilling needed\n",
      "\n",
      "ala__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['FDH7', 'ABUTD', 'r0085', 'PTRCTA', 'SO4CLtex', 'CLFORtex', 'ACOAD1z', 'PDYXPT_c', 'SO4t_e', 'ACChex', 'r0156', 'LDH_D', 'PYDXO', 'PTRCOX1', 'SSALx']\n",
      "consume\n",
      "possible_rxns\n",
      "ALAPAT4te: ala__L_e --> ala__L_c\n",
      "\n",
      "trp__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['EX_trpmetval_e', 'ASPTA6', 'LEUSERTRPt', 'TRPTHRGLUr', 'GLUTRPALAt', 'EX_sertrphis_e', 'TRPTHRTYRt', 'TRPILETRPt', 'TRPSERTYRr', 'LEUSERTRPr', 'LYSVALTRPt', 'EX_glutrpala_e', 'EX_serargtrp_e', 'TRPILETRPr', 'TRPMETVALr', 'MACCOAT', 'PPS', 'PROTRPTHRt', 'LYSVALTRPr', 'MCA', 'EX_protrpthr_e', 'PDYXPT_c', 'EX_phesertrp_e', 'EX_trpaspasp_e', 'PHESERTRPr', 'SERARGTRPr', 'TRPSERTYRt', 'EX_trpthrtyr_e', 'TRPLYSr', 'EX_trpthrglu_e', 'SERTRPHISt', 'SERTRPHISr', 'EX_trpsertyr_e', 'MCITL2', 'ACChex', 'TRPASPASPr', 'PHESERTRPt', 'TRPASPASPt', 'TRPTHRTYRr', 'PROTEIN_BS', 'EX_leusertrp_e', 'CYSGLUTRPr', 'HCO3E', 'SERARGTRPt', 'ECOAH9ir', 'EX_lysvaltrp_e', 'TRPLYSt', 'EX_cysglutrp_e', 'PYDXO', 'PROTRPTHRr', 'EX_trplys_e', 'HACD9', 'CYSGLUTRPt', 'ACOAD20', 'EX_trpiletrp_e', 'TRPTHRGLUt', 'GLUTRPALAr', 'TRPMETVALt']\n",
      "\n",
      "met__L\n",
      "produce\n",
      "no gapfilling needed\n",
      "consume\n",
      "no gapfilling needed\n",
      "\n",
      "pydxn\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['FDH7', 'FACOAE60', 'GPDDA4', 'LPLIPAL2E181', 'UREA', 'ETHAAL', 'SO4CLtex', 'CLFORtex', 'LPLIPAL2G161', 'LPLIPAL2G120', 'DCATDc', 'PYDXOR', 'ARTCOAL1', 'G3PT', 'ARTPLM1', 'TSULR3e', 'LPLIPAL2G180', 'ACAFDOR', 'LPLIPAL2G141', 'LPLIPAL2G140', 'SO4t_e', 'SO3ti', 'ACChex', 'LPLIPAL2E180', 'HCO3E', 'OCTAt']\n",
      "\n",
      "succ\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['ALCD19y', 'MALS', 'ICL']\n",
      "\n",
      "pnto__R\n",
      "add exchange\n",
      "consume\n",
      "possible_rxns\n",
      "PNTOte: pnto__R_e --> pnto__R_c\n",
      "\n",
      "val__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['VALVALr', 'VALVALt', 'EX_valval_e', 'ACChex']\n",
      "\n",
      "fru\n",
      "add exchange\n",
      "consume\n",
      "possible_rxns\n",
      "FRUt1r: fru_e --> fru_c\n",
      "\n",
      "phe__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['EX_phephe_e', 'PHEPHEr', 'ACChex', 'PHEPHEt']\n",
      "\n",
      "ipdp\n",
      "add extracel met\n",
      "add exchange\n",
      "\n",
      "thr__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['PGCD', 'PUTA3', 'EX_hom__L_e', 'THRTHRARGt', 'UREA', 'FAO10', 'SPODMm', 'PSP_L', 'THRTHRARGr', 'RBPC', 'STPT1', 'THRS', 'CYS', 'ACChex', 'r1983', 'PSERT', 'PPDK', 'EX_arg__L_e', 'EX_thrthrarg_e', 'PRUK', 'LDH_D', 'PYDXO', 'THFGLUH']\n",
      "\n",
      "leu__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['LEULEUPEPT1tc', 'EX_leuleu_e', 'ACChex', 'LEULEULAPc']\n",
      "\n",
      "ptrc\n",
      "consume\n",
      "no gapfilling needed\n",
      "\n",
      "asp__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['ASPASNGLUr', 'ASPTA6', 'ASNN', 'FAO10', 'MALS', 'PDYXPT_c', 'ACChex', 'PPDK', 'O2tm', 'EX_aspasnglu_e', 'ICL', 'ASPASNGLUt', 'PYDXO', 'SUCOAACTr']\n",
      "\n",
      "pro__L\n",
      "produce\n",
      "no gapfilling needed\n",
      "consume\n",
      "no gapfilling needed\n",
      "\n",
      "tyr__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['PHETHPTOX', 'EX_tyrphetyr_e', 'ACChex', 'O2t', 'TYRPHETYRt', 'TYRPHETYRr', 'PYDXO', 'r0398']\n",
      "consume\n",
      "possible_rxns\n",
      "TYRt: tyr__L_e --> tyr__L_c\n",
      "\n",
      "arg__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['EX_argarg_e', 'ARGARGt', 'ARGARGr', 'ACChex']\n",
      "consume\n",
      "possible_rxns\n",
      "ARGt5r: arg__L_e --> arg__L_c\n",
      "\n",
      "lys__L\n",
      "add exchange\n",
      "produce\n",
      "solution\n",
      "['LYSLYSLYSr', 'LYSLYSLYSt', 'EX_lyslyslys_e', 'ACChex']\n",
      "\n",
      "pdx5p\n",
      "add extracel met\n",
      "add exchange\n",
      "\n",
      "gln__L\n",
      "add exchange\n",
      "produce\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-19-476ea00e50ba>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    125\u001b[0m                 \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreactions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_by_id\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'EX_'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mextracel_met\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupper_bound\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    126\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mslim_optimize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m0.01\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;31m# gapfill if can't produce\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 127\u001b[0;31m                     \u001b[0msolution\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpfba_gapfill_implementation\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0muniversal_model\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DM_'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mcyto_met\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    128\u001b[0m                     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'solution'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    129\u001b[0m                     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msolution\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-19-476ea00e50ba>\u001b[0m in \u001b[0;36mpfba_gapfill_implementation\u001b[0;34m(input_model, universal_model_ex, objective_reaction)\u001b[0m\n\u001b[1;32m     25\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0muniversal_model_pfba\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvariables\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'DM_'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'SK_'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'EX_'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 27\u001b[0;31m             \u001b[0muniversal_model_pfba\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobjective\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_linear_coefficients\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcoef\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     28\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     29\u001b[0m     \u001b[0;31m# get reaction IDs for reactions in the original model, then remove those reactions from the\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/cobrapy/lib/python3.6/site-packages/optlang/glpk_interface.py\u001b[0m in \u001b[0;36mset_linear_coefficients\u001b[0;34m(self, coefficients)\u001b[0m\n\u001b[1;32m    367\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mproblem\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    368\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mvariable\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcoefficient\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcoefficients\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 369\u001b[0;31m                 \u001b[0mglp_set_obj_coef\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mproblem\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mproblem\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariable\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_index\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfloat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcoefficient\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    370\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_expression_expired\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    371\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.virtualenvs/cobrapy/lib/python3.6/site-packages/optlang/glpk_interface.py\u001b[0m in \u001b[0;36m_index\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     91\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     92\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mproblem\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 93\u001b[0;31m             \u001b[0mi\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mglp_find_col\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mproblem\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mproblem\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     94\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mi\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     95\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "from cobra.flux_analysis.parsimonious import add_pfba\n",
    "\n",
    "def flatten_mixed_list(list_of_interest):\n",
    "    new_list = list()\n",
    "    for x in list_of_interest:\n",
    "        if isinstance(x,list):\n",
    "            new_list.extend(x)\n",
    "        else:\n",
    "            new_list.append(x)\n",
    "    return(new_list)\n",
    "\n",
    "# pFBA based gapfilling, implementation from Greg Medlock\n",
    "def pfba_gapfill_implementation(input_model, universal_model_ex, objective_reaction):\n",
    "    # objective_reaction is a reaction id\n",
    "    \n",
    "    universal_model_pfba = universal_model_ex.copy()\n",
    "    add_pfba(universal_model_pfba)\n",
    "    # penalize adding demand reactions\n",
    "    coef = universal_model_pfba.objective.get_linear_coefficients(universal_model_pfba.variables)\n",
    "    for key,value in coef.items():\n",
    "        if key.name.startswith('DM_') or key.name.startswith('SK_'):\n",
    "            coef[key] = 1000.\n",
    "        elif key.name.startswith('EX_'):\n",
    "            coef[key] = 50.\n",
    "    for x in universal_model_pfba.variables: \n",
    "        if x.name.startswith('DM_') or x.name.startswith('SK_') or x.name.startswith('EX_'):\n",
    "            universal_model_pfba.objective.set_linear_coefficients(coef)\n",
    "    \n",
    "    # get reaction IDs for reactions in the original model, then remove those reactions from the\n",
    "    # universal and replace them with the reactions from the model (this resets the sum of flux \n",
    "    # penalty for the reactions in the original model)\n",
    "    rxns_to_remove = [rxn for rxn in universal_model_pfba.reactions if rxn.id \\\n",
    "                      in [rxn.id for rxn in input_model.reactions]]\n",
    "    universal_model_pfba.remove_reactions(rxns_to_remove)\n",
    "    universal_model_pfba.add_reactions([rxn for rxn in input_model.reactions])\n",
    "\n",
    "    # Set your desired bound on the original objective (here, biomass, which has been added to \n",
    "    # the universal model from the model to be gapfilled)\n",
    "    universal_model_pfba.reactions.get_by_id(objective_reaction).lower_bound = 0.05\n",
    "    \n",
    "    # optimize and grab active reactions. 1E-8 is an arbitrary threshold, and in my example nothing\n",
    "    # had fluxes between 1E-5 and 1E-10, suggesting the threshold isn't as big of a deal as in the\n",
    "    # MILP formulation. This code will also return flux through exchange reactions (and any reaction\n",
    "    # from the model that was not present in the universal; you'll have to find another way to filter these).\n",
    "    solution = universal_model_pfba.optimize()\n",
    "    # get_fluxes = set([r.id for r in universal_model_pfba.reactions]) - set([r.id for r in rxns_to_remove])\n",
    "    get_fluxes = set([r.id for r in universal_model_pfba.reactions]) - set([rxn.id for rxn in model.reactions])\n",
    "    add_reactions_to_model = [rxn for rxn in get_fluxes if abs(solution.x_dict[rxn]) > 1E-8]\n",
    "    return(add_reactions_to_model) \n",
    "\n",
    "#for species, model in model_dict.items():\n",
    "for species, model in pf_model_dict.items():\n",
    "\n",
    "    if species != 'Pfalciparum3D7':\n",
    "        continue\n",
    "    print(species)\n",
    "    model.solver = 'gurobi'\n",
    "    add_reactions_list = list()\n",
    "    add_mets_list = list()\n",
    "    \n",
    "    mets_to_prod = tasks_dict[species][\"production\"]\n",
    "    mets_to_consume = tasks_dict[species][\"consumption\"]\n",
    "    produce_met = list()\n",
    "    consume_met = list()\n",
    "    all_mets = list()\n",
    "    cannot_gapfill = list()\n",
    "    for met in mets_to_prod:\n",
    "        if met+'_c' in universal_model.metabolites:\n",
    "            produce_met.append(met)\n",
    "            all_mets.append(met)\n",
    "        else:\n",
    "            cannot_gapfill.append(met)\n",
    "    for met in mets_to_consume:\n",
    "        if met+'_c' in universal_model.metabolites:\n",
    "            consume_met.append(met)\n",
    "            all_mets.append(met)\n",
    "        else:\n",
    "            cannot_gapfill.append(met)\n",
    "    all_mets = list(set(all_mets))\n",
    "            \n",
    "\n",
    "    for met in all_mets:\n",
    "        print('')\n",
    "        print(met)\n",
    "        extracel_met = met+'_e'\n",
    "        cyto_met = met+'_c'\n",
    "        \n",
    "        with model:\n",
    "            \n",
    "            # make sure you are making any mets by reversing biomass production\n",
    "            model.reactions.get_by_id('biomass').lb = 0. \n",
    "            model.reactions.get_by_id('biomass').ub = 0.\n",
    "\n",
    "            # add metabolites if necessary\n",
    "            if cyto_met not in model.metabolites: # already filtered list for only mets_c in universal\n",
    "                print('add cyto met')\n",
    "                model.add_metabolites([universal_model.metabolites.get_by_id(cyto_met).copy()])\n",
    "                add_mets_list.append(universal_model.metabolites.get_by_id(cyto_met).copy())\n",
    "            if extracel_met not in model.metabolites:\n",
    "                print('add extracel met')\n",
    "                if extracel_met in universal_model.metabolites:\n",
    "                    model.add_metabolites([universal_model.metabolites.get_by_id(extracel_met).copy()])\n",
    "                    add_mets_list.append(universal_model.metabolites.get_by_id(extracel_met).copy())\n",
    "                else:\n",
    "                    met = Metabolite(extracel_met,\n",
    "                                     formula = universal_model.metabolites.get_by_id(cyto_met).formula,\n",
    "                                     name = universal_model.metabolites.get_by_id(cyto_met).name,\n",
    "                                     compartment='e')\n",
    "                    model.add_metabolites([met])\n",
    "                    add_mets_list.append(met)\n",
    "\n",
    "            # add objective reaction\n",
    "            model.add_boundary(model.metabolites.get_by_id(cyto_met), type = \"demand\")\n",
    "            model.objective = 'DM_'+cyto_met\n",
    "\n",
    "            # add exchange\n",
    "            if 'EX_'+extracel_met not in model.reactions: \n",
    "                print('add exchange')\n",
    "                model.add_boundary(model.metabolites.get_by_id(extracel_met), type=\"exchange\")\n",
    "                add_reactions_list.append(model.reactions.get_by_id('EX_'+extracel_met).copy())\n",
    "\n",
    "            if met in list(produce_met):\n",
    "                print('produce')\n",
    "                model.reactions.get_by_id('EX_'+extracel_met).lower_bound = 0.\n",
    "                model.reactions.get_by_id('EX_'+extracel_met).upper_bound = 0.\n",
    "                if model.slim_optimize() < 0.01: # gapfill if can't produce\n",
    "                    solution = pfba_gapfill_implementation(model, universal_model, 'DM_'+cyto_met)\n",
    "                    print('solution')\n",
    "                    print(solution)\n",
    "                    if len(solution) > 1:\n",
    "                        for rxn_id in solution:\n",
    "                            if rxn_id != 'DM_'+cyto_met:\n",
    "                                add_reactions_list.append(universal_model.reactions.get_by_id(rxn_id).copy())\n",
    "                    else:\n",
    "                        if solution[0] != 'DM_'+cyto_met:\n",
    "                            add_reactions_list.append(universal_model.reactions.get_by_id(solution[0]).copy())\n",
    "                else:\n",
    "                    print('no gapfilling needed')\n",
    "\n",
    "            if met in list(consume_met):\n",
    "                print('consume')\n",
    "                model.reactions.get_by_id('EX_'+extracel_met).lb = -1000.\n",
    "                model.reactions.get_by_id('EX_'+extracel_met).ub = 1000.\n",
    "                t = False\n",
    "                if model.slim_optimize() < 0.01: # optimize and no growth\n",
    "                    # add import reaction (don't use gapfill because we want an import, not biosynthetic solution)\n",
    "                    if extracel_met in universal_model.metabolites:\n",
    "                        extracellular_reactions = universal_model.metabolites.get_by_id(extracel_met).reactions\n",
    "                        intracellular_reactions = universal_model.metabolites.get_by_id(cyto_met).reactions\n",
    "                        possible_rxns = list(extracellular_reactions.intersection(intracellular_reactions))\n",
    "                        if len(possible_rxns) >0:\n",
    "                            if len(possible_rxns) > 1:\n",
    "                                rxn_keep = ''\n",
    "                                for rxn in possible_rxns:\n",
    "                                    if len(rxn.metabolites) == 2: # add reaction that has no other substrates\n",
    "                                        rxn_keep = rxn\n",
    "                                    else:\n",
    "                                        pass\n",
    "                                if rxn_keep != '':\n",
    "                                    add_reactions_list.append(rxn_keep)\n",
    "                                else: add_reactions_list.append(possible_rxns[0])\n",
    "                                print('possible_rxns')\n",
    "                                print(rxn_keep)\n",
    "                            else:\n",
    "                                add_reactions_list.append(possible_rxns[0])\n",
    "                                print('possible_rxns')\n",
    "                                print(possible_rxns[0])\n",
    "                        \n",
    "                        else:\n",
    "                            t = True # if no transport in universal_model, make transport rxn\n",
    "                    else:\n",
    "                        t = True # if extracel met isnt in universal_model, make transport rxn\n",
    "                else:\n",
    "                    print('no gapfilling needed')\n",
    "                    # WORKS\n",
    "\n",
    "                if t == True:\n",
    "                    reaction = Reaction(met.upper()+'t')\n",
    "                    reaction.name = met + 'transport'\n",
    "                    reaction.subsystem = 'Transport'\n",
    "                    reaction.lower_bound = -1000.\n",
    "                    reaction.upper_bound = 1000.\n",
    "                    reaction.add_metabolites({model.metabolites.get_by_id(extracel_met): -1.0,\n",
    "                                              model.metabolites.get_by_id(cyto_met): 1.0 })\n",
    "                    add_reactions_list.append(reaction)\n",
    "\n",
    "\n",
    "                # is it used in any reaction intracellularly? if not, cannot do anything about it, except curate in future\n",
    "\n",
    "        add_reactions_list = flatten_mixed_list(add_reactions_list)                                      \n",
    "\n",
    "#     model.add_reactions([list(set(add_mets_list))])\n",
    "#     model.add_reactions([list(set(add_reactions_list))])\n",
    "    print(\"---------------------------------------------------\")\n",
    "    print(species)\n",
    "    pritn('no. reactions added:')\n",
    "    print(len(set(add_reactions_list)))\n",
    "    pritn('no. metabolites added:')\n",
    "    print(len(set(add_mets_list)))\n",
    "    #model_dict[species] = model\n",
    "    # save models\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# phe__L\n",
    "# add exchange\n",
    "# produce\n",
    "# solution\n",
    "# ['EX_phephe_e', 'PHEPHEr', 'ACChex', 'PHEPHEt']\n",
    "# how to limit exchange reactions - specifically small peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# for key, model in models_dict.items():\n",
    "name_conv = dict(\n",
    "    'NparisiiERTm3':{'name':'Nematocida_parisii__ERTm3',\n",
    "                     'tissue_type':'broad','type_para':'Microsporidia',\n",
    "                     'host':'Caenorhabditis elegans','vector':''},\n",
    "     'Vculicisfloridensis':{'name':'Vavraia_culicis_floridensis',\n",
    "                            'tissue_type':'','type_para':'Microsporidia',\n",
    "                            'host':'mosquito', 'vector':},\n",
    "     'CveliaCCMP2878':{'name':'Chromera_velia_CCMP2878',\n",
    "                    'tissue_type':'','type_para':'chromerid',\n",
    "                    'host':'corals (symbiotic)','vector':}, \n",
    "     'PcynomolgiM': {'name':'Plasmodium_cynomolgi_M',\n",
    "                    'tissue_type':'RBC, liver','type_para':'apicomplexan',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TvaginalisG3': {'name':'Trichomonas_vaginalis_G3',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'PneurophiliaMK1':{'name':'Pseudoloma_neurophilia_MK1',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':},  \n",
    "     'TgondiiRH':{'name':'Toxoplasma_gondii_RH',\n",
    "                    'tissue_type':'brain, gut','type_para':'apicompelxan',\n",
    "                    'host':'cats, mammals','vector':}, \n",
    "     'Chominis30976': {'name':'Cryptosporidium_hominis_30976',\n",
    "                    'tissue_type':'gut','type_para':'apicomplexan',\n",
    "                    'host':'humans','vector':'none'}, \n",
    "     'BmicrotiRI': {'name':'Babesia_microti_RI',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':},  \n",
    "     'TgondiiARI': {'name':'Toxoplasma_gondii_ARI',\n",
    "                    'tissue_type':'','type_para':'apicomplexan',\n",
    "                    'host':'','vector':}, \n",
    "     'LspMARLEM2494':{'name':'unclassified_Leishmania_sp._MAR_LEM2494',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':}, \n",
    "     'LmajorFriedlin':{'name':'Leishmania_major_Friedlin',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':}, \n",
    "     'TcruziCLBrener':{'name':'Trypanosoma_cruzi_CL_Brener',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':}, \n",
    "     'Pyoeliiyoelii17X': {'name':'Plasmodium_yoelii_yoelii_17X',\n",
    "                    'tissue_type':'RBC, liver','type_para':'apicomplexan',\n",
    "                    'host':'mice','vector':'mosquito'}, \n",
    "     'EromaleaeSJ2008':{'name':'Encephalitozoon_romaleae_SJ_2008',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'PvivaxP01': {'name':'Plasmodium_vivax_P01',\n",
    "                    'tissue_type':'RBC, liver','type_para':'apicomplexan',\n",
    "                    'host':'humans','vector':'mosquito'}, \n",
    "     'EnecatrixHoughton':{'name':'Eimeria_necatrix_Houghton',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'HhammondiHH34': {'name':'Hammondia_hammondi_HH34',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'TgondiiVAND':{'name':'Toxoplasma_gondii_VAND',\n",
    "                    'tissue_type':'','type_para':'apicomplexan',\n",
    "                    'host':'','vector':}, \n",
    "     'CmeleagridisUKMEL1':'Cryptosporidium_meleagridis_UKMEL1', {'name':'Babesia_microti_RI',\n",
    "                    'tissue_type':'','type_para':'apicomplexan',\n",
    "                    'host':'','vector':}, \n",
    "     'TcruzimarinkelleiB7':'Trypanosoma_cruzi_marinkellei_B7', {'name':'Babesia_microti_RI',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':}, \n",
    "     'EcuniculiEC2':'Encephalitozoon_cuniculi_EC2', {'name':'Babesia_microti_RI',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'NceranaeBRL01':'Nosema_ceranae_BRL01', {'name':'Babesia_microti_RI',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':},  \n",
    "     'MdaphniaeUGP3':'Mitosporidium_daphniae_UGP3', {'name':'Babesia_microti_RI',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'AcastellaniiNeff':'Acanthamoeba_castellanii_Neff', {'name':'Babesia_microti_RI',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'EintestinalisATCC50506':'Encephalitozoon_intestinalis_ATCC_50506', {'name':'Babesia_microti_RI',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'TequiWA':{'name':'Theileria_equi_WA',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'LpanamensisMHOMPA94PSC1': {'name':'Leishmania_panamensis_MHOM/PA/94/PSC-1',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':}, \n",
    "     'LbraziliensisMHOMBR75M2903': {'name':'Leishmania_braziliensis_MHOM/BR/75/M2903',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':}, \n",
    "     'PvinckeipetteriCR': {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'apicomplexan',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'BbovisT2Bo': {'name':'Babesia_bovis_T2Bo',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'CsuisWienI': {'name':'Cystoisospora_suis_Wein_I',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'EhistolyticaHM1IMSS-B': {'name':'Entamoeba_histolytica_HM1I:MSS-B',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'GintestinalisAssemblageADH': {'name':'Giardia_intestinalis_Assemblage_A_DH',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'TgondiiGT1':  {'name':'Toxoplasma_gondii_GT1',\n",
    "                    'tissue_type':'','type_para':'apicomplexan',\n",
    "                    'host':'','vector':}, \n",
    "     'TvivaxY486':  {'name':'Trypanosoma_vivax_Y486',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':}, \n",
    "     'LtarentolaeParrotTarII':  {'name':'Leishmania_tarentolae_Parrot_TarII',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':}, \n",
    "     'PblacklockiG01':   {'name':'Plasmodium_blacklocki_G01',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'EcuniculiGBM1': {'name':'Encephalitozoon_cuniculi_GBM1',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'LpanamensisMHOMCOL81L13': {'name':'Leishmania_panamensis_MHOM/COL/81/L13',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'LbraziliensisMHOMBR75M2904': {'name':'Leishmania_braziliensis_MHOM/BR/75/M2904',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':''}, \n",
    "     'Pyoeliiyoelii17XNL': {'name':'Plasmodium_yoelii_yoelii_17XNL',\n",
    "                    'tissue_type':'RBC, liver','type_para':'apicomplexan',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'Cubiquitum39726': {'name':'Cryptosporidium_ubiquitum_39726',\n",
    "                    'tissue_type':'gut','type_para':'apicomplexan',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TparvaMuguga': {'name':'Theileria_parva_Muguga',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'TcruziSylvioX10-1': {'name':'Trypanosoma_cruzi_Sylvio_X10/1',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':''}, \n",
    "     'TcruziDm28c': {'name':'Trypanosoma_cruzi_Dm28c',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':''}, \n",
    "     'EhellemSwiss': {'name':'Encephalitozoon_hellem_Swiss',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'PmalariaeUG01':  {'name':'Plasmodium_malariae_UG01',\n",
    "                    'tissue_type':'','type_para':'apicomplexan',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TtheileriEdinburgh':{'name':'Trypanosoma_theileri_Edinburgh',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'EhistolyticaHM3IMSS': {'name':'Entamoeba_histolytica_HM3I:MSS',\n",
    "                    'tissue_type':'gut','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'EnuttalliP19': {'name':'Entamoeba_nuttalli_P19',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'GintestinalisAssemblageAWB': {'name':'Giardia_intestinalis_Assemblage_A_WB',\n",
    "                    'tissue_type':'gut','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'NausubeliERTm2': {'name':'Nematocida_ausubeli_ERTm2',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'PcoatneyiHackeri':  {'name':'Plasmodium_coatneyi_Hackeri',\n",
    "                    'tissue_type':'','type_para':'apicomplexan',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PfalciparumIT': {'name':'Plasmodium_falciparum_IT',\n",
    "                    'tissue_type':'RBC, liver','type_para':'apicomplexan',\n",
    "                    'host':'human','vector':'mosquito'}, \n",
    "     'BovataMiyake': {'name':'Babesia_ovata_Miyake',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'ThominisUnknown': {'name':'Trichomonas_hominis_Unknown',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'LarabicaLEM1108': {'name':'Leishmania_arabica_LEM1108',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':''}, \n",
    "     'LturanicaLEM423':{'name':'Leishmania_turanica_LEM423',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':''}, \n",
    "     'EmonterogeiiLV88': {'name':'Endotrypanum_monterogeii_LV88',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'EaedisUSNM41457': {'name':'Edhazardia_aedis_USNM_41457',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'EcuniculiEC3': {'name':'Encephalitozoon_cuniculi_EC3',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'BbigeminaBOND': {'name':'Babesia_bigemina_BOND',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'EbieneusiH348': {'name':'Enterocytozoon_bieneusi_H348',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'SneuronaSOSN1': {'name':'Sarcocystis_neurona_SO_SN1',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'TgondiiRUB': {'name':'Toxoplasma_gondii_RUB',\n",
    "                    'tissue_type':'gut, brain','type_para':'apicomplexan',\n",
    "                    'host':'cat, mammals','vector':''}, \n",
    "     'EmoshkovskiiLaredo': {'name':'Entamoeba_moshkovskii_Laredo',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'LmajorLV39c5': {'name':'Leishmania_major_LV39c5',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':''}, \n",
    "     'PovalecurtisiGH01': 'Plasmodium_ovale_curtisi_GH01', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'ChominisTU502_2012': {'name':'Cryptosporidium_hominis_TU502',\n",
    "                    'tissue_type':'gut','type_para':'apicomplexan',\n",
    "                    'host':'humans, mammals','vector':'non'}, \n",
    "     'VbrassicaformisCCMP3155': {'name':'Vitrella_brassicaformis_CCMP3155',\n",
    "                    'tissue_type':'','type_para':'chromerid',\n",
    "                    'host':'','vector':}, \n",
    "     'EinvadensIP1': {'name':'Entamoeba_invadens_IP1',\n",
    "                    'tissue_type':'gut','type_para':'',\n",
    "                    'host':'','vector':}, \n",
    "     'GintestinalisAssemblageBGS': {'name':'Giardia_intestinalis_AssemblageB_GS',\n",
    "                    'tissue_type':'gut','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'PfragileNilgiri': {'name':'Plasmodium_fragile_Nilgiri',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'NfowleriATCC30863': {'name':'Naegleria_fowleri_ATCC30863',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'TcongolenseIL3000': {'name':'Trypanosoma_congolense_IL3000',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':''}, \n",
    "     'Pvinckeivinckeivinckei': 'Plasmodium_vinckei_vinckei_vinckei', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TorientalisShintoku':'Theileria_orientalis_Shintoku', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'PvivaxSal1': 'Plasmodium_vivax_Sal1', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'NdisplodereJUm2807':'Nematocida_displodere_JUm2807', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'CmurisRN66': {'name':'Cryptosporidium_muris_RN66',\n",
    "                    'tissue_type':'gut','type_para':'apicomplexan',\n",
    "                    'host':'mouse, mammals','vector':'none'}, \n",
    "     'CparvumIowaII': {'name':'Cryptosporidium_parvum_Iowa_II',\n",
    "                    'tissue_type':'mammals','type_para':'apicomplexan',\n",
    "                    'host':'','vector':'none'}, \n",
    "     'VcorneaeATCC50505':'Vittaforma_corneae_ATCC_50505', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'LtropicaL590':'Leishmania_tropica_L590', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'GniphandrodesUnknown':'Gregarina_niphandrodes_Unknown', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'NparisiiERTm1':'Nematocida_parisii_ERTm1', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TgondiiTgCatPRC2':'Toxoplasma_gondii_TgCatPRC2', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'EhistolyticaHM1IMSS':'Entamoeba_histolytica_HM1IMSS', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'EpraecoxHoughton':'Eimeria_praecox_Houghton', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PreichenowiG01': 'Plasmodium_reichenowi_G01', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'SneuronaSN3':'Sarcocystis_neurona_SN3', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'SsalmonicidaATCC50377':'Spironucleus_salmonicida_ATCC_50377', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PpraefalciparumG01': 'Plasmodium_praefalciparum_G01', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TgondiiGAB2-2007-GAL-DOM2':'Toxoplasma_gondii_GAB2_2007_GAL_DOM2', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TrangeliSC58':'Trypanosoma_rangeli_SC58', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'CcayetanensisCHN_HEN01', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'ChominisTU502':'Cryptosporidium_hominis_TU502',{'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'ChominisUdeA01':'Cryptosporidium_hominis_UdeA01', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PcynomolgiB': 'Plasmodium_cynomolgi_B', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'OcolligataOC4':'Cyclospora_cayetanensis_CHN_HEN01', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TgondiiMAS':'Toxoplasma_gondii_MAS', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PgaboniG01': 'Plasmodium_gaboni_G01', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PreichenowiCDC': 'Plasmodium_reichenowi_CDC', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'NbombycisCQ1':'Nosema_bombycis_CQ1', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'BayalaiB08-376':'Blechomonas_ayalai_B08-376', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'NcaninumLIV':'Neospora_caninum_Liverpool', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TgondiiFOU':'Toxoplasma_gondii_FOU', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'LenriettiiLEM3045':'Leishmania_enriettii_LEM3045', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TannulataAnkara':'Theileria_annulata_Ankara', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TbruceiLister427':'Trypanosoma_brucei_Lister427', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TgondiiME49':'Toxoplasma_gondii_ME49', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PknowlesiMalayanPk1A': 'Plasmodium_knowlesi_Malayan_Pk1A', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'LmexicanaMHOMGT2001U1103':'Leishmania_mexicana_MHOM/GT/2001/U1103', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'LmajorSD75.1':'Leishmania_major_SD75.1', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'EmitisHoughton':'Eimeria_mitis_Houghton', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PrelictumSGS1-like': 'Plasmodium_relictum_SGS1-like', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TevansiSTIB805':'Trypanosoma_evansi_STIB805', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PgaboniSY75': 'Plasmodium_gaboni_SY75', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'CtyzzeriUGA55':'Cryptosporidium_tyzzeri_UGA55', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TbruceiTREU927':'Trypanosoma_brucei_TREU927', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'EmaximaWeybridge':'Eimeria_maxima_Weybridge', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TcruziSylvioX10-1-2012':'Trypanosoma_cruzi_Sylvio_X10/1-2012', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'EfalciformisBayerHaberkorn1970':'Eimeria_falciformis_Bayer_Haberkorn_1970', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'GintestinalisAssemblageBGS_B':'Giardia_intestinalis_AssemblageB_GS-B', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'CfelisWinnie':'Cytauxzoon_felis_Winnie', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'CfasciculataCfCl':'Crithidia_fasciculata_CfCl', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'EhellemATCC50504':'Encephalitozoon_hellem_ATCC50504', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'LgerbilliLEM452':'Leishmania_gerbilli_LEM452', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'LseymouriATCC30220':'Leptomonas_seymouri_ATCC_30220', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TgrayiANR4':'Trypanosoma_grayi_ANR4', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TbruceigambienseDAL972':'Trypanosoma_brucei_gambiense_DAL972',{'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PconfusumCUL13': 'Paratrypanosoma_confusum_CUL13', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'Pchabaudichabaudi': 'Plasmodium_chabaudi_chabaudi', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'EhistolyticaKU27':'Entamoeba_histolytica_KU27', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PyoeliiyoeliiYM': 'Plasmodium_yoelii_yoelii_YM', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'EtenellaHoughton':'Eimeria_tenella_Houghton', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PknowlesiH': 'Plasmodium_knowlesi_H', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TcruziCLBrenerEsmeraldo-like':'Trypanosoma_cruzi_CL_Brener_Esmeraldo-like', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PbergheiANKA': 'Plasmodium_berghei_ANKA', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'Tgondiip89':'Toxoplasma_gondii_p89', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'AalgeraePRA339':'Annacaliia_algerae_PRA339', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'LpyrrhocorisH10':'Leptomonas_pyrrhocoris_H10', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PinuiSanAntonio1': 'Plasmodium_inui_San_Antonio_1', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'TcruziCLBrenerNon-Esmeraldo-like':'Trypanosoma_cruzi_CL_Brener_Non-Esmeraldo-like', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'EacervulinaHoughton':'Eimeria_acervulina_Houghton', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'Pfalciparum3D7': 'Plasmodium_falciparum_3D7', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'EcuniculiEC1':'Encephalitozoon_cuniculi_EC1', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'LinfantumJPCM5':'Leishmania_infantum_JPCM5', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'Candersoni30847':'Cryptosporidium_andersoni_isolate_30847', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PadleriG01': 'Plasmodium_adleri_G01', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'Pgallinaceum8A': 'Plasmodium_gallinaceum_8A', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'EdisparSAW760':'Entamoeba_dispar_SAW760', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'EbrunettiHoughton':'Eimeria_brunetti_Houghton',{'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'LdonovaniBPK282A1':'Leishmania_donovani_BPK282A1', {'name':'Plasmodium_vinckei_petteri_CR',\n",
    "                    'tissue_type':'','type_para':'trypanosomatid',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'PbillcollinsiG01':  {'name':'Plasmodium_billcollinsi_G01',\n",
    "                    'tissue_type':'RBC, liver','type_para':'apicomplexan',\n",
    "                    'host':'','vector':'mosquito'}, \n",
    "     'AalgeraePRA109': {'name':'Anncaliia_algerae_PRA109',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'LaethiopicaL147':{'name':'Leishmania_aethiopica_L147',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'NausubeliERTm6':{'name':'Nematocida_ausubeli_ERTm6',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'TgondiiVEG': {'name':'Toxoplasma_gondii_VEG',\n",
    "                    'tissue_type':'','type_para':'apicomplexan',\n",
    "                    'host':'','vector':'none'}, \n",
    "     'Slophii42_110': {'name':'Spraguea_lophii_42_110',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':''}, \n",
    "     'EhistolyticaHM1IMSS-A': {'name':'Entamoeba_histolytica_HM1I:MSS-A',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'none'}, \n",
    "     'GintestinalisAssemblageEP15':{'name':'Giardia_intestinalis_Assemblage_E_P15',\n",
    "                    'tissue_type':'','type_para':'',\n",
    "                    'host':'','vector':'none'}, )\n",
    "\n",
    "\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
